<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Will Gammerdinger">
<meta name="author" content="Noor Sohail">
<meta name="author" content="Zhu Zhuo">
<meta name="author" content="James Billingsley">
<meta name="author" content="Shannan Ho Sui">
<meta name="dcterms.date" content="2025-07-22">

<title>Theory of PCA – Introduction to single-cell RNA-seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-e5cb4ee4193d6a2f28a249434e39fb42.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../schedule/schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/"> 
<span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/Intro-to-spatial-scRNA-seq"> 
<span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu"> 
<span class="menu-text">Contact us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/03_quality-control.html">Day 1 Self-learning:</a></li><li class="breadcrumb-item"><a href="../lessons/04_theory_of_PCA.html">Theory of PCA</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false">
 <span class="menu-text">Pre-reading:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Pre-reading_01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pre-reading 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Pre-reading_02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pre-reading 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Pre-reading_03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pre-reading 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/01_introduction-to-spatial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction to Spatial Single Cell RNA-sequencing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/02_preprocessing-of-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pre-processing of Spatial Data</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/03_quality-control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_theory_of_PCA.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Theory of PCA</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 2:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/05_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Normalization and Regressing Out Unwanted Variation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 2 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/07_clustering-quality-control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Clustering Quality Control</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/08_marker-identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Marker Identification</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/09_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Integration</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 3:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/10_spatially-derived-values.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Spatially Derived Values</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/11_deconvolution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Deconvolution</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 3 Self-learning</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/12_dge.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Differential Gene Expression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/Cheatsheet_seurat.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Seurat Cheatsheet</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="false">
 <span class="menu-text">Day 4:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/13_cell-cell-communication.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Cell-Cell Comunication</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#setting-up-to-calculate-eigenvalues-and-eigenvectors" id="toc-setting-up-to-calculate-eigenvalues-and-eigenvectors" class="nav-link" data-scroll-target="#setting-up-to-calculate-eigenvalues-and-eigenvectors">Setting up to calculate Eigenvalues and Eigenvectors</a>
  <ul class="collapse">
  <li><a href="#creating-an-example-data-set" id="toc-creating-an-example-data-set" class="nav-link" data-scroll-target="#creating-an-example-data-set">Creating an example data set</a></li>
  <li><a href="#re-centering-the-dataset" id="toc-re-centering-the-dataset" class="nav-link" data-scroll-target="#re-centering-the-dataset">Re-centering the dataset</a></li>
  <li><a href="#create-a-covariance-matrix" id="toc-create-a-covariance-matrix" class="nav-link" data-scroll-target="#create-a-covariance-matrix">Create a covariance matrix</a></li>
  </ul></li>
  <li><a href="#calculating-the-principal-components" id="toc-calculating-the-principal-components" class="nav-link" data-scroll-target="#calculating-the-principal-components">Calculating the Principal Components</a>
  <ul class="collapse">
  <li><a href="#calculate-eigenvalues-and-eigenvectors-from-the-covariance-matrix" id="toc-calculate-eigenvalues-and-eigenvectors-from-the-covariance-matrix" class="nav-link" data-scroll-target="#calculate-eigenvalues-and-eigenvectors-from-the-covariance-matrix">Calculate Eigenvalues and Eigenvectors from the Covariance Matrix</a></li>
  <li><a href="#principal-component-scores" id="toc-principal-component-scores" class="nav-link" data-scroll-target="#principal-component-scores">Principal Component Scores</a></li>
  <li><a href="#percent-variance-explained" id="toc-percent-variance-explained" class="nav-link" data-scroll-target="#percent-variance-explained">Percent Variance Explained</a></li>
  <li><a href="#plotting-our-principal-components" id="toc-plotting-our-principal-components" class="nav-link" data-scroll-target="#plotting-our-principal-components">Plotting our Principal Components</a></li>
  </ul></li>
  <li><a href="#prcomp-comparison" id="toc-prcomp-comparison" class="nav-link" data-scroll-target="#prcomp-comparison"><code>prcomp()</code> comparison</a>
  <ul class="collapse">
  <li><a href="#center-of-the-data" id="toc-center-of-the-data" class="nav-link" data-scroll-target="#center-of-the-data">Center of the Data</a></li>
  <li><a href="#eigenvalues" id="toc-eigenvalues" class="nav-link" data-scroll-target="#eigenvalues">Eigenvalues</a></li>
  <li><a href="#eigenvectors" id="toc-eigenvectors" class="nav-link" data-scroll-target="#eigenvectors">Eigenvectors</a></li>
  <li><a href="#pc-scores" id="toc-pc-scores" class="nav-link" data-scroll-target="#pc-scores">PC Scores</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/03_quality-control.html">Day 1 Self-learning:</a></li><li class="breadcrumb-item"><a href="../lessons/04_theory_of_PCA.html">Theory of PCA</a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Theory of PCA</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Will Gammerdinger </p>
             <p>Noor Sohail </p>
             <p>Zhu Zhuo </p>
             <p>James Billingsley </p>
             <p>Shannan Ho Sui </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 22, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>Approximate time: 45 minutes</p>
<section id="learning-objectives" class="level1">
<h1>Learning Objectives</h1>
<ul>
<li>Derive the covariance matrix used for Principal Components Analysis</li>
<li>Explain the roles of eigenvectors and eigenvalues within a Principal Components Analysis</li>
<li>Compare and contrast our Principal Components Analysis to the output of <code>prcomp()</code></li>
</ul>
</section>
<section id="background" class="level1">
<h1>Background</h1>
<p>Principal Components Analysis is a dimensionality-reduction method employed to stratify data by their variance. Thus, points that are more closely together on a PCA are more similar to each other, while points that are more dissimilar to each other will be further apart. As a result, you would likely expect similar cell from the same cell type to cluster near each other and apart from distinctly different cell types. In order to demonstrate how Principal Components Analysis works we are going to describe the process roughly works in high-dimensional data, then work through it a two-dimensional example data set.</p>
<p>Consider a single-cell experiment where you have the expression value for every gene in each cell. Let’s hypothetically say that our sample has data from 25,000 genes. We could plot this 25,000-dimensional space where each dimension represents a given gene’s expression and each point represents a cell in that 25,000-dimensional space. However, that is rather unintuitive and many of those dimensions are uninteresting, so we would like to reduce the dimensionality to the axes that separate the variance the most. This is what we will be doing with a Principal Components Analysis.</p>
<p>From a high-level perspective, what we will be doing is:</p>
<ul>
<li>Finding the vector in that 25,000-dimensional space that has the most variance in it.</li>
<li>Looking for another vector that is orthogonal (a 90° angle but in multi-dimensional space) to the first vector that explains the most of the remaining variance. We this process until, in theory, you have have as many vectors as dimensions in your original dataset. However, in practice, people oftentimes stop after a few dozen.</li>
<li>Using these vectors to transform your original data into Principal Components space.</li>
</ul>
</section>
<section id="setting-up-to-calculate-eigenvalues-and-eigenvectors" class="level1">
<h1>Setting up to calculate Eigenvalues and Eigenvectors</h1>
<p>We will be using some functions from <code>tidyverse</code> in the lesson, so we will need to load it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="creating-an-example-data-set" class="level2">
<h2 class="anchored" data-anchor-id="creating-an-example-data-set">Creating an example data set</h2>
<p>We are going to discuss the steps for deriving a Principal Components analysis, but in order to do it, we are going to use an example dataset. Let’s start by creating an tibble that has gene expression values for two genes from four cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector for Cell IDs</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>cells <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Cell_1"</span>, <span class="st">"Cell_2"</span>, <span class="st">"Cell_3"</span>, <span class="st">"Cell_4"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector to hold expression values for Gene A across all of the cells</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>Gene_A <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">12</span>, <span class="dv">65</span>, <span class="dv">23</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector to hold expression values for Gene B across all of the cells</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>Gene_B <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">30</span>, <span class="dv">57</span>, <span class="dv">18</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a tibble to hold the cell names and expression values</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>expression_tibble <span class="ot">&lt;-</span> <span class="fu">tibble</span>(cells, Gene_A, Gene_B)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># View the expression tibble</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>expression_tibble</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  cells  Gene_A Gene_B
  &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt;
1 Cell_1      0      4
2 Cell_2     12     30
3 Cell_3     65     57
4 Cell_4     23     18</code></pre>
</div>
</div>
<p>Let’s get an idea of what our data looks like by plotting it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a plot to view the raw expression data</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(expression_tibble, <span class="fu">aes</span>(<span class="at">x =</span> Gene_A, <span class="at">y =</span> Gene_B, <span class="at">label =</span> cells)) <span class="sc">+</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"cornflowerblue"</span>) <span class="sc">+</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">80</span>) <span class="sc">+</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">80</span>) <span class="sc">+</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene A"</span>) <span class="sc">+</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Gene B"</span>) <span class="sc">+</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Example Expression Values from Four Cells"</span>) <span class="sc">+</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04_theory_of_PCA_files/figure-html/plot_raw_data-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now if were were to consider our data set that has <strong><em>X</em></strong> genes and <strong><em>Y</em></strong> cells, then this plot would have <strong><em>X</em></strong> dimensions and <strong><em>Y</em></strong> points.</p>
</section>
<section id="re-centering-the-dataset" class="level2">
<h2 class="anchored" data-anchor-id="re-centering-the-dataset">Re-centering the dataset</h2>
<p>When performing PCA, the first step is to re-center the data so that the center of the data is the origin. In order to do this, we need to do two steps:</p>
<p><strong>1. Find the center of the data</strong></p>
<p>We can find the center of the data by taking the average gene expression across each gene (dimension).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the center of the data by:</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Finding the average expression of gene A</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>Gene_A_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(expression_tibble<span class="sc">$</span>Gene_A)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Finding the average expression of gene B</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>Gene_B_mean <span class="ot">&lt;-</span> <span class="fu">mean</span>(expression_tibble<span class="sc">$</span>Gene_B)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector to hold the center of the data</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>center_of_data <span class="ot">&lt;-</span> <span class="fu">c</span>(Gene_A_mean, Gene_B_mean)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign names to the components of the vector</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(center_of_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene_A"</span>, <span class="st">"Gene_B"</span>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the center_of_data vector</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>center_of_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Gene_A Gene_B 
 25.00  27.25 </code></pre>
</div>
</div>
<p>The new center of the data will be where <strong><em>( 25, 27.25 )</em></strong> is currently. We can visualize this center and inspect that this is does in fact appear to be roughly in the middle of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View where the center of the data is located</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(expression_tibble, <span class="fu">aes</span>(<span class="at">x =</span> Gene_A, <span class="at">y =</span> Gene_B, <span class="at">label =</span> cells)) <span class="sc">+</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"cornflowerblue"</span>) <span class="sc">+</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotate</span>(<span class="st">"point"</span>, <span class="at">x =</span> Gene_A_mean, <span class="at">y =</span> Gene_B_mean, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> Gene_A_mean, <span class="at">y =</span> Gene_B_mean, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">label=</span><span class="st">"New Center"</span>, <span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>, <span class="dv">80</span>) <span class="sc">+</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">80</span>) <span class="sc">+</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene A"</span>) <span class="sc">+</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Gene B"</span>) <span class="sc">+</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Example Expression Values from Four Cells"</span>) <span class="sc">+</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04_theory_of_PCA_files/figure-html/new_center_plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>2. Translate the points from their raw expression coordinates to coordinates centered around the center of the data</strong></p>
<p>Next, we need to shift the points so that the center of the data corresponds to the origin. We can do this by subtracting the mean value from each gene from the raw values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Shift the data points so that they data is centered on the origin</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>recentered_expression_tibble <span class="ot">&lt;-</span> expression_tibble <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">Gene_A =</span> Gene_A <span class="sc">-</span> Gene_A_mean,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">Gene_B =</span> Gene_B <span class="sc">-</span> Gene_B_mean</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the re-centered data</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>recentered_expression_tibble</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  cells  Gene_A Gene_B
  &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt;
1 Cell_1    -25 -23.2 
2 Cell_2    -13   2.75
3 Cell_3     40  29.8 
4 Cell_4     -2  -9.25</code></pre>
</div>
</div>
<p>As with any data analysis, it is always a good idea to visualize your data to ensure that it appears to behaving how you intended. So let’s visualize our re-centered data to ensure that the data appears centered around the origin.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the raw data after it has been shifted to have the center of the data align with the origin</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(recentered_expression_tibble, <span class="fu">aes</span>(<span class="at">x =</span> Gene_A, <span class="at">y =</span> Gene_B, <span class="at">label =</span> cells)) <span class="sc">+</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="st">"cornflowerblue"</span>) <span class="sc">+</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"point"</span>, <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> <span class="dv">0</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">label=</span><span class="st">"New Center"</span>, <span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Gene A"</span>) <span class="sc">+</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Gene B"</span>) <span class="sc">+</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Example Re-centered Expression Values from Four Cells"</span>) <span class="sc">+</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04_theory_of_PCA_files/figure-html/plot_recentered_data-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="create-a-covariance-matrix" class="level2">
<h2 class="anchored" data-anchor-id="create-a-covariance-matrix">Create a covariance matrix</h2>
<p>The next step that we will after we have re-centered our data around the origin is to calculate the covariance matrix. As a reminder, the covariance is the joint variability of two random variables. In this case our random variables are the genes. The illustration below we provide examples for positive and negative covariance values along with a covariance that is near zero.</p>
<p><img src="../img/Covariance.png" width="800"></p>
<p>The equation to estimate the sample covariance is:</p>
<p><img src="https://latex.codecogs.com/svg.image?&amp;space;cov(X,Y)=\frac{1}{n - 1}\sum_{i=1}^{n}(x_{i}-\bar{x})(y_{i}-\bar{y})" width="300"></p>
<p>Where:</p>
<ul>
<li><em>X</em> is Gene A</li>
<li><em>Y</em> is Gene B</li>
<li><em>n</em> is the number of cells</li>
<li><em>x<sub>i</sub></em> is an expression value for Gene A</li>
<li><em><span class="math inline">\(\bar{x}\)</span></em> is the mean expression value for Gene A</li>
<li><em>y<sub>i</sub></em> is an expression value for Gene B</li>
<li><em><span class="math inline">\(\bar{y}\)</span></em> is the mean expression value for Gene B</li>
</ul>
<p>Now we will calculate the covariance for all pairwise comparisons for our re-centered expression data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Move the cell IDs to the rownames and convert the tibble to a matrix</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>recentered_expression_matrix <span class="ot">&lt;-</span> recentered_expression_tibble <span class="sc">%&gt;%</span> </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"cells"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a covariance matrix</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>cov_matrix <span class="ot">&lt;-</span> <span class="fu">cov</span>(recentered_expression_matrix)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the covariance matrix</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>cov_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Gene_A   Gene_B
Gene_A 799.3333 584.6667
Gene_B 584.6667 506.2500</code></pre>
</div>
</div>
<p>Let’s go ahead and manually check a covariance by hand to ensure that we trust the outputted covariance matrix. Let’s estimate the covariance of Gene A and Gene B.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate the covariance of Gene A and Gene B by hand</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span><span class="sc">/</span>(<span class="dv">4-1</span>))<span class="sc">*</span>((<span class="dv">0-25</span>)<span class="sc">*</span>(<span class="dv">4</span><span class="fl">-27.25</span>) <span class="sc">+</span> (<span class="dv">12-25</span>)<span class="sc">*</span>(<span class="dv">30</span><span class="fl">-27.25</span>) <span class="sc">+</span> (<span class="dv">65-25</span>)<span class="sc">*</span>(<span class="dv">57</span><span class="fl">-27.25</span>) <span class="sc">+</span> (<span class="dv">23-25</span>)<span class="sc">*</span>(<span class="dv">18</span><span class="fl">-27.25</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 584.6667</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or you can use the covariance function</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cov</span>(recentered_expression_matrix[, <span class="st">"Gene_A"</span>], recentered_expression_matrix[, <span class="st">"Gene_B"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 584.6667</code></pre>
</div>
</div>
<p>This value should match the covariance found in our covariance matrix for the covariance between Gene A and Gene B.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<a href="../lessons/04_theory_of_PCA_Answer-key.html#exercise-1"><strong>Exercise 1</strong></a>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Exercise 1</p>
<p>There are a couple properties of variance and co-variance that we can verify:</p>
<p><em>cov(X,X)</em> is equal to <em>var(X)</em>. We can observe this is mathematically below:</p>
<p><img src="https://latex.codecogs.com/svg.image?&amp;space;cov(X,X)=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})(x_{i}-\bar{x})=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})^{2}=var(X)" width="500"></p>
<p>As a result, you will sometimes see covariance matrices written as:</p>
<p><img src="https://latex.codecogs.com/svg.image?\begin{bmatrix}var(X)&amp;cov(X,Y)&amp;...&amp;cov(X,Z)\\cov(Y,X)&amp;var(Y)&amp;...&amp;cov(Y,Z)\\...&amp;...&amp;...&amp;...\\cov(Z,X)&amp;cov(Z,Y)&amp;...&amp;var(Z)\\\end{bmatrix}" width="300"></p>
<ol type="1">
<li>Confirm this property by estimating the variance for Gene A.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate the variance for Gene A</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(recentered_expression_matrix[, <span class="st">"Gene_A"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 799.3333</code></pre>
</div>
</div>
<ol start="2" type="1">
<li>Now estimate the covariance for Gene A and Gene A</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate the covariance for Gene A and Gene A</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cov</span>(recentered_expression_matrix[, <span class="st">"Gene_A"</span>], recentered_expression_matrix[, <span class="st">"Gene_A"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 799.3333</code></pre>
</div>
</div>
<ol start="3" type="1">
<li>Is the value the same? Does it match the value in the covariance matrix for Gene A and Gene A?</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the covariance estimate of Gene A and Gene A from the covariance matrix</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>cov_matrix[<span class="st">"Gene_A"</span>, <span class="st">"Gene_A"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 799.3333</code></pre>
</div>
</div>
<p><em>cov(X,Y)</em> is equal to <em>cov(Y,X)</em>. We can observe this is mathematically below:</p>
<p><img src="https://latex.codecogs.com/svg.image?&amp;space;cov(X,Y)=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})(y_{i}-\bar{y})=\frac{1}{n-1}\sum_{i=1}^{n}(y_{i}-\bar{y})(x_{i}-\bar{x})=cov(Y,X)" width="500"></p>
<ol start="4" type="1">
<li>Estimate the covariance of Gene B and Gene A.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate the covariance of Gene B and Gene A</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cov</span>(recentered_expression_tibble[, <span class="st">"Gene_B"</span>], recentered_expression_tibble[, <span class="st">"Gene_A"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Gene_A
Gene_B 584.6667</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>How does this compare to the covariance that we estimated by hand?</li>
</ol>
</div>
</div>
</section>
</section>
<section id="calculating-the-principal-components" class="level1">
<h1>Calculating the Principal Components</h1>
<section id="calculate-eigenvalues-and-eigenvectors-from-the-covariance-matrix" class="level2">
<h2 class="anchored" data-anchor-id="calculate-eigenvalues-and-eigenvectors-from-the-covariance-matrix">Calculate Eigenvalues and Eigenvectors from the Covariance Matrix</h2>
<p>First, we will need to calculate our eigenvalues, which are a measure of variance for a principal component. We can then use eigenvalues to calculate our eigenvectors, but R will do this all in a single step for us. We can think of the eigenvectors as the weights, or influence, for transforming our re-centered expression values into PCA space.</p>
<p>Now that we have a covariance matrix for the re-centered expression values, we need to estimate the eigenvalues and eigenvectors for the covariance matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find the eigenvalues and eigenvectors of the covariance matrix</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>eig <span class="ot">&lt;-</span> <span class="fu">eigen</span>(cov_matrix)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the output from eigen()</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>eig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>eigen() decomposition
$values
[1] 1255.543   50.040

$vectors
           [,1]       [,2]
[1,] -0.7883911  0.6151743
[2,] -0.6151743 -0.7883911</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Click here to see how to calculate eigenvalues and eigenvectors">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click here to see how to calculate eigenvalues and eigenvectors
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong>Calculating Eigenvalues</strong></p>
<p>If we wanted to do this by hand then we need to set the determinant of the covariance matrix minus <span class="math inline">\(\lambda\)</span> times the identify matrix equal to 0 and solving for <span class="math inline">\(\lambda\)</span>.</p>
<p><img src="https://latex.codecogs.com/svg.image?det(\begin{bmatrix}var(X)&amp;cov(X,Y)\\cov(Y,X)&amp;var(Y)\\\end{bmatrix}-\lambda\begin{bmatrix}1&amp;0\\0&amp;1\\\end{bmatrix})=0" width="500"></p>
<p>In this case it can be simplified to:</p>
<p><img src="https://latex.codecogs.com/svg.image?(var(X)-\lambda)(var(Y)-\lambda)-(cov(X,Y))(cov(Y,X))=0" width="500"></p>
<p>We can now replace the variance and covariance values with the values that we have from the covariance matrix and solve for <span class="math inline">\(\lambda\)</span>. This results in two lambda values: <strong><span class="math inline">\(\lambda\)</span><sub>1</sub> = 1255.543</strong> and <strong><span class="math inline">\(\lambda\)</span><sub>2</sub> = 50.04</strong>.</p>
<p>We can see that this is equal to the eigenvalues returned from our <code>eig</code> object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvalues</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>eig<span class="sc">$</span>values</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1255.543   50.040</code></pre>
</div>
</div>
<p><strong>Calculating Eigenvectors</strong></p>
<p>In order to find the eigenvectors, we need to substitute the variacne and covariance values from the covariance matrix along with one of our <em><span class="math inline">\(\lambda\)</span></em> values, <em><span class="math inline">\(\lambda\)</span><sub>1</sub></em> to find the associated eigenvector, <em>v<sub>1</sub></em>. This results in <em>v<sub>1</sub></em> being equal to -0.7883911, -0.6151743.</p>
<p><img src="https://latex.codecogs.com/svg.image?(\begin{bmatrix}var(X)&amp;cov(X,Y)\\cov(Y,X)&amp;var(Y)\\\end{bmatrix}-\lambda&amp;space;_{1}\begin{bmatrix}1&amp;0\\0&amp;1\\\end{bmatrix})v_{1}=0" width="500"></p>
<p>We can confirm this by plugging our covariance matrix, eigenvalues and eigenvectors back into the above equation and it should return 0, or near 0 (due to rounding errors).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the eigenvalues and eigenvectors</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>(cov_matrix <span class="sc">-</span> eig<span class="sc">$</span>values[<span class="dv">1</span>]<span class="sc">*</span><span class="fu">diag</span>(<span class="dv">2</span>)) <span class="sc">%*%</span> eig<span class="sc">$</span>vectors[,<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                [,1]
Gene_A -5.484729e-14
Gene_B -1.507339e-14</code></pre>
</div>
</div>
<p>We then repeat this process for <em><span class="math inline">\(\lambda\)</span><sub>2</sub></em> in order to to find its associated eigenvector, <em>v<sub>2</sub></em>, which results in 0.6151743, -0.7883911. Thus, our resulting matrix of eigenvectors appears as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvectors</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>eig<span class="sc">$</span>vectors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           [,1]       [,2]
[1,] -0.7883911  0.6151743
[2,] -0.6151743 -0.7883911</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="principal-component-scores" class="level2">
<h2 class="anchored" data-anchor-id="principal-component-scores">Principal Component Scores</h2>
<p>Now that we have our re-centered expression matrix and eigenvectors, we can perform matrix multiplication to obtain our principal component scores. This is where we are transforming our data from the re-centered expression into Principal Components space by the weight, or influence, of the eigenvector.</p>
<p><img src="../img/PC_generation.png" width="800"></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform the data into PC space by multiply the re-centered expression matrix by the eigenvectors</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>pc_scores <span class="ot">&lt;-</span> recentered_expression_matrix <span class="sc">%*%</span> eig<span class="sc">$</span>vectors</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Name the columns in pc_scores object</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(pc_scores) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"PC_1"</span>, <span class="st">"PC_2"</span>)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the pc_scores object</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>pc_scores</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             PC_1       PC_2
Cell_1  34.012581   2.950736
Cell_2   8.557355 -10.165342
Cell_3 -49.837081   1.152337
Cell_4   7.267145   6.062269</code></pre>
</div>
</div>
</section>
<section id="percent-variance-explained" class="level2">
<h2 class="anchored" data-anchor-id="percent-variance-explained">Percent Variance Explained</h2>
<p>We are likely interested in knowing in knowing the amount of variance explained in our data by each principal component. As we alluded to earlier, we can use the eigenvalues to help us with this. The sum of all of the eigenvalues captures the total variance explained by the principal components analysis.</p>
<p>Thus, if we want to know the proportion of the variance explained by each principal component, then we would need to divide each eigenvalue by the total variance explained (the sum of the eigenvalues) and if we wanted this as a percentage then we would need to multiply it by 100.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the percent of variance explained by each PC using the eigenvalues</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>pct_var_explained <span class="ot">&lt;-</span> (eig<span class="sc">$</span>values <span class="sc">/</span> <span class="fu">sum</span>(eig<span class="sc">$</span>values)) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Name the elements of the pct_var_explained by their PC</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pct_var_explained) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"PC_1"</span>, <span class="st">"PC_2"</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out the percent of variance explained by each PC</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>pct_var_explained</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    PC_1     PC_2 
96.16723  3.83277 </code></pre>
</div>
</div>
</section>
<section id="plotting-our-principal-components" class="level2">
<h2 class="anchored" data-anchor-id="plotting-our-principal-components">Plotting our Principal Components</h2>
<p>With out principal components in hand, we will create a visualization of the principal components.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a tibble to hold the PC scores we found and also make the Cell IDs into a column</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>pc_scores_tibble <span class="ot">&lt;-</span> pc_scores <span class="sc">%&gt;%</span> </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"cells"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PC scores we found</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pc_scores_tibble, <span class="fu">aes</span>(<span class="at">x =</span> PC_1, <span class="at">y =</span> PC_2, <span class="at">label =</span> cells)) <span class="sc">+</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="at">color =</span> <span class="st">"cornflowerblue"</span>) <span class="sc">+</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">12</span>, <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC 1 (Variance Explained "</span>, <span class="fu">round</span>(pct_var_explained[<span class="st">"PC_1"</span>], <span class="at">digits =</span> <span class="dv">2</span>),<span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC 2 (Variance Explained "</span>, <span class="fu">round</span>(pct_var_explained[<span class="st">"PC_2"</span>], <span class="at">digits =</span> <span class="dv">2</span>),<span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"PCA of Expression Values from Four Cells"</span>) <span class="sc">+</span></span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04_theory_of_PCA_files/figure-html/plotting_PCA_hand-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<a href="../lessons/04_theory_of_PCA_Answer-key.html#exercise-2"><strong>Exercise 2</strong></a>
</div>
</div>
<div class="callout-body-container callout-body">
<p>When looking at the percent explained by each principal component, the first principal component should explain the most and each of the following principal components should explain less than the previous principal component. Let’s have a look at our <code>pct_var_explained</code> object, are our results congruent with the expectation this expectation?</p>
</div>
</div>
</section>
</section>
<section id="prcomp-comparison" class="level1">
<h1><code>prcomp()</code> comparison</h1>
<p>At this point we have completed a principal components analysis, but we can also compare it to how R does a principal components analysis and see how our results compare. One way to do a principal components analysis in R is to us the <code>prcomp()</code> function in R.</p>
<p>In order to run <code>prcomp()</code>, we need to start from an expression matrix and move the cell IDs to the rownames:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run prcomp() on the expression tibble after moving the Cell IDs to be rownames</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>prcomp_PCA <span class="ot">&lt;-</span> expression_tibble <span class="sc">%&gt;%</span> </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">column_to_rownames</span>(<span class="st">"cells"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s inspect the output from <code>prcomp()</code> to ensure that we are getting the same output:</p>
<section id="center-of-the-data" class="level2">
<h2 class="anchored" data-anchor-id="center-of-the-data">Center of the Data</h2>
<p>First we can ensure that we have the same center of the data. <code>prcomp()</code> says that the center of the data is:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the center of the data found by prcomp()</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>prcomp_PCA<span class="sc">$</span>center</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Gene_A Gene_B 
 25.00  27.25 </code></pre>
</div>
</div>
<p>We can compare this with what we derived as the center of the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the center of the data we found</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>center_of_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Gene_A Gene_B 
 25.00  27.25 </code></pre>
</div>
</div>
<p>They match which is a great sign that we did this correctly!</p>
</section>
<section id="eigenvalues" class="level2">
<h2 class="anchored" data-anchor-id="eigenvalues">Eigenvalues</h2>
<p>Note that eigenvalues are the variance for a particular principal component and that variance is the square of the standard deviation as shown below.</p>
<p><img src="https://latex.codecogs.com/svg.image?var(X)=sd(X)^{2}" width="150"></p>
<p><code>prcomp()</code> reports the standard deviation, so we will need to square the items in the <code>prcomp_PCA$sdev</code> slot in order to recover our variance, or eigenvalues.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvalues found by prcomp() by squaring prcomp_PCA$sdev</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>prcomp_eigenvalues <span class="ot">&lt;-</span> prcomp_PCA<span class="sc">$</span>sdev <span class="sc">**</span> <span class="dv">2</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Name the elements of the prcomp_eigenvalues by their PC</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(prcomp_eigenvalues) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"PC_1"</span>, <span class="st">"PC_2"</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Print out prcomp_eigenvalues</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>prcomp_eigenvalues</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    PC_1     PC_2 
1255.543   50.040 </code></pre>
</div>
</div>
<p>We can compare this with our estimation of eiganvalues by hand:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvalues we found</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>eig<span class="sc">$</span>values</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1255.543   50.040</code></pre>
</div>
</div>
</section>
<section id="eigenvectors" class="level2">
<h2 class="anchored" data-anchor-id="eigenvectors">Eigenvectors</h2>
<p>Let’s go ahead and compare the eigenvectors from <code>prcomp()</code> to the eigenvectors that we calculated. <code>prcomp()</code> calculated the eigenvectors as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvectors found by prcomp()</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>prcomp_PCA<span class="sc">$</span>rotation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             PC1        PC2
Gene_A 0.7883911  0.6151743
Gene_B 0.6151743 -0.7883911</code></pre>
</div>
</div>
<p>While we calculated the eigenvectors as:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the eigenvectors we found</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>eig<span class="sc">$</span>vectors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           [,1]       [,2]
[1,] -0.7883911  0.6151743
[2,] -0.6151743 -0.7883911</code></pre>
</div>
</div>
<p>We can see that the values are the same, but the sign is flipped between the two methods. This is because eigenvectors are only defined “up to sign”. This means for a given eigenvalue, the eigenvector <em>v<sub>1</sub></em> and <em>-v<sub>1</sub></em> are the same line, but point in different directions. This does not impact our principal components calculations. However, it could mean that a plot derived <em>v<sub>1</sub></em> will look rotated 180° when compared to a plot derived from <em>-v<sub>1</sub></em>. Regardless, it looks like we have the same eigenvectors in both approaches.</p>
</section>
<section id="pc-scores" class="level2">
<h2 class="anchored" data-anchor-id="pc-scores">PC Scores</h2>
<p>Finally, we can check our PC scores to ensure that they are also equivalent. We should note that since the eigenvector from <code>prcomp()</code> and the eigenvector we calculated differed by sign and eigenvectors are multiplied by the re-centered expression, then we should expect our principal components to also differ by sign. The principal components from <code>prcomp()</code> are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the PC scores found by prcomp()</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>prcomp_PCA<span class="sc">$</span>x</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              PC1        PC2
Cell_1 -34.012581   2.950736
Cell_2  -8.557355 -10.165342
Cell_3  49.837081   1.152337
Cell_4  -7.267145   6.062269</code></pre>
</div>
</div>
<p>While our principal components were:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the PC scores we found</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>pc_scores</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             PC_1       PC_2
Cell_1  34.012581   2.950736
Cell_2   8.557355 -10.165342
Cell_3 -49.837081   1.152337
Cell_4   7.267145   6.062269</code></pre>
</div>
</div>
<p>Once again, the values are the same, but the only difference is the sign, which comes from the arbitrary direction of the eigenvector.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<a href="../lessons/04_theory_of_PCA_Answer-key.html#exercise-3"><strong>Exercise 3</strong></a>
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a plot of the Principal Components Analysis derived from <code>prcomp()</code>. Is it the same as the plot we derived except only rotated 180°?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a tibble to hold the PC scores prcomp() found and also make the Cell IDs into a column</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>prcomp_pc_scores_tibble <span class="ot">&lt;-</span> prcomp_PCA<span class="sc">$</span>x <span class="sc">%&gt;%</span> </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">"cells"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>()</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the PC scores found by prcomp()</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(prcomp_pc_scores_tibble, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">label =</span> cells)) <span class="sc">+</span></span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="at">color =</span> <span class="st">"cornflowerblue"</span>) <span class="sc">+</span></span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="dv">0</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="sc">-</span><span class="dv">50</span>, <span class="dv">50</span>) <span class="sc">+</span></span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="sc">-</span><span class="dv">12</span>, <span class="dv">12</span>) <span class="sc">+</span></span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="fu">paste0</span>(<span class="st">"PC 1 (Variance Explained "</span>, <span class="fu">round</span>(prcomp_eigenvalues[<span class="st">"PC_1"</span>], <span class="at">digits =</span> <span class="dv">2</span>),<span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"PC 2 (Variance Explained "</span>, <span class="fu">round</span>(prcomp_eigenvalues[<span class="st">"PC_2"</span>], <span class="at">digits =</span> <span class="dv">2</span>),<span class="st">"%)"</span>)) <span class="sc">+</span></span>
<span id="cb55-16"><a href="#cb55-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"PCA of Expression Values from Four Cells"</span>) <span class="sc">+</span></span>
<span id="cb55-17"><a href="#cb55-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="04_theory_of_PCA_files/figure-html/plotting_PCA_prcomp-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb56" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Theory of PCA"</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> </span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co">  - Will Gammerdinger</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="co">  - Noor Sohail</span></span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a><span class="co">  - Zhu Zhuo</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a><span class="co">  - James Billingsley</span></span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="co">  - Shannan Ho Sui</span></span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "Tuesday, July 22, 2025"</span></span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>Approximate time: 45 minutes</span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a><span class="fu"># Learning Objectives</span></span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Derive the covariance matrix used for Principal Components Analysis</span>
<span id="cb56-20"><a href="#cb56-20" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Explain the roles of eigenvectors and eigenvalues within a Principal Components Analysis</span>
<span id="cb56-21"><a href="#cb56-21" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Compare and contrast our Principal Components Analysis to the output of <span class="in">`prcomp()`</span></span>
<span id="cb56-22"><a href="#cb56-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-23"><a href="#cb56-23" aria-hidden="true" tabindex="-1"></a><span class="fu"># Background</span></span>
<span id="cb56-24"><a href="#cb56-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-25"><a href="#cb56-25" aria-hidden="true" tabindex="-1"></a>Principal Components Analysis is a dimensionality-reduction method employed to stratify data by their variance. Thus, points that are more closely together on a PCA are more similar to each other, while points that are more dissimilar to each other will be further apart. As a result, you would likely expect similar cell from the same cell type to cluster near each other and apart from distinctly different cell types. In order to demonstrate how Principal Components Analysis works we are going to describe the process roughly works in high-dimensional data, then work through it a two-dimensional example data set.</span>
<span id="cb56-26"><a href="#cb56-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-27"><a href="#cb56-27" aria-hidden="true" tabindex="-1"></a>Consider a single-cell experiment where you have the expression value for every gene in each cell. Let's hypothetically say that our sample has data from 25,000 genes. We could plot this 25,000-dimensional space where each dimension represents a given gene's expression and each point represents a cell in that 25,000-dimensional space. However, that is rather unintuitive and many of those dimensions are uninteresting, so we would like to reduce the dimensionality to the axes that separate the variance the most. This is what we will be doing with a Principal Components Analysis.</span>
<span id="cb56-28"><a href="#cb56-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-29"><a href="#cb56-29" aria-hidden="true" tabindex="-1"></a>From a high-level perspective, what we will be doing is:</span>
<span id="cb56-30"><a href="#cb56-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-31"><a href="#cb56-31" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Finding the vector in that 25,000-dimensional space that has the most variance in it.</span>
<span id="cb56-32"><a href="#cb56-32" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Looking for another vector that is orthogonal (a 90<span class="dv">&amp;deg;</span> angle but in multi-dimensional space) to the first vector that explains the most of the remaining variance. We this process until, in theory, you have have as many vectors as dimensions in your original dataset. However, in practice, people oftentimes stop after a few dozen. </span>
<span id="cb56-33"><a href="#cb56-33" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Using these vectors to transform your original data into Principal Components space.</span>
<span id="cb56-34"><a href="#cb56-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-35"><a href="#cb56-35" aria-hidden="true" tabindex="-1"></a><span class="fu"># Setting up to calculate Eigenvalues and Eigenvectors</span></span>
<span id="cb56-36"><a href="#cb56-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-37"><a href="#cb56-37" aria-hidden="true" tabindex="-1"></a>We will be using some functions from <span class="in">`tidyverse`</span> in the lesson, so we will need to load it:</span>
<span id="cb56-38"><a href="#cb56-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-39"><a href="#cb56-39" aria-hidden="true" tabindex="-1"></a><span class="in">```{r load_tidyverse}</span></span>
<span id="cb56-40"><a href="#cb56-40" aria-hidden="true" tabindex="-1"></a><span class="in">library(tidyverse)</span></span>
<span id="cb56-41"><a href="#cb56-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-42"><a href="#cb56-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-43"><a href="#cb56-43" aria-hidden="true" tabindex="-1"></a><span class="fu">## Creating an example data set</span></span>
<span id="cb56-44"><a href="#cb56-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-45"><a href="#cb56-45" aria-hidden="true" tabindex="-1"></a>We are going to discuss the steps for deriving a Principal Components analysis, but in order to do it, we are going to use an example dataset. Let's start by creating an tibble that has gene expression values for two genes from four cells.</span>
<span id="cb56-46"><a href="#cb56-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-47"><a href="#cb56-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r create_tibble}</span></span>
<span id="cb56-48"><a href="#cb56-48" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a vector for Cell IDs</span></span>
<span id="cb56-49"><a href="#cb56-49" aria-hidden="true" tabindex="-1"></a><span class="in">cells &lt;- c("Cell_1", "Cell_2", "Cell_3", "Cell_4")</span></span>
<span id="cb56-50"><a href="#cb56-50" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a vector to hold expression values for Gene A across all of the cells</span></span>
<span id="cb56-51"><a href="#cb56-51" aria-hidden="true" tabindex="-1"></a><span class="in">Gene_A &lt;- c(0, 12, 65, 23)</span></span>
<span id="cb56-52"><a href="#cb56-52" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a vector to hold expression values for Gene B across all of the cells</span></span>
<span id="cb56-53"><a href="#cb56-53" aria-hidden="true" tabindex="-1"></a><span class="in">Gene_B &lt;- c(4, 30, 57, 18)</span></span>
<span id="cb56-54"><a href="#cb56-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-55"><a href="#cb56-55" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a tibble to hold the cell names and expression values</span></span>
<span id="cb56-56"><a href="#cb56-56" aria-hidden="true" tabindex="-1"></a><span class="in">expression_tibble &lt;- tibble(cells, Gene_A, Gene_B)</span></span>
<span id="cb56-57"><a href="#cb56-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-58"><a href="#cb56-58" aria-hidden="true" tabindex="-1"></a><span class="in"># View the expression tibble</span></span>
<span id="cb56-59"><a href="#cb56-59" aria-hidden="true" tabindex="-1"></a><span class="in">expression_tibble</span></span>
<span id="cb56-60"><a href="#cb56-60" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-61"><a href="#cb56-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-62"><a href="#cb56-62" aria-hidden="true" tabindex="-1"></a>Let's get an idea of what our data looks like by plotting it:</span>
<span id="cb56-63"><a href="#cb56-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-64"><a href="#cb56-64" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_raw_data}</span></span>
<span id="cb56-65"><a href="#cb56-65" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a plot to view the raw expression data</span></span>
<span id="cb56-66"><a href="#cb56-66" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(expression_tibble, aes(x = Gene_A, y = Gene_B, label = cells)) +</span></span>
<span id="cb56-67"><a href="#cb56-67" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(color = "cornflowerblue") +</span></span>
<span id="cb56-68"><a href="#cb56-68" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(hjust = 0, vjust = -1) +</span></span>
<span id="cb56-69"><a href="#cb56-69" aria-hidden="true" tabindex="-1"></a><span class="in">  xlim(0, 80) +</span></span>
<span id="cb56-70"><a href="#cb56-70" aria-hidden="true" tabindex="-1"></a><span class="in">  ylim(0, 80) +</span></span>
<span id="cb56-71"><a href="#cb56-71" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() +</span></span>
<span id="cb56-72"><a href="#cb56-72" aria-hidden="true" tabindex="-1"></a><span class="in">  xlab("Gene A") +</span></span>
<span id="cb56-73"><a href="#cb56-73" aria-hidden="true" tabindex="-1"></a><span class="in">  ylab("Gene B") +</span></span>
<span id="cb56-74"><a href="#cb56-74" aria-hidden="true" tabindex="-1"></a><span class="in">  ggtitle("Example Expression Values from Four Cells") +</span></span>
<span id="cb56-75"><a href="#cb56-75" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.title = element_text(hjust = 0.5))</span></span>
<span id="cb56-76"><a href="#cb56-76" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-77"><a href="#cb56-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-78"><a href="#cb56-78" aria-hidden="true" tabindex="-1"></a>Now if were were to consider our data set that has ***X*** genes and ***Y*** cells, then this plot would have ***X*** dimensions and ***Y*** points.</span>
<span id="cb56-79"><a href="#cb56-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-80"><a href="#cb56-80" aria-hidden="true" tabindex="-1"></a><span class="fu">## Re-centering the dataset</span></span>
<span id="cb56-81"><a href="#cb56-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-82"><a href="#cb56-82" aria-hidden="true" tabindex="-1"></a>When performing PCA, the first step is to re-center the data so that the center of the data is the origin. In order to do this, we need to do two steps:</span>
<span id="cb56-83"><a href="#cb56-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-84"><a href="#cb56-84" aria-hidden="true" tabindex="-1"></a>**1. Find the center of the data**</span>
<span id="cb56-85"><a href="#cb56-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-86"><a href="#cb56-86" aria-hidden="true" tabindex="-1"></a>We can find the center of the data by taking the average gene expression across each gene (dimension).</span>
<span id="cb56-87"><a href="#cb56-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-88"><a href="#cb56-88" aria-hidden="true" tabindex="-1"></a><span class="in">```{r center_of_data}</span></span>
<span id="cb56-89"><a href="#cb56-89" aria-hidden="true" tabindex="-1"></a><span class="in"># Determine the center of the data by:</span></span>
<span id="cb56-90"><a href="#cb56-90" aria-hidden="true" tabindex="-1"></a><span class="in"># Finding the average expression of gene A</span></span>
<span id="cb56-91"><a href="#cb56-91" aria-hidden="true" tabindex="-1"></a><span class="in">Gene_A_mean &lt;- mean(expression_tibble$Gene_A)</span></span>
<span id="cb56-92"><a href="#cb56-92" aria-hidden="true" tabindex="-1"></a><span class="in"># Finding the average expression of gene B</span></span>
<span id="cb56-93"><a href="#cb56-93" aria-hidden="true" tabindex="-1"></a><span class="in">Gene_B_mean &lt;- mean(expression_tibble$Gene_B)</span></span>
<span id="cb56-94"><a href="#cb56-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-95"><a href="#cb56-95" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a vector to hold the center of the data</span></span>
<span id="cb56-96"><a href="#cb56-96" aria-hidden="true" tabindex="-1"></a><span class="in">center_of_data &lt;- c(Gene_A_mean, Gene_B_mean)</span></span>
<span id="cb56-97"><a href="#cb56-97" aria-hidden="true" tabindex="-1"></a><span class="in"># Assign names to the components of the vector</span></span>
<span id="cb56-98"><a href="#cb56-98" aria-hidden="true" tabindex="-1"></a><span class="in">names(center_of_data) &lt;- c("Gene_A", "Gene_B")</span></span>
<span id="cb56-99"><a href="#cb56-99" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out the center_of_data vector</span></span>
<span id="cb56-100"><a href="#cb56-100" aria-hidden="true" tabindex="-1"></a><span class="in">center_of_data</span></span>
<span id="cb56-101"><a href="#cb56-101" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-102"><a href="#cb56-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-103"><a href="#cb56-103" aria-hidden="true" tabindex="-1"></a>The new center of the data will be where ***( `r center_of_data` )*** is currently. We can visualize this center and inspect that this is does in fact appear to be roughly in the middle of the data.</span>
<span id="cb56-104"><a href="#cb56-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-105"><a href="#cb56-105" aria-hidden="true" tabindex="-1"></a><span class="in">``` {r new_center_plot}</span></span>
<span id="cb56-106"><a href="#cb56-106" aria-hidden="true" tabindex="-1"></a><span class="in"># View where the center of the data is located</span></span>
<span id="cb56-107"><a href="#cb56-107" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(expression_tibble, aes(x = Gene_A, y = Gene_B, label = cells)) +</span></span>
<span id="cb56-108"><a href="#cb56-108" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(color = "cornflowerblue") +</span></span>
<span id="cb56-109"><a href="#cb56-109" aria-hidden="true" tabindex="-1"></a><span class="in">    annotate("point", x = Gene_A_mean, y = Gene_B_mean, color = "red", size = 3) +</span></span>
<span id="cb56-110"><a href="#cb56-110" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(hjust = 0, vjust = -1) +</span></span>
<span id="cb56-111"><a href="#cb56-111" aria-hidden="true" tabindex="-1"></a><span class="in">  annotate("text", x = Gene_A_mean, y = Gene_B_mean, color = "red", label="New Center", hjust = 0, vjust = -1) +</span></span>
<span id="cb56-112"><a href="#cb56-112" aria-hidden="true" tabindex="-1"></a><span class="in">  xlim(0, 80) +</span></span>
<span id="cb56-113"><a href="#cb56-113" aria-hidden="true" tabindex="-1"></a><span class="in">  ylim(0, 80) +</span></span>
<span id="cb56-114"><a href="#cb56-114" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() +</span></span>
<span id="cb56-115"><a href="#cb56-115" aria-hidden="true" tabindex="-1"></a><span class="in">  xlab("Gene A") +</span></span>
<span id="cb56-116"><a href="#cb56-116" aria-hidden="true" tabindex="-1"></a><span class="in">  ylab("Gene B") +</span></span>
<span id="cb56-117"><a href="#cb56-117" aria-hidden="true" tabindex="-1"></a><span class="in">  ggtitle("Example Expression Values from Four Cells") +</span></span>
<span id="cb56-118"><a href="#cb56-118" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.title = element_text(hjust = 0.5))</span></span>
<span id="cb56-119"><a href="#cb56-119" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-120"><a href="#cb56-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-121"><a href="#cb56-121" aria-hidden="true" tabindex="-1"></a>**2. Translate the points from their raw expression coordinates to coordinates centered around the center of the data**</span>
<span id="cb56-122"><a href="#cb56-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-123"><a href="#cb56-123" aria-hidden="true" tabindex="-1"></a>Next, we need to shift the points so that the center of the data corresponds to the origin. We can do this by subtracting the mean value from each gene from the raw values.</span>
<span id="cb56-124"><a href="#cb56-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-125"><a href="#cb56-125" aria-hidden="true" tabindex="-1"></a><span class="in">```{r recentered_data}</span></span>
<span id="cb56-126"><a href="#cb56-126" aria-hidden="true" tabindex="-1"></a><span class="in"># Shift the data points so that they data is centered on the origin</span></span>
<span id="cb56-127"><a href="#cb56-127" aria-hidden="true" tabindex="-1"></a><span class="in">recentered_expression_tibble &lt;- expression_tibble %&gt;%</span></span>
<span id="cb56-128"><a href="#cb56-128" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(</span></span>
<span id="cb56-129"><a href="#cb56-129" aria-hidden="true" tabindex="-1"></a><span class="in">    Gene_A = Gene_A - Gene_A_mean,</span></span>
<span id="cb56-130"><a href="#cb56-130" aria-hidden="true" tabindex="-1"></a><span class="in">    Gene_B = Gene_B - Gene_B_mean</span></span>
<span id="cb56-131"><a href="#cb56-131" aria-hidden="true" tabindex="-1"></a><span class="in">  )</span></span>
<span id="cb56-132"><a href="#cb56-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-133"><a href="#cb56-133" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out the re-centered data</span></span>
<span id="cb56-134"><a href="#cb56-134" aria-hidden="true" tabindex="-1"></a><span class="in">recentered_expression_tibble</span></span>
<span id="cb56-135"><a href="#cb56-135" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-136"><a href="#cb56-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-137"><a href="#cb56-137" aria-hidden="true" tabindex="-1"></a>As with any data analysis, it is always a good idea to visualize your data to ensure that it appears to behaving how you intended. So let's visualize our re-centered data to ensure that the data appears centered around the origin. </span>
<span id="cb56-138"><a href="#cb56-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-139"><a href="#cb56-139" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plot_recentered_data}</span></span>
<span id="cb56-140"><a href="#cb56-140" aria-hidden="true" tabindex="-1"></a><span class="in"># Plot the raw data after it has been shifted to have the center of the data align with the origin</span></span>
<span id="cb56-141"><a href="#cb56-141" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(recentered_expression_tibble, aes(x = Gene_A, y = Gene_B, label = cells)) +</span></span>
<span id="cb56-142"><a href="#cb56-142" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(color = "cornflowerblue") +</span></span>
<span id="cb56-143"><a href="#cb56-143" aria-hidden="true" tabindex="-1"></a><span class="in">  annotate("point", x = 0, y = 0, color = "red", size = 3) +</span></span>
<span id="cb56-144"><a href="#cb56-144" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(hjust = 0, vjust = -1) +</span></span>
<span id="cb56-145"><a href="#cb56-145" aria-hidden="true" tabindex="-1"></a><span class="in">  annotate("text", x = 0, y = 0, color = "red", label="New Center", hjust = 0, vjust = -1) +</span></span>
<span id="cb56-146"><a href="#cb56-146" aria-hidden="true" tabindex="-1"></a><span class="in">  xlim(-50, 50) +</span></span>
<span id="cb56-147"><a href="#cb56-147" aria-hidden="true" tabindex="-1"></a><span class="in">  ylim(-50, 50) +</span></span>
<span id="cb56-148"><a href="#cb56-148" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() +</span></span>
<span id="cb56-149"><a href="#cb56-149" aria-hidden="true" tabindex="-1"></a><span class="in">  xlab("Gene A") +</span></span>
<span id="cb56-150"><a href="#cb56-150" aria-hidden="true" tabindex="-1"></a><span class="in">  ylab("Gene B") +</span></span>
<span id="cb56-151"><a href="#cb56-151" aria-hidden="true" tabindex="-1"></a><span class="in">  ggtitle("Example Re-centered Expression Values from Four Cells") +</span></span>
<span id="cb56-152"><a href="#cb56-152" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.title = element_text(hjust = 0.5))</span></span>
<span id="cb56-153"><a href="#cb56-153" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-154"><a href="#cb56-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-155"><a href="#cb56-155" aria-hidden="true" tabindex="-1"></a><span class="fu">## Create a covariance matrix</span></span>
<span id="cb56-156"><a href="#cb56-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-157"><a href="#cb56-157" aria-hidden="true" tabindex="-1"></a>The next step that we will after we have re-centered our data around the origin is to calculate the covariance matrix. As a reminder, the covariance is the joint variability of two random variables. In this case our random variables are the genes. The illustration below we provide examples for positive and negative covariance values along with a covariance that is near zero.</span>
<span id="cb56-158"><a href="#cb56-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-159"><a href="#cb56-159" aria-hidden="true" tabindex="-1"></a>&lt;img src="../img/Covariance.png" width="800"&gt;</span>
<span id="cb56-160"><a href="#cb56-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-161"><a href="#cb56-161" aria-hidden="true" tabindex="-1"></a>The equation to estimate the sample covariance is:</span>
<span id="cb56-162"><a href="#cb56-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-163"><a href="#cb56-163" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?<span class="dv">&amp;space;</span>cov(X,Y)=\frac{1}{n - 1}\sum_{i=1}^{n}(x_{i}-\bar{x})(y_{i}-\bar{y})" width="300"&gt;</span>
<span id="cb56-164"><a href="#cb56-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-165"><a href="#cb56-165" aria-hidden="true" tabindex="-1"></a>Where:</span>
<span id="cb56-166"><a href="#cb56-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-167"><a href="#cb56-167" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*X* is Gene A</span>
<span id="cb56-168"><a href="#cb56-168" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*Y* is Gene B</span>
<span id="cb56-169"><a href="#cb56-169" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*n* is the number of cells</span>
<span id="cb56-170"><a href="#cb56-170" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*x~i~* is an expression value for Gene A</span>
<span id="cb56-171"><a href="#cb56-171" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*$\bar{x}$* is the mean expression value for Gene A</span>
<span id="cb56-172"><a href="#cb56-172" aria-hidden="true" tabindex="-1"></a><span class="ss"> - </span>*y~i~* is an expression value for Gene B</span>
<span id="cb56-173"><a href="#cb56-173" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>*$\bar{y}$* is the mean expression value for Gene B</span>
<span id="cb56-174"><a href="#cb56-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-175"><a href="#cb56-175" aria-hidden="true" tabindex="-1"></a>Now we will calculate the covariance for all pairwise comparisons for our re-centered expression data.</span>
<span id="cb56-176"><a href="#cb56-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-177"><a href="#cb56-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r create_covariance_matrix}</span></span>
<span id="cb56-178"><a href="#cb56-178" aria-hidden="true" tabindex="-1"></a><span class="in"># Move the cell IDs to the rownames and convert the tibble to a matrix</span></span>
<span id="cb56-179"><a href="#cb56-179" aria-hidden="true" tabindex="-1"></a><span class="in">recentered_expression_matrix &lt;- recentered_expression_tibble %&gt;% </span></span>
<span id="cb56-180"><a href="#cb56-180" aria-hidden="true" tabindex="-1"></a><span class="in">  column_to_rownames("cells") %&gt;% </span></span>
<span id="cb56-181"><a href="#cb56-181" aria-hidden="true" tabindex="-1"></a><span class="in">  as.matrix()</span></span>
<span id="cb56-182"><a href="#cb56-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-183"><a href="#cb56-183" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a covariance matrix</span></span>
<span id="cb56-184"><a href="#cb56-184" aria-hidden="true" tabindex="-1"></a><span class="in">cov_matrix &lt;- cov(recentered_expression_matrix)</span></span>
<span id="cb56-185"><a href="#cb56-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-186"><a href="#cb56-186" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out the covariance matrix</span></span>
<span id="cb56-187"><a href="#cb56-187" aria-hidden="true" tabindex="-1"></a><span class="in">cov_matrix</span></span>
<span id="cb56-188"><a href="#cb56-188" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-189"><a href="#cb56-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-190"><a href="#cb56-190" aria-hidden="true" tabindex="-1"></a>Let's go ahead and manually check a covariance by hand to ensure that we trust the outputted covariance matrix. Let's estimate the covariance of Gene A and Gene B.</span>
<span id="cb56-191"><a href="#cb56-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-192"><a href="#cb56-192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r handcheck_covariance_matrix}</span></span>
<span id="cb56-193"><a href="#cb56-193" aria-hidden="true" tabindex="-1"></a><span class="in"># Estimate the covariance of Gene A and Gene B by hand</span></span>
<span id="cb56-194"><a href="#cb56-194" aria-hidden="true" tabindex="-1"></a><span class="in">(1/(4-1))*((0-25)*(4-27.25) + (12-25)*(30-27.25) + (65-25)*(57-27.25) + (23-25)*(18-27.25))</span></span>
<span id="cb56-195"><a href="#cb56-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-196"><a href="#cb56-196" aria-hidden="true" tabindex="-1"></a><span class="in"># Or you can use the covariance function</span></span>
<span id="cb56-197"><a href="#cb56-197" aria-hidden="true" tabindex="-1"></a><span class="in">cov(recentered_expression_matrix[, "Gene_A"], recentered_expression_matrix[, "Gene_B"])</span></span>
<span id="cb56-198"><a href="#cb56-198" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-199"><a href="#cb56-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-200"><a href="#cb56-200" aria-hidden="true" tabindex="-1"></a>This value should match the covariance found in our covariance matrix for the covariance between Gene A and Gene B.</span>
<span id="cb56-201"><a href="#cb56-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-202"><a href="#cb56-202" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb56-203"><a href="#cb56-203" aria-hidden="true" tabindex="-1"></a><span class="fu"># [**Exercise 1**](04_theory_of_PCA_Answer-key.qmd#exercise-1)</span></span>
<span id="cb56-204"><a href="#cb56-204" aria-hidden="true" tabindex="-1"></a>Exercise 1</span>
<span id="cb56-205"><a href="#cb56-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-206"><a href="#cb56-206" aria-hidden="true" tabindex="-1"></a>There are a couple properties of variance and co-variance that we can verify:</span>
<span id="cb56-207"><a href="#cb56-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-208"><a href="#cb56-208" aria-hidden="true" tabindex="-1"></a>*cov(X,X)* is equal to *var(X)*. We can observe this is mathematically below:</span>
<span id="cb56-209"><a href="#cb56-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-210"><a href="#cb56-210" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?<span class="dv">&amp;space;</span>cov(X,X)=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})(x_{i}-\bar{x})=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})^{2}=var(X)" width="500"&gt;</span>
<span id="cb56-211"><a href="#cb56-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-212"><a href="#cb56-212" aria-hidden="true" tabindex="-1"></a>As a result, you will sometimes see covariance matrices written as:</span>
<span id="cb56-213"><a href="#cb56-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-214"><a href="#cb56-214" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?\begin{bmatrix}var(X)&amp;cov(X,Y)&amp;...&amp;cov(X,Z)\\cov(Y,X)&amp;var(Y)&amp;...&amp;cov(Y,Z)\\...&amp;...&amp;...&amp;...\\cov(Z,X)&amp;cov(Z,Y)&amp;...&amp;var(Z)\\\end{bmatrix}" width="300"&gt;</span>
<span id="cb56-215"><a href="#cb56-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-216"><a href="#cb56-216" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Confirm this property by estimating the variance for Gene A.</span>
<span id="cb56-217"><a href="#cb56-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-218"><a href="#cb56-218" aria-hidden="true" tabindex="-1"></a><span class="in">```{r covariance_check_1}</span></span>
<span id="cb56-219"><a href="#cb56-219" aria-hidden="true" tabindex="-1"></a><span class="in"># Estimate the variance for Gene A</span></span>
<span id="cb56-220"><a href="#cb56-220" aria-hidden="true" tabindex="-1"></a><span class="in">var(recentered_expression_matrix[, "Gene_A"])</span></span>
<span id="cb56-221"><a href="#cb56-221" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-222"><a href="#cb56-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-223"><a href="#cb56-223" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Now estimate the covariance for Gene A and Gene A</span>
<span id="cb56-224"><a href="#cb56-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-225"><a href="#cb56-225" aria-hidden="true" tabindex="-1"></a><span class="in">```{r covariance_check_2}</span></span>
<span id="cb56-226"><a href="#cb56-226" aria-hidden="true" tabindex="-1"></a><span class="in"># Estimate the covariance for Gene A and Gene A</span></span>
<span id="cb56-227"><a href="#cb56-227" aria-hidden="true" tabindex="-1"></a><span class="in">cov(recentered_expression_matrix[, "Gene_A"], recentered_expression_matrix[, "Gene_A"])</span></span>
<span id="cb56-228"><a href="#cb56-228" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-229"><a href="#cb56-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-230"><a href="#cb56-230" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Is the value the same? Does it match the value in the covariance matrix for Gene A and Gene A?</span>
<span id="cb56-231"><a href="#cb56-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-232"><a href="#cb56-232" aria-hidden="true" tabindex="-1"></a><span class="in">```{r covariance_matrix_check}</span></span>
<span id="cb56-233"><a href="#cb56-233" aria-hidden="true" tabindex="-1"></a><span class="in"># Extract the covariance estimate of Gene A and Gene A from the covariance matrix</span></span>
<span id="cb56-234"><a href="#cb56-234" aria-hidden="true" tabindex="-1"></a><span class="in">cov_matrix["Gene_A", "Gene_A"]</span></span>
<span id="cb56-235"><a href="#cb56-235" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-236"><a href="#cb56-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-237"><a href="#cb56-237" aria-hidden="true" tabindex="-1"></a>*cov(X,Y)* is equal to *cov(Y,X)*. We can observe this is mathematically below:</span>
<span id="cb56-238"><a href="#cb56-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-239"><a href="#cb56-239" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?<span class="dv">&amp;space;</span>cov(X,Y)=\frac{1}{n-1}\sum_{i=1}^{n}(x_{i}-\bar{x})(y_{i}-\bar{y})=\frac{1}{n-1}\sum_{i=1}^{n}(y_{i}-\bar{y})(x_{i}-\bar{x})=cov(Y,X)" width="500"&gt;</span>
<span id="cb56-240"><a href="#cb56-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-241"><a href="#cb56-241" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Estimate the covariance of Gene B and Gene A.</span>
<span id="cb56-242"><a href="#cb56-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-243"><a href="#cb56-243" aria-hidden="true" tabindex="-1"></a><span class="in">```{r covariance_check_3}</span></span>
<span id="cb56-244"><a href="#cb56-244" aria-hidden="true" tabindex="-1"></a><span class="in"># Estimate the covariance of Gene B and Gene A</span></span>
<span id="cb56-245"><a href="#cb56-245" aria-hidden="true" tabindex="-1"></a><span class="in">cov(recentered_expression_tibble[, "Gene_B"], recentered_expression_tibble[, "Gene_A"])</span></span>
<span id="cb56-246"><a href="#cb56-246" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-247"><a href="#cb56-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-248"><a href="#cb56-248" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>How does this compare to the covariance that we estimated by hand?</span>
<span id="cb56-249"><a href="#cb56-249" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb56-250"><a href="#cb56-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-251"><a href="#cb56-251" aria-hidden="true" tabindex="-1"></a><span class="fu"># Calculating the Principal Components</span></span>
<span id="cb56-252"><a href="#cb56-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-253"><a href="#cb56-253" aria-hidden="true" tabindex="-1"></a><span class="fu">## Calculate Eigenvalues and Eigenvectors from the Covariance Matrix</span></span>
<span id="cb56-254"><a href="#cb56-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-255"><a href="#cb56-255" aria-hidden="true" tabindex="-1"></a>First, we will need to calculate our eigenvalues, which are a measure of variance for a principal component. We can then use eigenvalues to calculate our eigenvectors, but R will do this all in a single step for us. We can think of the eigenvectors as the weights, or influence, for transforming our re-centered expression values into PCA space. </span>
<span id="cb56-256"><a href="#cb56-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-257"><a href="#cb56-257" aria-hidden="true" tabindex="-1"></a>Now that we have a covariance matrix for the re-centered expression values, we need to estimate the eigenvalues and eigenvectors for the covariance matrix. </span>
<span id="cb56-258"><a href="#cb56-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-259"><a href="#cb56-259" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigen_calculation}</span></span>
<span id="cb56-260"><a href="#cb56-260" aria-hidden="true" tabindex="-1"></a><span class="in"># Find the eigenvalues and eigenvectors of the covariance matrix</span></span>
<span id="cb56-261"><a href="#cb56-261" aria-hidden="true" tabindex="-1"></a><span class="in">eig &lt;- eigen(cov_matrix)</span></span>
<span id="cb56-262"><a href="#cb56-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-263"><a href="#cb56-263" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the output from eigen()</span></span>
<span id="cb56-264"><a href="#cb56-264" aria-hidden="true" tabindex="-1"></a><span class="in">eig</span></span>
<span id="cb56-265"><a href="#cb56-265" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-266"><a href="#cb56-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-267"><a href="#cb56-267" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true" title="Click here to see how to calculate eigenvalues and eigenvectors"}</span>
<span id="cb56-268"><a href="#cb56-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-269"><a href="#cb56-269" aria-hidden="true" tabindex="-1"></a>**Calculating Eigenvalues**</span>
<span id="cb56-270"><a href="#cb56-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-271"><a href="#cb56-271" aria-hidden="true" tabindex="-1"></a>If we wanted to do this by hand then we need to set the determinant of the covariance matrix minus $\lambda$ times the identify matrix equal to 0 and solving for $\lambda$.</span>
<span id="cb56-272"><a href="#cb56-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-273"><a href="#cb56-273" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?det(\begin{bmatrix}var(X)&amp;cov(X,Y)\\cov(Y,X)&amp;var(Y)\\\end{bmatrix}-\lambda\begin{bmatrix}1&amp;0\\0&amp;1\\\end{bmatrix})=0" width="500"&gt;</span>
<span id="cb56-274"><a href="#cb56-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-275"><a href="#cb56-275" aria-hidden="true" tabindex="-1"></a>In this case it can be simplified to:</span>
<span id="cb56-276"><a href="#cb56-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-277"><a href="#cb56-277" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?(var(X)-\lambda)(var(Y)-\lambda)-(cov(X,Y))(cov(Y,X))=0" width="500"&gt;</span>
<span id="cb56-278"><a href="#cb56-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-279"><a href="#cb56-279" aria-hidden="true" tabindex="-1"></a>We can now replace the variance and covariance values with the values that we have from the covariance matrix and solve for $\lambda$. This results in two lambda values: **$\lambda$~1~ = `r round(eig$values[1], digits = 3)`** and **$\lambda$~2~ = `r round(eig$values[2], digits = 3)`**.</span>
<span id="cb56-280"><a href="#cb56-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-281"><a href="#cb56-281" aria-hidden="true" tabindex="-1"></a>We can see that this is equal to the eigenvalues returned from our <span class="in">`eig`</span> object:</span>
<span id="cb56-282"><a href="#cb56-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-283"><a href="#cb56-283" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvalues}</span></span>
<span id="cb56-284"><a href="#cb56-284" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvalues</span></span>
<span id="cb56-285"><a href="#cb56-285" aria-hidden="true" tabindex="-1"></a><span class="in">eig$values</span></span>
<span id="cb56-286"><a href="#cb56-286" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-287"><a href="#cb56-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-288"><a href="#cb56-288" aria-hidden="true" tabindex="-1"></a>**Calculating Eigenvectors**</span>
<span id="cb56-289"><a href="#cb56-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-290"><a href="#cb56-290" aria-hidden="true" tabindex="-1"></a>In order to find the eigenvectors, we need to substitute the variacne and covariance values from the covariance matrix along with one of our *$\lambda$* values, *$\lambda$~1~* to find the associated eigenvector, *v~1~*. This results in *v~1~* being equal to <span class="in">`r eig$vectors[,1]`</span>.</span>
<span id="cb56-291"><a href="#cb56-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-292"><a href="#cb56-292" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?(\begin{bmatrix}var(X)&amp;cov(X,Y)\\cov(Y,X)&amp;var(Y)\\\end{bmatrix}-\lambda<span class="dv">&amp;space;</span>_{1}\begin{bmatrix}1&amp;0\\0&amp;1\\\end{bmatrix})v_{1}=0" width="500"&gt;</span>
<span id="cb56-293"><a href="#cb56-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-294"><a href="#cb56-294" aria-hidden="true" tabindex="-1"></a>We can confirm this by plugging our covariance matrix, eigenvalues and eigenvectors back into the above equation and it should return 0, or near 0 (due to rounding errors).</span>
<span id="cb56-295"><a href="#cb56-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-296"><a href="#cb56-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r check_eigen}</span></span>
<span id="cb56-297"><a href="#cb56-297" aria-hidden="true" tabindex="-1"></a><span class="in"># Check the eigenvalues and eigenvectors</span></span>
<span id="cb56-298"><a href="#cb56-298" aria-hidden="true" tabindex="-1"></a><span class="in">(cov_matrix - eig$values[1]*diag(2)) %*% eig$vectors[,1]</span></span>
<span id="cb56-299"><a href="#cb56-299" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-300"><a href="#cb56-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-301"><a href="#cb56-301" aria-hidden="true" tabindex="-1"></a>We then repeat this process for *$\lambda$~2~* in order to to find its associated eigenvector, *v~2~*, which results in <span class="in">`r eig$vectors[,2]`</span>. Thus, our resulting matrix of eigenvectors appears as:</span>
<span id="cb56-302"><a href="#cb56-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-303"><a href="#cb56-303" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvectors}</span></span>
<span id="cb56-304"><a href="#cb56-304" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvectors</span></span>
<span id="cb56-305"><a href="#cb56-305" aria-hidden="true" tabindex="-1"></a><span class="in">eig$vectors</span></span>
<span id="cb56-306"><a href="#cb56-306" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-307"><a href="#cb56-307" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb56-308"><a href="#cb56-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-309"><a href="#cb56-309" aria-hidden="true" tabindex="-1"></a><span class="fu">## Principal Component Scores</span></span>
<span id="cb56-310"><a href="#cb56-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-311"><a href="#cb56-311" aria-hidden="true" tabindex="-1"></a>Now that we have our re-centered expression matrix and eigenvectors, we can perform matrix multiplication to obtain our principal component scores. This is where we are transforming our data from the re-centered expression into Principal Components space by the weight, or influence, of the eigenvector.</span>
<span id="cb56-312"><a href="#cb56-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-313"><a href="#cb56-313" aria-hidden="true" tabindex="-1"></a>&lt;img src="../img/PC_generation.png" width="800"&gt;</span>
<span id="cb56-314"><a href="#cb56-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-315"><a href="#cb56-315" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pc_scores}</span></span>
<span id="cb56-316"><a href="#cb56-316" aria-hidden="true" tabindex="-1"></a><span class="in"># Transform the data into PC space by multiply the re-centered expression matrix by the eigenvectors</span></span>
<span id="cb56-317"><a href="#cb56-317" aria-hidden="true" tabindex="-1"></a><span class="in">pc_scores &lt;- recentered_expression_matrix %*% eig$vectors</span></span>
<span id="cb56-318"><a href="#cb56-318" aria-hidden="true" tabindex="-1"></a><span class="in"># Name the columns in pc_scores object</span></span>
<span id="cb56-319"><a href="#cb56-319" aria-hidden="true" tabindex="-1"></a><span class="in">colnames(pc_scores) &lt;- c("PC_1", "PC_2")</span></span>
<span id="cb56-320"><a href="#cb56-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-321"><a href="#cb56-321" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out the pc_scores object</span></span>
<span id="cb56-322"><a href="#cb56-322" aria-hidden="true" tabindex="-1"></a><span class="in">pc_scores</span></span>
<span id="cb56-323"><a href="#cb56-323" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-324"><a href="#cb56-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-325"><a href="#cb56-325" aria-hidden="true" tabindex="-1"></a><span class="fu">## Percent Variance Explained</span></span>
<span id="cb56-326"><a href="#cb56-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-327"><a href="#cb56-327" aria-hidden="true" tabindex="-1"></a>We are likely interested in knowing in knowing the amount of variance explained in our data by each principal component. As we alluded to earlier, we can use the eigenvalues to help us with this. The sum of all of the eigenvalues captures the total variance explained by the principal components analysis.</span>
<span id="cb56-328"><a href="#cb56-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-329"><a href="#cb56-329" aria-hidden="true" tabindex="-1"></a>Thus, if we want to know the proportion of the variance explained by each principal component, then we would need to divide each eigenvalue by the total variance explained (the sum of the eigenvalues) and if we wanted this as a percentage then we would need to multiply it by 100.</span>
<span id="cb56-330"><a href="#cb56-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-331"><a href="#cb56-331" aria-hidden="true" tabindex="-1"></a><span class="in">```{r percent_explained}</span></span>
<span id="cb56-332"><a href="#cb56-332" aria-hidden="true" tabindex="-1"></a><span class="in"># Calculate the percent of variance explained by each PC using the eigenvalues</span></span>
<span id="cb56-333"><a href="#cb56-333" aria-hidden="true" tabindex="-1"></a><span class="in">pct_var_explained &lt;- (eig$values / sum(eig$values)) * 100</span></span>
<span id="cb56-334"><a href="#cb56-334" aria-hidden="true" tabindex="-1"></a><span class="in"># Name the elements of the pct_var_explained by their PC</span></span>
<span id="cb56-335"><a href="#cb56-335" aria-hidden="true" tabindex="-1"></a><span class="in">names(pct_var_explained) &lt;- c("PC_1", "PC_2")</span></span>
<span id="cb56-336"><a href="#cb56-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-337"><a href="#cb56-337" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out the percent of variance explained by each PC</span></span>
<span id="cb56-338"><a href="#cb56-338" aria-hidden="true" tabindex="-1"></a><span class="in">pct_var_explained</span></span>
<span id="cb56-339"><a href="#cb56-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-340"><a href="#cb56-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-341"><a href="#cb56-341" aria-hidden="true" tabindex="-1"></a><span class="fu">## Plotting our Principal Components</span></span>
<span id="cb56-342"><a href="#cb56-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-343"><a href="#cb56-343" aria-hidden="true" tabindex="-1"></a>With out principal components in hand, we will create a visualization of the principal components.</span>
<span id="cb56-344"><a href="#cb56-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-345"><a href="#cb56-345" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plotting_PCA_hand}</span></span>
<span id="cb56-346"><a href="#cb56-346" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a tibble to hold the PC scores we found and also make the Cell IDs into a column</span></span>
<span id="cb56-347"><a href="#cb56-347" aria-hidden="true" tabindex="-1"></a><span class="in">pc_scores_tibble &lt;- pc_scores %&gt;% </span></span>
<span id="cb56-348"><a href="#cb56-348" aria-hidden="true" tabindex="-1"></a><span class="in">  as.data.frame() %&gt;% </span></span>
<span id="cb56-349"><a href="#cb56-349" aria-hidden="true" tabindex="-1"></a><span class="in">  rownames_to_column("cells") %&gt;% </span></span>
<span id="cb56-350"><a href="#cb56-350" aria-hidden="true" tabindex="-1"></a><span class="in">  as_tibble()</span></span>
<span id="cb56-351"><a href="#cb56-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-352"><a href="#cb56-352" aria-hidden="true" tabindex="-1"></a><span class="in"># Plot the PC scores we found</span></span>
<span id="cb56-353"><a href="#cb56-353" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(pc_scores_tibble, aes(x = PC_1, y = PC_2, label = cells)) +</span></span>
<span id="cb56-354"><a href="#cb56-354" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point( color = "cornflowerblue") +</span></span>
<span id="cb56-355"><a href="#cb56-355" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(hjust = 0, vjust = -1) +</span></span>
<span id="cb56-356"><a href="#cb56-356" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() +</span></span>
<span id="cb56-357"><a href="#cb56-357" aria-hidden="true" tabindex="-1"></a><span class="in">  xlim(-50, 50) +</span></span>
<span id="cb56-358"><a href="#cb56-358" aria-hidden="true" tabindex="-1"></a><span class="in">  ylim(-12, 12) +</span></span>
<span id="cb56-359"><a href="#cb56-359" aria-hidden="true" tabindex="-1"></a><span class="in">  xlab(paste0("PC 1 (Variance Explained ", round(pct_var_explained["PC_1"], digits = 2),"%)")) +</span></span>
<span id="cb56-360"><a href="#cb56-360" aria-hidden="true" tabindex="-1"></a><span class="in">  ylab(paste0("PC 2 (Variance Explained ", round(pct_var_explained["PC_2"], digits = 2),"%)")) +</span></span>
<span id="cb56-361"><a href="#cb56-361" aria-hidden="true" tabindex="-1"></a><span class="in">  ggtitle("PCA of Expression Values from Four Cells") +</span></span>
<span id="cb56-362"><a href="#cb56-362" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.title = element_text(hjust = 0.5))</span></span>
<span id="cb56-363"><a href="#cb56-363" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-364"><a href="#cb56-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-365"><a href="#cb56-365" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb56-366"><a href="#cb56-366" aria-hidden="true" tabindex="-1"></a><span class="fu"># [**Exercise 2**](04_theory_of_PCA_Answer-key.qmd#exercise-2)</span></span>
<span id="cb56-367"><a href="#cb56-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-368"><a href="#cb56-368" aria-hidden="true" tabindex="-1"></a>When looking at the percent explained by each principal component, the first principal component should explain the most and each of the following principal components should explain less than the previous principal component. Let's have a look at our <span class="in">`pct_var_explained`</span> object, are our results congruent with the expectation this expectation?</span>
<span id="cb56-369"><a href="#cb56-369" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb56-370"><a href="#cb56-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-371"><a href="#cb56-371" aria-hidden="true" tabindex="-1"></a><span class="fu"># `prcomp()` comparison</span></span>
<span id="cb56-372"><a href="#cb56-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-373"><a href="#cb56-373" aria-hidden="true" tabindex="-1"></a>At this point we have completed a principal components analysis, but we can also compare it to how R does a principal components analysis and see how our results compare. One way to do a principal components analysis in R is to us the <span class="in">`prcomp()`</span> function in R.</span>
<span id="cb56-374"><a href="#cb56-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-375"><a href="#cb56-375" aria-hidden="true" tabindex="-1"></a>In order to run <span class="in">`prcomp()`</span>, we need to start from an expression matrix and move the cell IDs to the rownames:</span>
<span id="cb56-376"><a href="#cb56-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-377"><a href="#cb56-377" aria-hidden="true" tabindex="-1"></a><span class="in">```{r prcomp_check}</span></span>
<span id="cb56-378"><a href="#cb56-378" aria-hidden="true" tabindex="-1"></a><span class="in"># Run prcomp() on the expression tibble after moving the Cell IDs to be rownames</span></span>
<span id="cb56-379"><a href="#cb56-379" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_PCA &lt;- expression_tibble %&gt;% </span></span>
<span id="cb56-380"><a href="#cb56-380" aria-hidden="true" tabindex="-1"></a><span class="in">  column_to_rownames("cells") %&gt;% </span></span>
<span id="cb56-381"><a href="#cb56-381" aria-hidden="true" tabindex="-1"></a><span class="in">  prcomp()</span></span>
<span id="cb56-382"><a href="#cb56-382" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-383"><a href="#cb56-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-384"><a href="#cb56-384" aria-hidden="true" tabindex="-1"></a>Let's inspect the output from <span class="in">`prcomp()`</span> to ensure that we are getting the same output:</span>
<span id="cb56-385"><a href="#cb56-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-386"><a href="#cb56-386" aria-hidden="true" tabindex="-1"></a><span class="fu">## Center of the Data</span></span>
<span id="cb56-387"><a href="#cb56-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-388"><a href="#cb56-388" aria-hidden="true" tabindex="-1"></a>First we can ensure that we have the same center of the data. <span class="in">`prcomp()`</span> says that the center of the data is:</span>
<span id="cb56-389"><a href="#cb56-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-390"><a href="#cb56-390" aria-hidden="true" tabindex="-1"></a><span class="in">```{r center_prcomp}</span></span>
<span id="cb56-391"><a href="#cb56-391" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the center of the data found by prcomp()</span></span>
<span id="cb56-392"><a href="#cb56-392" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_PCA$center</span></span>
<span id="cb56-393"><a href="#cb56-393" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-394"><a href="#cb56-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-395"><a href="#cb56-395" aria-hidden="true" tabindex="-1"></a>We can compare this with what we derived as the center of the data:</span>
<span id="cb56-396"><a href="#cb56-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-397"><a href="#cb56-397" aria-hidden="true" tabindex="-1"></a><span class="in">```{r center_hand}</span></span>
<span id="cb56-398"><a href="#cb56-398" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the center of the data we found</span></span>
<span id="cb56-399"><a href="#cb56-399" aria-hidden="true" tabindex="-1"></a><span class="in">center_of_data</span></span>
<span id="cb56-400"><a href="#cb56-400" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-401"><a href="#cb56-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-402"><a href="#cb56-402" aria-hidden="true" tabindex="-1"></a>They match which is a great sign that we did this correctly!</span>
<span id="cb56-403"><a href="#cb56-403" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-404"><a href="#cb56-404" aria-hidden="true" tabindex="-1"></a><span class="fu">## Eigenvalues</span></span>
<span id="cb56-405"><a href="#cb56-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-406"><a href="#cb56-406" aria-hidden="true" tabindex="-1"></a>Note that eigenvalues are the variance for a particular principal component and that variance is the square of the standard deviation as shown below. </span>
<span id="cb56-407"><a href="#cb56-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-408"><a href="#cb56-408" aria-hidden="true" tabindex="-1"></a>&lt;img src="https://latex.codecogs.com/svg.image?var(X)=sd(X)^{2}" width="150"&gt;</span>
<span id="cb56-409"><a href="#cb56-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-410"><a href="#cb56-410" aria-hidden="true" tabindex="-1"></a><span class="in">`prcomp()`</span> reports the standard deviation, so we will need to square the items in the <span class="in">`prcomp_PCA$sdev`</span> slot in order to recover our variance, or eigenvalues.</span>
<span id="cb56-411"><a href="#cb56-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-412"><a href="#cb56-412" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvalues_prcomp}</span></span>
<span id="cb56-413"><a href="#cb56-413" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvalues found by prcomp() by squaring prcomp_PCA$sdev</span></span>
<span id="cb56-414"><a href="#cb56-414" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_eigenvalues &lt;- prcomp_PCA$sdev ** 2</span></span>
<span id="cb56-415"><a href="#cb56-415" aria-hidden="true" tabindex="-1"></a><span class="in"># Name the elements of the prcomp_eigenvalues by their PC</span></span>
<span id="cb56-416"><a href="#cb56-416" aria-hidden="true" tabindex="-1"></a><span class="in">names(prcomp_eigenvalues) &lt;- c("PC_1", "PC_2")</span></span>
<span id="cb56-417"><a href="#cb56-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-418"><a href="#cb56-418" aria-hidden="true" tabindex="-1"></a><span class="in"># Print out prcomp_eigenvalues</span></span>
<span id="cb56-419"><a href="#cb56-419" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_eigenvalues</span></span>
<span id="cb56-420"><a href="#cb56-420" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-421"><a href="#cb56-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-422"><a href="#cb56-422" aria-hidden="true" tabindex="-1"></a>We can compare this with our estimation of eiganvalues by hand:</span>
<span id="cb56-423"><a href="#cb56-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-424"><a href="#cb56-424" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvalue_hand}</span></span>
<span id="cb56-425"><a href="#cb56-425" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvalues we found</span></span>
<span id="cb56-426"><a href="#cb56-426" aria-hidden="true" tabindex="-1"></a><span class="in">eig$values</span></span>
<span id="cb56-427"><a href="#cb56-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-428"><a href="#cb56-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-429"><a href="#cb56-429" aria-hidden="true" tabindex="-1"></a><span class="fu">## Eigenvectors</span></span>
<span id="cb56-430"><a href="#cb56-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-431"><a href="#cb56-431" aria-hidden="true" tabindex="-1"></a>Let's go ahead and compare the eigenvectors from <span class="in">`prcomp()`</span> to the eigenvectors that we calculated. <span class="in">`prcomp()`</span> calculated the eigenvectors as:</span>
<span id="cb56-432"><a href="#cb56-432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-433"><a href="#cb56-433" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvector_prcomp}</span></span>
<span id="cb56-434"><a href="#cb56-434" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvectors found by prcomp()</span></span>
<span id="cb56-435"><a href="#cb56-435" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_PCA$rotation</span></span>
<span id="cb56-436"><a href="#cb56-436" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-437"><a href="#cb56-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-438"><a href="#cb56-438" aria-hidden="true" tabindex="-1"></a>While we calculated the eigenvectors as:</span>
<span id="cb56-439"><a href="#cb56-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-440"><a href="#cb56-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eigenvector_hand}</span></span>
<span id="cb56-441"><a href="#cb56-441" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the eigenvectors we found</span></span>
<span id="cb56-442"><a href="#cb56-442" aria-hidden="true" tabindex="-1"></a><span class="in">eig$vectors</span></span>
<span id="cb56-443"><a href="#cb56-443" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-444"><a href="#cb56-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-445"><a href="#cb56-445" aria-hidden="true" tabindex="-1"></a>We can see that the values are the same, but the sign is flipped between the two methods. This is because eigenvectors are only defined "up to sign". This means for a given eigenvalue, the eigenvector *v~1~* and *-v~1~* are the same line, but point in different directions. This does not impact our principal components calculations. However, it could mean that a plot derived *v~1~* will look rotated 180&amp;deg; when compared to a plot derived from *-v~1~*. Regardless, it looks like we have the same eigenvectors in both approaches.</span>
<span id="cb56-446"><a href="#cb56-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-447"><a href="#cb56-447" aria-hidden="true" tabindex="-1"></a><span class="fu">## PC Scores</span></span>
<span id="cb56-448"><a href="#cb56-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-449"><a href="#cb56-449" aria-hidden="true" tabindex="-1"></a>Finally, we can check our PC scores to ensure that they are also equivalent. We should note that since the eigenvector from <span class="in">`prcomp()`</span> and the eigenvector we calculated differed by sign and eigenvectors are multiplied by the re-centered expression, then we should expect our principal components to also differ by sign. The principal components from <span class="in">`prcomp()`</span> are:</span>
<span id="cb56-450"><a href="#cb56-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-451"><a href="#cb56-451" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pc_scores_prcomp}</span></span>
<span id="cb56-452"><a href="#cb56-452" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the PC scores found by prcomp()</span></span>
<span id="cb56-453"><a href="#cb56-453" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_PCA$x</span></span>
<span id="cb56-454"><a href="#cb56-454" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-455"><a href="#cb56-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-456"><a href="#cb56-456" aria-hidden="true" tabindex="-1"></a>While our principal components were:</span>
<span id="cb56-457"><a href="#cb56-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-458"><a href="#cb56-458" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pc_scores_hand}</span></span>
<span id="cb56-459"><a href="#cb56-459" aria-hidden="true" tabindex="-1"></a><span class="in"># Print the PC scores we found</span></span>
<span id="cb56-460"><a href="#cb56-460" aria-hidden="true" tabindex="-1"></a><span class="in">pc_scores</span></span>
<span id="cb56-461"><a href="#cb56-461" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-462"><a href="#cb56-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-463"><a href="#cb56-463" aria-hidden="true" tabindex="-1"></a>Once again, the values are the same, but the only difference is the sign, which comes from the arbitrary direction of the eigenvector.</span>
<span id="cb56-464"><a href="#cb56-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-465"><a href="#cb56-465" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb56-466"><a href="#cb56-466" aria-hidden="true" tabindex="-1"></a><span class="fu"># [**Exercise 3**](04_theory_of_PCA_Answer-key.qmd#exercise-3)</span></span>
<span id="cb56-467"><a href="#cb56-467" aria-hidden="true" tabindex="-1"></a>Create a plot of the Principal Components Analysis derived from <span class="in">`prcomp()`</span>. Is it the same as the plot we derived except only rotated 180<span class="dv">&amp;deg;</span>?</span>
<span id="cb56-468"><a href="#cb56-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-469"><a href="#cb56-469" aria-hidden="true" tabindex="-1"></a><span class="in">```{r plotting_PCA_prcomp}</span></span>
<span id="cb56-470"><a href="#cb56-470" aria-hidden="true" tabindex="-1"></a><span class="in"># Create a tibble to hold the PC scores prcomp() found and also make the Cell IDs into a column</span></span>
<span id="cb56-471"><a href="#cb56-471" aria-hidden="true" tabindex="-1"></a><span class="in">prcomp_pc_scores_tibble &lt;- prcomp_PCA$x %&gt;% </span></span>
<span id="cb56-472"><a href="#cb56-472" aria-hidden="true" tabindex="-1"></a><span class="in">  as.data.frame() %&gt;% </span></span>
<span id="cb56-473"><a href="#cb56-473" aria-hidden="true" tabindex="-1"></a><span class="in">  rownames_to_column("cells") %&gt;% </span></span>
<span id="cb56-474"><a href="#cb56-474" aria-hidden="true" tabindex="-1"></a><span class="in">  as_tibble()</span></span>
<span id="cb56-475"><a href="#cb56-475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-476"><a href="#cb56-476" aria-hidden="true" tabindex="-1"></a><span class="in"># Plot the PC scores found by prcomp()</span></span>
<span id="cb56-477"><a href="#cb56-477" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(prcomp_pc_scores_tibble, aes(x = PC1, y = PC2, label = cells)) +</span></span>
<span id="cb56-478"><a href="#cb56-478" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point( color = "cornflowerblue") +</span></span>
<span id="cb56-479"><a href="#cb56-479" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(hjust = 0, vjust = -1) +</span></span>
<span id="cb56-480"><a href="#cb56-480" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() +</span></span>
<span id="cb56-481"><a href="#cb56-481" aria-hidden="true" tabindex="-1"></a><span class="in">  xlim(-50, 50) +</span></span>
<span id="cb56-482"><a href="#cb56-482" aria-hidden="true" tabindex="-1"></a><span class="in">  ylim(-12, 12) +</span></span>
<span id="cb56-483"><a href="#cb56-483" aria-hidden="true" tabindex="-1"></a><span class="in">  xlab(paste0("PC 1 (Variance Explained ", round(prcomp_eigenvalues["PC_1"], digits = 2),"%)")) +</span></span>
<span id="cb56-484"><a href="#cb56-484" aria-hidden="true" tabindex="-1"></a><span class="in">  ylab(paste0("PC 2 (Variance Explained ", round(prcomp_eigenvalues["PC_2"], digits = 2),"%)")) +</span></span>
<span id="cb56-485"><a href="#cb56-485" aria-hidden="true" tabindex="-1"></a><span class="in">  ggtitle("PCA of Expression Values from Four Cells") +</span></span>
<span id="cb56-486"><a href="#cb56-486" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(plot.title = element_text(hjust = 0.5))</span></span>
<span id="cb56-487"><a href="#cb56-487" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb56-488"><a href="#cb56-488" aria-hidden="true" tabindex="-1"></a>:::</span></code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> A portion of these materials and hands-on activities were adapted from the <a href="https://satijalab.org/">Satija Lab’s</a> <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html">Seurat - Guided Clustering Tutorial</a></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>