<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Will Gammerdinger">
<meta name="author" content="Noor Sohail">
<meta name="dcterms.date" content="2025-08-25">

<title>Subsetting a Slide – Introduction to single-cell RNA-seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-67e3698c9357a87fd1a0e9aab97ce577.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="../site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<link href="../site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="../site_libs/datatables-binding-0.34.0/datatables.js"></script>
<script src="../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="../site_libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="../site_libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script>
<link href="../site_libs/crosstalk-1.2.2/css/crosstalk.min.css" rel="stylesheet">
<script src="../site_libs/crosstalk-1.2.2/js/crosstalk.min.js"></script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Introduction to single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../schedule/schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/"> 
<span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/Intro-to-spatial-scRNA-seq"> 
<span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu"> 
<span class="menu-text">Contact us</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#tasks-left-on-lesson" id="toc-tasks-left-on-lesson" class="nav-link active" data-scroll-target="#tasks-left-on-lesson">Tasks left on lesson</a></li>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#set-up" id="toc-set-up" class="nav-link" data-scroll-target="#set-up">Set-up</a></li>
  <li><a href="#loading-the-seurat-object" id="toc-loading-the-seurat-object" class="nav-link" data-scroll-target="#loading-the-seurat-object">Loading the Seurat Object</a></li>
  <li><a href="#the-two-coordinate-spaces" id="toc-the-two-coordinate-spaces" class="nav-link" data-scroll-target="#the-two-coordinate-spaces">The Two Coordinate Spaces</a>
  <ul class="collapse">
  <li><a href="#pixel-coordinate-space" id="toc-pixel-coordinate-space" class="nav-link" data-scroll-target="#pixel-coordinate-space">Pixel Coordinate Space</a></li>
  <li><a href="#array-coordinate-space" id="toc-array-coordinate-space" class="nav-link" data-scroll-target="#array-coordinate-space">Array Coordinate Space</a></li>
  <li><a href="#drawing-connections-between-the-two-coordinate-spaces" id="toc-drawing-connections-between-the-two-coordinate-spaces" class="nav-link" data-scroll-target="#drawing-connections-between-the-two-coordinate-spaces">Drawing Connections Between the Two Coordinate Spaces</a></li>
  </ul></li>
  <li><a href="#scope-of-data" id="toc-scope-of-data" class="nav-link" data-scroll-target="#scope-of-data">Scope of data</a></li>
  <li><a href="#subsetting-the-seurat-object" id="toc-subsetting-the-seurat-object" class="nav-link" data-scroll-target="#subsetting-the-seurat-object">Subsetting the Seurat Object</a></li>
  <li><a href="#writing-subsetted-data-directory" id="toc-writing-subsetted-data-directory" class="nav-link" data-scroll-target="#writing-subsetted-data-directory">Writing Subsetted Data Directory</a></li>
  <li><a href="#automate-subsetting-by-looping-through-samples-and-bins" id="toc-automate-subsetting-by-looping-through-samples-and-bins" class="nav-link" data-scroll-target="#automate-subsetting-by-looping-through-samples-and-bins">Automate subsetting by looping through samples and bins</a>
  <ul class="collapse">
  <li><a href="#setting-variables" id="toc-setting-variables" class="nav-link" data-scroll-target="#setting-variables">Setting variables</a></li>
  <li><a href="#test-reading-in-fresh_frozen-bin-sizes" id="toc-test-reading-in-fresh_frozen-bin-sizes" class="nav-link" data-scroll-target="#test-reading-in-fresh_frozen-bin-sizes">Test Reading in <code>fresh_frozen</code> Bin Sizes</a></li>
  <li><a href="#test-reading-in-fixed_frozen-bin-sizes" id="toc-test-reading-in-fixed_frozen-bin-sizes" class="nav-link" data-scroll-target="#test-reading-in-fixed_frozen-bin-sizes">Test Reading in <code>fixed_frozen</code> Bin Sizes</a></li>
  </ul></li>
  <li><a href="#merging-a-bin-size" id="toc-merging-a-bin-size" class="nav-link" data-scroll-target="#merging-a-bin-size">Merging a Bin Size</a>
  <ul class="collapse">
  <li><a href="#merging-the-8μm-bin-size" id="toc-merging-the-8μm-bin-size" class="nav-link" data-scroll-target="#merging-the-8μm-bin-size">Merging the 8μm Bin Size</a></li>
  <li><a href="#merging-the-16μm-bin-size" id="toc-merging-the-16μm-bin-size" class="nav-link" data-scroll-target="#merging-the-16μm-bin-size">Merging the 16μm Bin Size</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Subsetting a Slide</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Will Gammerdinger </p>
             <p>Noor Sohail </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 25, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="tasks-left-on-lesson" class="level1">
<h1>Tasks left on lesson</h1>
<ul>
<li>Improve exercise</li>
<li>Run lesson on O2 in order to get directory tree and images following the loop then put those parts in the download.sh and embed them in the lesson</li>
<li>Polish</li>
</ul>
<p>Approximate time: 45 minutes</p>
</section>
<section id="learning-objectives" class="level1">
<h1>Learning Objectives</h1>
<ul>
<li>Describe the coordinate systems for Visium HD data</li>
<li>Subset a spatial scRNA-seq slide</li>
</ul>
</section>
<section id="background" class="level1">
<h1>Background</h1>
<p>When organizing this workshop, spatial single-cell data can be quite large and the result is that analyzing it can be quite memory intensive. In order to focus on the techniques, methods and interpretation of results, we needed to reduce the size of our dataset to work better on our participants computers. We provide the downstream output from this subsetting for our analysis, but we have created these materials should you want to know how to do this yourself.</p>
</section>
<section id="set-up" class="level1">
<h1>Set-up</h1>
<p>Open a new R Script by going <kbd>File</kbd> → <kbd>New File </kbd> → <kbd>R Script</kbd>. Save the file as “subsetting_slide.R”. Provide the R Script with some information about the script:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subsetting_slide.R is for subsetting the spatial single-cell data utilized in the Introduction to Spatial Single-Cell RNA-seq workshop provided by the Harvard Chan Bioinformatics Core</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Written by: &lt;Your name&gt;</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Written on: &lt;Date&gt;</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#  Rscript subsetting_slide.R</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In order to subset the spatial single-cell data, we will need to load three packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DropletUtils)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(png)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magick)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(arrow)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="loading-the-seurat-object" class="level1">
<h1>Loading the Seurat Object</h1>
<p>We can start the subsetting process by loading in our entire dataset into a Seurat object called <code>seurat_obj</code> using the <code>Load10X_Spatial()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load spatial data into Seurat object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.dir =</span> <span class="st">"data/fresh_frozen"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">slice =</span> <span class="st">"fresh_frozen"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The parameters we are using are:</p>
<ul>
<li><code>data.dir</code> - The directory where the H5 file is located should contain a subdirectory called “spatial” that holds the image data</li>
<li><code>bin.size</code> - The size in µm that we want to use for the binning</li>
<li><code>slice</code> - The name of the image</li>
</ul>
<p>We can inspect this Seurat object with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the Seurat object</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>seurat_obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
19059 features across 113885 samples within 1 assay 
Active assay: Spatial.016um (19059 features, 0 variable features)
 1 layer present: counts
 1 spatial field of view present: fresh_frozen.016um</code></pre>
</div>
</div>
<p>We can see that this object has <strong>19059</strong> genes <strong>113885</strong> cells.</p>
</section>
<section id="the-two-coordinate-spaces" class="level1">
<h1>The Two Coordinate Spaces</h1>
<section id="pixel-coordinate-space" class="level2">
<h2 class="anchored" data-anchor-id="pixel-coordinate-space">Pixel Coordinate Space</h2>
<p>Within this Seurat object, we also have coordinate information stored that we will need to use to subset our slide. In order to extract this coordinate information from the Seurat object, we will use the <code>GetTissueCoordinates()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain coordinates for our spatial data</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(seurat_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s get an intuition for what this data looks like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the coordinates table</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(coordinates)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have subsetted the first few dozen lines in the table below:</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-46f9de3f2c2201c5c804" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-46f9de3f2c2201c5c804">{"x":{"filter":"none","vertical":false,"data":[["s_016um_00107_00066-1","s_016um_00126_00213-1","s_016um_00158_00414-1","s_016um_00242_00130-1","s_016um_00230_00364-1","s_016um_00329_00125-1","s_016um_00280_00313-1","s_016um_00184_00346-1","s_016um_00106_00359-1","s_016um_00343_00189-1","s_016um_00258_00092-1","s_016um_00245_00212-1","s_016um_00193_00225-1","s_016um_00283_00339-1","s_016um_00160_00289-1","s_016um_00163_00111-1","s_016um_00134_00038-1","s_016um_00148_00186-1","s_016um_00310_00015-1","s_016um_00191_00066-1","s_016um_00244_00289-1","s_016um_00139_00291-1","s_016um_00219_00284-1","s_016um_00234_00009-1","s_016um_00221_00112-1","s_016um_00266_00094-1","s_016um_00212_00262-1","s_016um_00305_00319-1","s_016um_00261_00327-1","s_016um_00258_00140-1","s_016um_00196_00292-1","s_016um_00198_00377-1","s_016um_00310_00270-1","s_016um_00166_00070-1","s_016um_00191_00292-1","s_016um_00325_00322-1","s_016um_00215_00294-1","s_016um_00137_00270-1","s_016um_00204_00141-1","s_016um_00314_00096-1","s_016um_00175_00015-1","s_016um_00241_00338-1","s_016um_00394_00293-1","s_016um_00280_00247-1","s_016um_00106_00266-1","s_016um_00210_00078-1","s_016um_00180_00237-1","s_016um_00327_00204-1","s_016um_00231_00099-1","s_016um_00118_00045-1","s_016um_00216_00052-1","s_016um_00318_00271-1","s_016um_00249_00398-1","s_016um_00336_00082-1","s_016um_00315_00187-1","s_016um_00338_00245-1","s_016um_00243_00051-1","s_016um_00350_00088-1","s_016um_00132_00082-1","s_016um_00273_00061-1"],[2977.618751211005,4082.946402205297,5946.160099658176,10864.77054951166,10155.54100707687,15949.1768840035,13079.26590342947,7467.917661601083,2909.158827409865,16765.16562738197,11801.09304799674,11037.30344481484,7997.997729764038,13253.70367459832,6067.30407377724,6248.694434273881,4556.440975718804,5369.541594953131,14842.55186003653,7886.533880228576,10976.24781944113,4839.99933752064,9515.423070177978,10401.36705409387,9638.152563672074,12268.54082353153,9107.093097707315,14540.05776142034,11968.4341918231,11799.46342942413,8171.03534959221,8285.02080824511,14833.91620971327,6425.411171255212,7878.836272941563,15708.75230773497,9281.323693401211,4723.836784146922,8643.693736513907,15073.56654340893,6953.241071349535,10799.26350472399,19742.07578381446,13081.50427211291,2912.339393928342,8996.474932601543,7237.872574789907,15829.62476166239,10222.98954925644,3621.170829993594,9347.996163750133,15301.40014693381,11264.74450365384,16359.70760226578,15128.92498527149,16471.07383601817,10925.89409396605,17177.65643918249,4438.059231250685,12678.73656586954],[23843.35694652802,15247.3001571513,3493.118364757471,20095.01817935324,6413.371730167526,20383.35720261078,9393.100406990923,7467.974883990907,6711.442339098875,16640.59166503679,22316.1586035951,15300.28336690137,14542.56020565318,7872.711269695884,10801.93858041793,21209.60300716088,25479.28000585435,16824.99733676617,26815.97139470767,23839.48050109053,10798.06643630626,10685.96441254582,11091.57422148393,27170.30055855369,21148.45651534921,22198.84861120383,12378.25956153127,9041.12122632663,8575.379729906108,19509.57437415773,10624.86581765512,5654.718701184776,11905.97358064532,23606.75273025324,10625.09629899603,8864.785947658416,10507.04779957433,11913.949106783,19453.59485851639,22079.6929642332,26822.20297204795,7933.118096431546,10557.26771206134,13252.18885053705,12149.26239905168,23136.95922664372,13841.5088637766,15764.26829268546,21908.11133512737,25070.72642948659,24656.91138510654,11847.13377327466,4424.469464147805,22897.2634354984,16758.82447631676,13366.45649495993,24714.13559572657,22545.79512105253,22906.67690461147,24128.04800114817],["s_016um_00107_00066-1","s_016um_00126_00213-1","s_016um_00158_00414-1","s_016um_00242_00130-1","s_016um_00230_00364-1","s_016um_00329_00125-1","s_016um_00280_00313-1","s_016um_00184_00346-1","s_016um_00106_00359-1","s_016um_00343_00189-1","s_016um_00258_00092-1","s_016um_00245_00212-1","s_016um_00193_00225-1","s_016um_00283_00339-1","s_016um_00160_00289-1","s_016um_00163_00111-1","s_016um_00134_00038-1","s_016um_00148_00186-1","s_016um_00310_00015-1","s_016um_00191_00066-1","s_016um_00244_00289-1","s_016um_00139_00291-1","s_016um_00219_00284-1","s_016um_00234_00009-1","s_016um_00221_00112-1","s_016um_00266_00094-1","s_016um_00212_00262-1","s_016um_00305_00319-1","s_016um_00261_00327-1","s_016um_00258_00140-1","s_016um_00196_00292-1","s_016um_00198_00377-1","s_016um_00310_00270-1","s_016um_00166_00070-1","s_016um_00191_00292-1","s_016um_00325_00322-1","s_016um_00215_00294-1","s_016um_00137_00270-1","s_016um_00204_00141-1","s_016um_00314_00096-1","s_016um_00175_00015-1","s_016um_00241_00338-1","s_016um_00394_00293-1","s_016um_00280_00247-1","s_016um_00106_00266-1","s_016um_00210_00078-1","s_016um_00180_00237-1","s_016um_00327_00204-1","s_016um_00231_00099-1","s_016um_00118_00045-1","s_016um_00216_00052-1","s_016um_00318_00271-1","s_016um_00249_00398-1","s_016um_00336_00082-1","s_016um_00315_00187-1","s_016um_00338_00245-1","s_016um_00243_00051-1","s_016um_00350_00088-1","s_016um_00132_00082-1","s_016um_00273_00061-1"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>x<\/th>\n      <th>y<\/th>\n      <th>cell<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[1,2]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"x","targets":1},{"name":"y","targets":2},{"name":"cell","targets":3}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>The <code>x</code> and <code>y</code> values correspond to the spot in pixel space that Seurat will be using.</p>
<p>We can compare the original image to the boundaries of the slide area to the coordinates obtained from the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create plot with coordinates</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>coordinate_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(coordinates) <span class="sc">+</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y), <span class="at">size =</span> <span class="fl">0.0001</span>, <span class="at">color =</span> <span class="st">"cornflowerblue"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in and orient the low resolution image</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>lowres_image <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="st">"data/fresh_frozen/spatial/tissue_lowres_image.png"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_rotate</span>(<span class="at">degrees =</span> <span class="dv">270</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.raster</span>(<span class="at">interpolate =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rasterGrob</span>()</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in and orient the detected tissue image</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>detected_image <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="st">"data/fresh_frozen/spatial/detected_tissue_image.jpg"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_rotate</span>(<span class="at">degrees =</span> <span class="dv">90</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_flop</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.raster</span>(<span class="at">interpolate =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rasterGrob</span>()</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot three image images next to each other</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(lowres_image, detected_image, coordinate_plot, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">align =</span> <span class="st">'h'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_subsetting_slide_files/figure-html/campare_images_coordinates-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="array-coordinate-space" class="level2">
<h2 class="anchored" data-anchor-id="array-coordinate-space">Array Coordinate Space</h2>
<p>Let’s dig a bit deeper into how the spatial data is stored. The storage of the data comes from a <code>parquet</code> file that is provided with the data. We can read this parquet file using <code>read_parquet</code> from the <code>arrow</code> package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read parquet file</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>parquet_df <span class="ot">&lt;-</span> <span class="fu">read_parquet</span>(<span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_positions.parquet"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can inspect what this <code>parquet</code> file looks like with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect parquet file</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(parquet_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The first few dozen lines will look like:</p>
<div class="cell">
<div class="cell-output-display">
<div class="datatables html-widget html-fill-item" id="htmlwidget-856506e5b0f73b8e62b6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-856506e5b0f73b8e62b6">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60"],["s_016um_00000_00000-1","s_016um_00000_00001-1","s_016um_00000_00002-1","s_016um_00000_00003-1","s_016um_00000_00004-1","s_016um_00000_00005-1","s_016um_00000_00006-1","s_016um_00000_00007-1","s_016um_00000_00008-1","s_016um_00000_00009-1","s_016um_00000_00010-1","s_016um_00000_00011-1","s_016um_00000_00012-1","s_016um_00000_00013-1","s_016um_00000_00014-1","s_016um_00000_00015-1","s_016um_00000_00016-1","s_016um_00000_00017-1","s_016um_00000_00018-1","s_016um_00000_00019-1","s_016um_00000_00020-1","s_016um_00000_00021-1","s_016um_00000_00022-1","s_016um_00000_00023-1","s_016um_00000_00024-1","s_016um_00000_00025-1","s_016um_00000_00026-1","s_016um_00000_00027-1","s_016um_00000_00028-1","s_016um_00000_00029-1","s_016um_00000_00030-1","s_016um_00000_00031-1","s_016um_00000_00032-1","s_016um_00000_00033-1","s_016um_00000_00034-1","s_016um_00000_00035-1","s_016um_00000_00036-1","s_016um_00000_00037-1","s_016um_00000_00038-1","s_016um_00000_00039-1","s_016um_00000_00040-1","s_016um_00000_00041-1","s_016um_00000_00042-1","s_016um_00000_00043-1","s_016um_00000_00044-1","s_016um_00000_00045-1","s_016um_00000_00046-1","s_016um_00000_00047-1","s_016um_00000_00048-1","s_016um_00000_00049-1","s_016um_00000_00050-1","s_016um_00000_00051-1","s_016um_00000_00052-1","s_016um_00000_00053-1","s_016um_00000_00054-1","s_016um_00000_00055-1","s_016um_00000_00056-1","s_016um_00000_00057-1","s_016um_00000_00058-1","s_016um_00000_00059-1"],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59],[-3273.14021091571,-3273.174583515683,-3273.208956117584,-3273.24332872141,-3273.277701327161,-3273.312073934838,-3273.346446544441,-3273.380819155971,-3273.415191769426,-3273.449564384808,-3273.483937002115,-3273.518309621348,-3273.552682242506,-3273.587054865591,-3273.621427490602,-3273.655800117539,-3273.690172746402,-3273.72454537719,-3273.758918009904,-3273.793290644545,-3273.827663281111,-3273.862035919604,-3273.896408560022,-3273.930781202366,-3273.965153846637,-3273.999526492832,-3274.033899140954,-3274.068271791002,-3274.102644442976,-3274.137017096876,-3274.171389752702,-3274.205762410453,-3274.24013507013,-3274.274507731734,-3274.308880395263,-3274.343253060718,-3274.3776257281,-3274.411998397407,-3274.44637106864,-3274.480743741799,-3274.515116416884,-3274.549489093895,-3274.583861772831,-3274.618234453695,-3274.652607136483,-3274.686979821198,-3274.721352507839,-3274.755725196404,-3274.790097886897,-3274.824470579316,-3274.85884327366,-3274.89321596993,-3274.927588668126,-3274.961961368248,-3274.996334070296,-3275.03070677427,-3275.06507948017,-3275.099452187996,-3275.133824897747,-3275.168197609425],[27707.33408063396,27648.86389502864,27590.39370614718,27531.9235139896,27473.45331855587,27414.98311984602,27356.51291786003,27298.04271259792,27239.57250405967,27181.10229224528,27122.63207715476,27064.16185878811,27005.69163714532,26947.2214122264,26888.75118403135,26830.28095256016,26771.81071781283,26713.34047978937,26654.87023848978,26596.39999391404,26537.92974606217,26479.45949493417,26420.98924053003,26362.51898284975,26304.04872189333,26245.57845766078,26187.10819015209,26128.63791936725,26070.16764530629,26011.69736796918,25953.22708735593,25894.75680346655,25836.28651630102,25777.81622585935,25719.34593214155,25660.8756351476,25602.40533487751,25543.93503133128,25485.46472450892,25426.99441441041,25368.52410103575,25310.05378438496,25251.58346445802,25193.11314125494,25134.64281477572,25076.17248502035,25017.70215198884,24959.23181568119,24900.7614760974,24842.29113323746,24783.82078710137,24725.35043768914,24666.88008500076,24608.40972903624,24549.93936979557,24491.46900727876,24432.9986414858,24374.5282724167,24316.05790007144,24257.58752445004]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>barcode<\/th>\n      <th>in_tissue<\/th>\n      <th>array_row<\/th>\n      <th>array_col<\/th>\n      <th>pxl_row_in_fullres<\/th>\n      <th>pxl_col_in_fullres<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"barcode","targets":1},{"name":"in_tissue","targets":2},{"name":"array_row","targets":3},{"name":"array_col","targets":4},{"name":"pxl_row_in_fullres","targets":5},{"name":"pxl_col_in_fullres","targets":6}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<p>In order to get a sense for this table, let’s investigate each column:</p>
<ul>
<li><code>barcode</code> - This is the barcode for the square on the slide. The format is:
<ul>
<li>[ident]_[bin_size]_[array_row]_[array_col]-[suffix] where:
<ul>
<li><code>ident</code> - The ident</li>
<li><code>bin_size</code> - The size of the bin</li>
<li><code>array_row</code> - The row number in the array for a given bin size</li>
<li><code>array_col</code> - The column number in the array for a given bin size</li>
<li><code>suffix</code> - An optional suffix reserved for technical reasons (usually is <code>-1</code>)</li>
</ul></li>
</ul></li>
<li><code>in_tissue</code> - A binary descriptor for whether Space Ranger believes this is in the tissue or not
<ul>
<li><code>0</code> - Not in the tissue</li>
<li><code>1</code> - In the tissue</li>
</ul></li>
<li><code>array_row</code> - The row number in the array for a given bin size</li>
<li><code>array_col</code> - The column number in the array for a given bin size</li>
<li><code>pxl_row_in_fullres</code> - Location of row in pixel space</li>
<li><code>pxl_col_in_fullres</code> - Location of column in pixel space</li>
</ul>
</section>
<section id="drawing-connections-between-the-two-coordinate-spaces" class="level2">
<h2 class="anchored" data-anchor-id="drawing-connections-between-the-two-coordinate-spaces">Drawing Connections Between the Two Coordinate Spaces</h2>
<p>The <code>array_row</code> and <code>array_col</code> are directly tied to <code>pxl_row_in_fullres</code> and <code>pxl_col_in_fullres</code>, respectively. However the data has undergone an affine transformation within Space Ranger that considers a scale factor (how many pixels per bin), an offset and a rotation. In short, the <code>array_row</code> and <code>array_col</code> are ralated to <code>pxl_row_in_fullres</code> and <code>pxl_col_in_fullres</code> and the <code>parquet</code> file ties them together. Let’s evaluate this by looking at the first line from the <code>parquet</code> file:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">pillar.sigfig =</span> <span class="dv">7</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>parquet_df[parquet_df<span class="sc">$</span>barcode <span class="sc">==</span> <span class="st">"s_016um_00107_00066-1"</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  barcode    in_tissue array_row array_col pxl_row_in_fullres pxl_col_in_fullres
  &lt;chr&gt;          &lt;int&gt;     &lt;int&gt;     &lt;int&gt;              &lt;dbl&gt;              &lt;dbl&gt;
1 s_016um_0…         1       107        66           2977.619           23843.36</code></pre>
</div>
</div>
<p>Now let’s look at that same cell in our <code>coordinates</code> object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>coordinates[coordinates<span class="sc">$</span>cell <span class="sc">==</span> <span class="st">"s_016um_00107_00066-1"</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                             x        y                  cell
s_016um_00107_00066-1 2977.619 23843.36 s_016um_00107_00066-1</code></pre>
</div>
</div>
<p>We can observe that <code>pxl_row_in_fullres</code> in the <code>parquet</code> file is the <code>x</code> column from Seurat’s <code>GetTissueCoordinates()</code> function and <code>pxl_col_in_fullres</code> in the <code>parquet</code> file is the <code>y</code> column from Seurat’s <code>GetTissueCoordinates()</code> function.</p>
</section>
</section>
<section id="scope-of-data" class="level1">
<h1>Scope of data</h1>
<p>We can inspect the number of bins that Space Ranger determine were in a tissue by inspecting the Seurat object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(seurat_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 113885</code></pre>
</div>
</div>
<p>This data is also encoded in the <code>parquet</code> file within the <code>in_tissue</code> column where <code>0</code> represent not in the tissue and <code>1</code> represents in the tissue.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(parquet_df<span class="sc">$</span>in_tissue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     0      1 
 61676 113885 </code></pre>
</div>
</div>
<p>From this we can see that Seurat has subsampled our dataset to only include the bins that Space Ranger detected as being within the tissue.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If we wanted to include all pixels (with and without tissue), we could have used the <code>filter.matrix = FALSE</code> option in <code>Load10X_Spatial()</code>)</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Run the <code>Load10X_Spatial()</code> function again and also add the <code>filter.matrix = FALSE</code> option and assign it to <code>exercise_obj</code>. Pull the coordinates with the <code>GetTissueCoordinates()</code> function and assigning them to <code>exercise_coordinates</code>.</p>
<ol type="1">
<li>Determine the <code>exercise_coordinates</code> object to determine the number of rows of and the number of.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load spatial data into Seurat object</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>exercise_obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.dir =</span> <span class="st">"data/fresh_frozen"</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">slice =</span> <span class="st">"fresh_frozen"</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">filter.matrix =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract tissue coordinates from the Seurat object</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>exercise_coordinates <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(exercise_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li><p>In a single 6.5mm x 6.5mm capture area from Visium HD, we would expect that to correspond to about 20,000 pixels by 20,000 pixels. Do our expectations match our observed data?</p></li>
<li><p>By comparing the dimensions of <code>coordinates</code>, which only includes coordinates for locations that Space Ranger detected tissue, and <code>exercise_coordinates</code>, which contains all coordinates with and without tissue detected by Space Ranger, how many pixels didn’t have tissue detected in them?</p></li>
</ol>
<p>After completing this exercise, be sure to remove <code>exercise_obj</code> and <code>exercise_coordinates</code> from our R environment to save on memory by using:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove exercise Seurat object</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(exercise_obj)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove exercise coordinate object</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(exercise_coordinates)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="subsetting-the-seurat-object" class="level1">
<h1>Subsetting the Seurat Object</h1>
<p>Now that we have an understanding of the coordinates, let’s go ahead and subset our coordinates based upon the region we will like to focus on in this workshop:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set x and y bounds for subsetting image</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>x_min <span class="ot">&lt;-</span> <span class="sc">-</span><span class="cn">Inf</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>x_max <span class="ot">&lt;-</span> <span class="dv">15114</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>y_min <span class="ot">&lt;-</span> <span class="sc">-</span><span class="cn">Inf</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>y_max <span class="ot">&lt;-</span> <span class="dv">21517</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract bin barcodes based on the defined x and y bounds</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>subsetted_bins <span class="ot">&lt;-</span> coordinates <span class="sc">%&gt;%</span> </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">subset</span>((x <span class="sc">&gt;</span> x_min <span class="sc">&amp;</span> x <span class="sc">&lt;</span> x_max) <span class="sc">&amp;</span> </span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>           (y <span class="sc">&gt;</span> y_min <span class="sc">&amp;</span> y <span class="sc">&lt;</span> y_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">row.names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The output of this subsetting should be bin names that fit this critera for <code>x</code> and <code>y</code> bounds. We can inspect the object with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect subsetted bin names</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(subsetted_bins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "s_016um_00126_00213-1" "s_016um_00158_00414-1" "s_016um_00242_00130-1"
[4] "s_016um_00230_00364-1" "s_016um_00280_00313-1" "s_016um_00184_00346-1"</code></pre>
</div>
</div>
<p>We can confirm that this subsetting was done appropriately, by subsetting the <code>coordinates</code> object by our <code>subsetted_bins</code> names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>coordinates[subsetted_bins,] <span class="sc">%&gt;%</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                              x         y                  cell
s_016um_00126_00213-1  4082.946 15247.300 s_016um_00126_00213-1
s_016um_00158_00414-1  5946.160  3493.118 s_016um_00158_00414-1
s_016um_00242_00130-1 10864.771 20095.018 s_016um_00242_00130-1
s_016um_00230_00364-1 10155.541  6413.372 s_016um_00230_00364-1
s_016um_00280_00313-1 13079.266  9393.100 s_016um_00280_00313-1
s_016um_00184_00346-1  7467.918  7467.975 s_016um_00184_00346-1</code></pre>
</div>
</div>
<p>From here we can see our pixels are all within the pixel range that we subsetted by, so we feel confident that our subsetting was appropriate.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Sanity checks like the above check may feel unnecessary but they are really important for good data analysis practices. It oftentimes takes less than a few moments to inspect and can save you lots of time downstream when you are debugging code.</p>
</div>
</div>
<p>Let’s go ahead and start subsetting our Seurat object. This will happen in three parts:</p>
<ol type="1">
<li>Creating a new column, <code>filter</code>, in <code>meta.data</code> of the Seurat object</li>
<li>Assigning the values of <code>TRUE</code> or <code>FALSE</code> for bins to be removed or kept, respectively</li>
<li>Subsetting the Seurat object based about the data in the <code>filter</code> column of <code>meta.data</code></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Label a column called filter for TRUE</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>seurat_obj<span class="sc">$</span>filter <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-assign the value of FALSE to the filter column if the bin is within subsetted_bins</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>seurat_obj<span class="sc">@</span>meta.data[subsetted_bins, <span class="st">"filter"</span>] <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Retain the bins which have FALSE for filter in the meta data</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>seurat_obj_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_obj, filter <span class="sc">==</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Just like we did previously when subsetting coordinates, let’s confirm that we are subsetting the correct bins.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the first few lines of coordinates to ensure they match our expectation</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">GetTissueCoordinates</span>(seurat_obj_filt) <span class="sc">%&gt;%</span>  </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                              x         y                  cell
s_016um_00126_00213-1  4082.946 15247.300 s_016um_00126_00213-1
s_016um_00158_00414-1  5946.160  3493.118 s_016um_00158_00414-1
s_016um_00242_00130-1 10864.771 20095.018 s_016um_00242_00130-1
s_016um_00230_00364-1 10155.541  6413.372 s_016um_00230_00364-1
s_016um_00280_00313-1 13079.266  9393.100 s_016um_00280_00313-1
s_016um_00184_00346-1  7467.918  7467.975 s_016um_00184_00346-1</code></pre>
</div>
</div>
<p>Last, we can use the <code>SpatialFeaturePlot()</code> function from Seurat to compare the before and after of subsetting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visual the original image and the subsetted image next to each other</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(seurat_obj, <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>       <span class="fu">SpatialFeaturePlot</span>(seurat_obj_filt, <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="Aside_subsetting_slide_files/figure-html/visualize_subsetted_bins-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="writing-subsetted-data-directory" class="level1">
<h1>Writing Subsetted Data Directory</h1>
<p>It looks like we have correctly subsetted out data and now we are ready to create a directory to hold our data. This will make it easier in the future to work with our data, because we can use the <code>Load10X_Spatial()</code> function from Seurat to read in our subsetted data rather than performing the subsetting again.</p>
<p>We are going to start by creating a directory structure to hold our subsetted data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/"</span>))</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/"</span>))</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/"</span>))</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The directory structure should look like:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>data_filtered/
└── fresh_frozen
    └── binned_outputs
        └── square_016um
            ├── filtered_feature_bc_matrix.h5
            └── spatial
                ├── scalefactors_json.json
                ├── tissue_lowres_image.png
                └── tissue_positions.parquet</code></pre>
</div>
</div>
<p>Now we will populate this directory with the files needed to be able to run the <code>Load10X_Spatial()</code> function in the future.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy scale factors JSON to new directory</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/scalefactors_json.json"</span>,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy tissue positions parquet file to new directory</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_positions.parquet"</span>,</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy low resolution PNG to new directory</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_lowres_image.png"</span>,</span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Write counts to hd5 file</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a><span class="fu">write10xCounts</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/filtered_feature_bc_matrix.h5"</span>,</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">LayerData</span>(seurat_obj_filt),</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">barcodes=</span><span class="fu">colnames</span>(seurat_obj_filt),</span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.id=</span><span class="fu">rownames</span>(seurat_obj_filt),</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">version=</span><span class="st">"3"</span>,</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="st">"HDF5"</span>,</span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After writing and copying the files the file directory should look like this:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>data_filtered/
└── fresh_frozen
    └── binned_outputs
        └── square_016um
            ├── filtered_feature_bc_matrix.h5
            └── spatial
                ├── scalefactors_json.json
                ├── tissue_lowres_image.png
                └── tissue_positions.parquet</code></pre>
</div>
</div>
</section>
<section id="automate-subsetting-by-looping-through-samples-and-bins" class="level1">
<h1>Automate subsetting by looping through samples and bins</h1>
<p>Now we have processed a single bin (16μm) for a single sample (fresh_frozen). However, we would like to process our data for 8μm and 16μm for both “fresh_frozen” and “fixed_frozen” This process can be automated by using the same process as above, but looping through both bin sizes and both samples. <strong>Due to memory constraints on some laptops, you many not be able to run the code for the remaining parts of the lesson</strong>. It requires ~32Gb of RAM to execute.</p>
<section id="setting-variables" class="level2">
<h2 class="anchored" data-anchor-id="setting-variables">Setting variables</h2>
<p>We are going to set-up the the variables that we would like to loop through (bin size and sample) and also set our boundaries for where to subset our sample for. The index of the <code>folders</code> vector is the sample index for the boundary vectors (<code>x_min</code>, <code>x_max</code>, <code>y_min</code>, <code>y_max</code>).For example, the <code>xmin</code> of <code>-Inf</code> corresponds to the fresh_frozen sample while the <code>xmin</code> of <code>7205.1</code> corresponds to the fixed_frozen sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"008"</span>, <span class="st">"016"</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>folders <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fresh_frozen"</span>, <span class="st">"fixed_frozen"</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>x_min <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">7205.1</span>) </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>x_max <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15114</span>, <span class="cn">Inf</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>y_min <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="sc">-</span><span class="dv">1592</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>y_max <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21517</span>, <span class="dv">17735</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The code below will allow us to loop through the different samples and bin sizes, adding them to a <code>data_filtered/</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (folder <span class="cf">in</span> folders) {</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>, folder, <span class="st">"/binned_outputs/"</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (bin <span class="cf">in</span> bins) {</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    path_bin <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, <span class="st">"um"</span>)</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(path_bin, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Copy spatial images from original folder</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(<span class="fu">paste0</span>(path_bin, <span class="st">"/spatial"</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>    path_json <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/scalefactors_json.json"</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>    path_parq <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/tissue_positions.parquet"</span>)</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>    path_png  <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/tissue_lowres_image.png"</span>)</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>    files_spatial <span class="ot">&lt;-</span> <span class="fu">c</span>(path_json, path_parq, path_png)</span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">file.copy</span>(files_spatial,<span class="fu">paste0</span>(path_bin, <span class="st">"/spatial"</span>))</span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Load object</span></span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>    obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">data.dir =</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder),</span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">bin.size =</span> <span class="fu">c</span>(<span class="fu">as.integer</span>(bin)),</span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a>      <span class="at">slice =</span> folder)</span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Cells to keep based on coordinates</span></span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>    coords <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(obj)</span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>    cells <span class="ot">&lt;-</span> coords <span class="sc">%&gt;%</span> </span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a>      <span class="fu">subset</span>((x <span class="sc">&gt;</span> x_min[<span class="fu">which</span>(folders <span class="sc">==</span> folder)] <span class="sc">&amp;</span> x <span class="sc">&lt;</span> x_max[<span class="fu">which</span>(folders <span class="sc">==</span> folder)]) <span class="sc">&amp;</span> </span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>             (y <span class="sc">&gt;</span> y_min[<span class="fu">which</span>(folders <span class="sc">==</span> folder)] <span class="sc">&amp;</span> y <span class="sc">&lt;</span> y_max[<span class="fu">which</span>(folders <span class="sc">==</span> folder)])) <span class="sc">%&gt;%</span> </span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>      <span class="fu">row.names</span>()</span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-35"><a href="#cb35-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset object</span></span>
<span id="cb35-36"><a href="#cb35-36" aria-hidden="true" tabindex="-1"></a>    obj<span class="sc">$</span>filter <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb35-37"><a href="#cb35-37" aria-hidden="true" tabindex="-1"></a>    obj<span class="sc">@</span>meta.data[cells, <span class="st">"filter"</span>] <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb35-38"><a href="#cb35-38" aria-hidden="true" tabindex="-1"></a>    obj_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, filter <span class="sc">==</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-39"><a href="#cb35-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-40"><a href="#cb35-40" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(obj, <span class="fu">paste0</span>(<span class="st">"nCount_Spatial."</span>, bin, <span class="st">"um"</span>), <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb35-41"><a href="#cb35-41" aria-hidden="true" tabindex="-1"></a>         <span class="fu">SpatialFeaturePlot</span>(obj_filt, <span class="fu">paste0</span>(<span class="st">"nCount_Spatial."</span>, bin, <span class="st">"um"</span>), <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb35-42"><a href="#cb35-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(p)</span>
<span id="cb35-43"><a href="#cb35-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-44"><a href="#cb35-44" aria-hidden="true" tabindex="-1"></a>    path_h5 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(path_bin, <span class="st">"/filtered_feature_bc_matrix.h5"</span>)</span>
<span id="cb35-45"><a href="#cb35-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write10xCounts</span>(path_h5,</span>
<span id="cb35-46"><a href="#cb35-46" aria-hidden="true" tabindex="-1"></a>                   <span class="at">x=</span><span class="fu">LayerData</span>(obj_filt),</span>
<span id="cb35-47"><a href="#cb35-47" aria-hidden="true" tabindex="-1"></a>                   <span class="at">barcodes=</span><span class="fu">colnames</span>(obj_filt),</span>
<span id="cb35-48"><a href="#cb35-48" aria-hidden="true" tabindex="-1"></a>                   <span class="at">gene.id=</span><span class="fu">rownames</span>(obj_filt),</span>
<span id="cb35-49"><a href="#cb35-49" aria-hidden="true" tabindex="-1"></a>                   <span class="at">version=</span><span class="st">"3"</span>,</span>
<span id="cb35-50"><a href="#cb35-50" aria-hidden="true" tabindex="-1"></a>                   <span class="at">type=</span><span class="st">"HDF5"</span>,</span>
<span id="cb35-51"><a href="#cb35-51" aria-hidden="true" tabindex="-1"></a>                   <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-52"><a href="#cb35-52" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-53"><a href="#cb35-53" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(obj)</span>
<span id="cb35-54"><a href="#cb35-54" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(obj_filt)</span>
<span id="cb35-55"><a href="#cb35-55" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb35-56"><a href="#cb35-56" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="test-reading-in-fresh_frozen-bin-sizes" class="level2">
<h2 class="anchored" data-anchor-id="test-reading-in-fresh_frozen-bin-sizes">Test Reading in <code>fresh_frozen</code> Bin Sizes</h2>
<p>If we had run been able to run the above for loops, our directory structure would now look like:</p>
<p>With this directory structure, let’s go ahead and load in the “fresh_frozen” sample for the 16μm bin size:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fresh_frozen 16μm</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>fresh_16 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fresh_frozen"</span>,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fresh"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next we will load in the “fresh_frozen” sample for the 8μm bin size:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fresh_frozen 8μm</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>fresh_8 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fresh_frozen"</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">8</span>,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fresh"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can confirm that the number of bins in the <code>fresh_8</code> is greater than the number of bins in the <code>fresh_16</code> as we would expect:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 8μm fresh_frozen sample</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fresh_8)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 16μm fresh_frozen sample</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fresh_16)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lastly, we can visualize both bin sizes for our “fresh_frozen” samples side-by-side using <code>SpatialFeaturePlot()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare SpatialFeaturePlot of fresh_8 and fresh_16</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(fresh_8,  <span class="st">"nCount_Spatial.008um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">|</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">SpatialFeaturePlot</span>(fresh_16,  <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="test-reading-in-fixed_frozen-bin-sizes" class="level2">
<h2 class="anchored" data-anchor-id="test-reading-in-fixed_frozen-bin-sizes">Test Reading in <code>fixed_frozen</code> Bin Sizes</h2>
<p>We will iterate those steps for the “fixed_frozen” sample:. First, we will read in the 16μm bin size:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fixed_frozen 16μm</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>fixed_16 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fixed_frozen"</span>,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fixed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we will read in the 8μm bin size for the “fixed_frozen” sample:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fixed_frozen 8μm</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>fixed_8 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fixed_frozen"</span>,</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">8</span>,</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fixed"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once again, we will do a check to ensure that the number of bins in the <code>fixed_8</code> is greater than the number of bins in the <code>fixed_16</code> as we would expect:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 8μm fixed_frozen sample</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fixed_8)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 16μm fixed_frozen sample</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fixed_16)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lastly, we can check that both bin sizes have been subsetted correctly and view them with <code>SpatialFeaturePlot()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare SpatialFeaturePlot of fixed_8 and fixed_16</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(fixed_8,  <span class="st">"nCount_Spatial.008um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">|</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">SpatialFeaturePlot</span>(fixed_16,  <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="merging-a-bin-size" class="level1">
<h1>Merging a Bin Size</h1>
<p>With our sample looking appropriate, we can now merge the samples with the same bin size into a single Seurat object. This can be done using Seurat’s <code>merge()</code> function.</p>
<section id="merging-the-8μm-bin-size" class="level2">
<h2 class="anchored" data-anchor-id="merging-the-8μm-bin-size">Merging the 8μm Bin Size</h2>
<p>First, we will merge the 8μm bin size:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge two sample together</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">merge</span>(fixed_8, fresh_8)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, when we use the <code>merge()</code> function in Seurat, it keeps the counts matrices of previously <code>fixed_8</code> and <code>fresh_8</code> in separate layers. We would like them to be in the same layer for our analysis, so we will use the <code>JoinLayers()</code> function from Seurat to do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Combines combines the count matrices from different layers into a single layer</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_8)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We would like to visualize our data to ensure that everything looks like it was done correctly, but in order to do this appropriately, we will need to normalize the data using Seuart’s <code>NormalizeData()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data for the merged 8μm bin size</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(merged_8)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visualize a gene that we are potentially interested in “Cck” using <code>SpatialFeaturePlot()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Cck for our merged 8μm bin size</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(merged_8, </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">features =</span> <span class="st">"Cck"</span>, </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="merging-the-16μm-bin-size" class="level2">
<h2 class="anchored" data-anchor-id="merging-the-16μm-bin-size">Merging the 16μm Bin Size</h2>
<p>Now, we will merge the 16μm bin size following the same steps:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge two sample together</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">merge</span>(fixed_16, fresh_16)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Again, we will use the <code>JoinLayers()</code> to combine the counts matrices of the previously <code>fixed_16</code> and <code>fresh_16</code> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Combines combines the count matrices from different layers into a single layer</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_16)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We will normalize our 16μm data for visualization with Seuart’s <code>NormalizeData()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data for the merged 16μm bin size</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(merged_16)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visualize a gene that we are potentially interested in “Cck” using <code>SpatialFeaturePlot()</code> on the 16μm data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Cck for our merged 16μm bin size</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(merged_16, </span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">features =</span> <span class="st">"Cck"</span>, </span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>At this point we have successfully subsetted out data for two different bin sizes and from two separate samples while performing appropriate checks along the way to ensure that our subsetting has been done correctly.</p>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb52" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Subsetting a Slide"</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> </span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="co">  - Will Gammerdinger</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a><span class="co">  - Noor Sohail</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "Monday, August 25, 2025"</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="an">editor_options:</span><span class="co"> </span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a><span class="co">  markdown: </span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="co">    wrap: 72</span></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: prep_quarto</span></span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fs)</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a><span class="fu"># Tasks left on lesson</span></span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Improve exercise</span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Run lesson on O2 in order to get directory tree and images following the loop then put those parts in the download.sh and embed them in the lesson</span>
<span id="cb52-25"><a href="#cb52-25" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Polish</span>
<span id="cb52-26"><a href="#cb52-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-27"><a href="#cb52-27" aria-hidden="true" tabindex="-1"></a>Approximate time: 45 minutes</span>
<span id="cb52-28"><a href="#cb52-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-29"><a href="#cb52-29" aria-hidden="true" tabindex="-1"></a><span class="fu"># Learning Objectives</span></span>
<span id="cb52-30"><a href="#cb52-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-31"><a href="#cb52-31" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Describe the coordinate systems for Visium HD data</span>
<span id="cb52-32"><a href="#cb52-32" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Subset a spatial scRNA-seq slide</span>
<span id="cb52-33"><a href="#cb52-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-34"><a href="#cb52-34" aria-hidden="true" tabindex="-1"></a><span class="fu"># Background</span></span>
<span id="cb52-35"><a href="#cb52-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-36"><a href="#cb52-36" aria-hidden="true" tabindex="-1"></a>When organizing this workshop, spatial single-cell data can be quite large and the result is that analyzing it can be quite memory intensive. In order to focus on the techniques, methods and interpretation of results, we needed to reduce the size of our dataset to work better on our participants computers. We provide the downstream output from this subsetting for our analysis, but we have created these materials should you want to know how to do this yourself.</span>
<span id="cb52-37"><a href="#cb52-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-38"><a href="#cb52-38" aria-hidden="true" tabindex="-1"></a><span class="fu"># Set-up</span></span>
<span id="cb52-39"><a href="#cb52-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-40"><a href="#cb52-40" aria-hidden="true" tabindex="-1"></a>Open a new R Script by going <span class="dt">&lt;</span><span class="kw">kbd</span><span class="dt">&gt;</span>File<span class="dt">&lt;/</span><span class="kw">kbd</span><span class="dt">&gt;</span> → <span class="dt">&lt;</span><span class="kw">kbd</span><span class="dt">&gt;</span>New File <span class="dt">&lt;/</span><span class="kw">kbd</span><span class="dt">&gt;</span> → <span class="dt">&lt;</span><span class="kw">kbd</span><span class="dt">&gt;</span>R Script<span class="dt">&lt;/</span><span class="kw">kbd</span><span class="dt">&gt;</span>. Save the file as "subsetting_slide.R". Provide the R Script with some information about the script:</span>
<span id="cb52-41"><a href="#cb52-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-44"><a href="#cb52-44" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-45"><a href="#cb52-45" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: metadata</span></span>
<span id="cb52-46"><a href="#cb52-46" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-47"><a href="#cb52-47" aria-hidden="true" tabindex="-1"></a><span class="co"># subsetting_slide.R is for subsetting the spatial single-cell data utilized in the Introduction to Spatial Single-Cell RNA-seq workshop provided by the Harvard Chan Bioinformatics Core</span></span>
<span id="cb52-48"><a href="#cb52-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Written by: &lt;Your name&gt;</span></span>
<span id="cb52-49"><a href="#cb52-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Written on: &lt;Date&gt;</span></span>
<span id="cb52-50"><a href="#cb52-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage example:</span></span>
<span id="cb52-51"><a href="#cb52-51" aria-hidden="true" tabindex="-1"></a><span class="co">#  Rscript subsetting_slide.R</span></span>
<span id="cb52-52"><a href="#cb52-52" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-53"><a href="#cb52-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-54"><a href="#cb52-54" aria-hidden="true" tabindex="-1"></a>In order to subset the spatial single-cell data, we will need to load</span>
<span id="cb52-55"><a href="#cb52-55" aria-hidden="true" tabindex="-1"></a>three packages:</span>
<span id="cb52-56"><a href="#cb52-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-59"><a href="#cb52-59" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-60"><a href="#cb52-60" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_packages</span></span>
<span id="cb52-61"><a href="#cb52-61" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb52-62"><a href="#cb52-62" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb52-63"><a href="#cb52-63" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb52-64"><a href="#cb52-64" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DropletUtils)</span>
<span id="cb52-65"><a href="#cb52-65" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb52-66"><a href="#cb52-66" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(png)</span>
<span id="cb52-67"><a href="#cb52-67" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grid)</span>
<span id="cb52-68"><a href="#cb52-68" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magick)</span>
<span id="cb52-69"><a href="#cb52-69" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(arrow)</span>
<span id="cb52-70"><a href="#cb52-70" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-71"><a href="#cb52-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-72"><a href="#cb52-72" aria-hidden="true" tabindex="-1"></a><span class="fu"># Loading the Seurat Object</span></span>
<span id="cb52-73"><a href="#cb52-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-74"><a href="#cb52-74" aria-hidden="true" tabindex="-1"></a>We can start the subsetting process by loading in our entire dataset</span>
<span id="cb52-75"><a href="#cb52-75" aria-hidden="true" tabindex="-1"></a>into a Seurat object called <span class="in">`seurat_obj`</span> using the <span class="in">`Load10X_Spatial()`</span></span>
<span id="cb52-76"><a href="#cb52-76" aria-hidden="true" tabindex="-1"></a>function:</span>
<span id="cb52-77"><a href="#cb52-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-80"><a href="#cb52-80" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-81"><a href="#cb52-81" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_spatial_object</span></span>
<span id="cb52-82"><a href="#cb52-82" aria-hidden="true" tabindex="-1"></a><span class="co"># Load spatial data into Seurat object</span></span>
<span id="cb52-83"><a href="#cb52-83" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-84"><a href="#cb52-84" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.dir =</span> <span class="st">"data/fresh_frozen"</span>,</span>
<span id="cb52-85"><a href="#cb52-85" aria-hidden="true" tabindex="-1"></a>  <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb52-86"><a href="#cb52-86" aria-hidden="true" tabindex="-1"></a>  <span class="at">slice =</span> <span class="st">"fresh_frozen"</span>)</span>
<span id="cb52-87"><a href="#cb52-87" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-88"><a href="#cb52-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-89"><a href="#cb52-89" aria-hidden="true" tabindex="-1"></a>The parameters we are using are: </span>
<span id="cb52-90"><a href="#cb52-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-91"><a href="#cb52-91" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`data.dir`</span> - The directory where the H5 file is located should contain a subdirectory called "spatial" that holds the image data </span>
<span id="cb52-92"><a href="#cb52-92" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`bin.size`</span> - The size in µm that we want to use for the binning </span>
<span id="cb52-93"><a href="#cb52-93" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`slice`</span> - The name of the image</span>
<span id="cb52-94"><a href="#cb52-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-95"><a href="#cb52-95" aria-hidden="true" tabindex="-1"></a>We can inspect this Seurat object with:</span>
<span id="cb52-96"><a href="#cb52-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-99"><a href="#cb52-99" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-100"><a href="#cb52-100" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: inspect_seurat_object</span></span>
<span id="cb52-101"><a href="#cb52-101" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the Seurat object</span></span>
<span id="cb52-102"><a href="#cb52-102" aria-hidden="true" tabindex="-1"></a>seurat_obj</span>
<span id="cb52-103"><a href="#cb52-103" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-104"><a href="#cb52-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-105"><a href="#cb52-105" aria-hidden="true" tabindex="-1"></a>We can see that this object has **`r format(nrow(seurat_obj), scientific = FALSE)`** genes **`r format(ncol(seurat_obj), scientific = FALSE)`** cells.</span>
<span id="cb52-106"><a href="#cb52-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-107"><a href="#cb52-107" aria-hidden="true" tabindex="-1"></a><span class="fu"># The Two Coordinate Spaces</span></span>
<span id="cb52-108"><a href="#cb52-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-109"><a href="#cb52-109" aria-hidden="true" tabindex="-1"></a><span class="fu">## Pixel Coordinate Space</span></span>
<span id="cb52-110"><a href="#cb52-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-111"><a href="#cb52-111" aria-hidden="true" tabindex="-1"></a>Within this Seurat object, we also have coordinate information stored that we will need to use to subset our slide. In order to extract this coordinate information from the Seurat object, we will use the <span class="in">`GetTissueCoordinates()`</span> function:</span>
<span id="cb52-112"><a href="#cb52-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-115"><a href="#cb52-115" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-116"><a href="#cb52-116" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: get_coordinates</span></span>
<span id="cb52-117"><a href="#cb52-117" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain coordinates for our spatial data</span></span>
<span id="cb52-118"><a href="#cb52-118" aria-hidden="true" tabindex="-1"></a>coordinates <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(seurat_obj)</span>
<span id="cb52-119"><a href="#cb52-119" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-120"><a href="#cb52-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-121"><a href="#cb52-121" aria-hidden="true" tabindex="-1"></a>Let's get an intuition for what this data looks like:</span>
<span id="cb52-122"><a href="#cb52-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-125"><a href="#cb52-125" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-126"><a href="#cb52-126" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: view_coordinates</span></span>
<span id="cb52-127"><a href="#cb52-127" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-128"><a href="#cb52-128" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the coordinates table</span></span>
<span id="cb52-129"><a href="#cb52-129" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(coordinates)</span>
<span id="cb52-130"><a href="#cb52-130" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-131"><a href="#cb52-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-132"><a href="#cb52-132" aria-hidden="true" tabindex="-1"></a>We have subsetted the first few dozen lines in the table below:</span>
<span id="cb52-133"><a href="#cb52-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-136"><a href="#cb52-136" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-137"><a href="#cb52-137" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: table_coordinates</span></span>
<span id="cb52-138"><a href="#cb52-138" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-139"><a href="#cb52-139" aria-hidden="true" tabindex="-1"></a><span class="co"># Load DT to make a table</span></span>
<span id="cb52-140"><a href="#cb52-140" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DT)</span>
<span id="cb52-141"><a href="#cb52-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-142"><a href="#cb52-142" aria-hidden="true" tabindex="-1"></a><span class="co"># View the first 60 lines in a data table</span></span>
<span id="cb52-143"><a href="#cb52-143" aria-hidden="true" tabindex="-1"></a>coordinates <span class="sc">%&gt;%</span> </span>
<span id="cb52-144"><a href="#cb52-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">60</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-145"><a href="#cb52-145" aria-hidden="true" tabindex="-1"></a>  <span class="fu">datatable</span>(<span class="at">options =</span> <span class="fu">list</span>(<span class="at">pageLength =</span> <span class="dv">5</span>))</span>
<span id="cb52-146"><a href="#cb52-146" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-147"><a href="#cb52-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-148"><a href="#cb52-148" aria-hidden="true" tabindex="-1"></a>The <span class="in">`x`</span> and <span class="in">`y`</span> values correspond to the spot in pixel space that Seurat will be using. </span>
<span id="cb52-149"><a href="#cb52-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-150"><a href="#cb52-150" aria-hidden="true" tabindex="-1"></a>We can compare the original image to the boundaries of the slide area to the coordinates obtained from the data. </span>
<span id="cb52-151"><a href="#cb52-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-154"><a href="#cb52-154" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-155"><a href="#cb52-155" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: campare_images_coordinates</span></span>
<span id="cb52-156"><a href="#cb52-156" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig.height: 3</span></span>
<span id="cb52-157"><a href="#cb52-157" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb52-158"><a href="#cb52-158" aria-hidden="true" tabindex="-1"></a><span class="co"># Create plot with coordinates</span></span>
<span id="cb52-159"><a href="#cb52-159" aria-hidden="true" tabindex="-1"></a>coordinate_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(coordinates) <span class="sc">+</span></span>
<span id="cb52-160"><a href="#cb52-160" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y), <span class="at">size =</span> <span class="fl">0.0001</span>, <span class="at">color =</span> <span class="st">"cornflowerblue"</span>)</span>
<span id="cb52-161"><a href="#cb52-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-162"><a href="#cb52-162" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in and orient the low resolution image</span></span>
<span id="cb52-163"><a href="#cb52-163" aria-hidden="true" tabindex="-1"></a>lowres_image <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="st">"data/fresh_frozen/spatial/tissue_lowres_image.png"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-164"><a href="#cb52-164" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_rotate</span>(<span class="at">degrees =</span> <span class="dv">270</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-165"><a href="#cb52-165" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.raster</span>(<span class="at">interpolate =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-166"><a href="#cb52-166" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rasterGrob</span>()</span>
<span id="cb52-167"><a href="#cb52-167" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-168"><a href="#cb52-168" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in and orient the detected tissue image</span></span>
<span id="cb52-169"><a href="#cb52-169" aria-hidden="true" tabindex="-1"></a>detected_image <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="st">"data/fresh_frozen/spatial/detected_tissue_image.jpg"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-170"><a href="#cb52-170" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_rotate</span>(<span class="at">degrees =</span> <span class="dv">90</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-171"><a href="#cb52-171" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_flop</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb52-172"><a href="#cb52-172" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.raster</span>(<span class="at">interpolate =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-173"><a href="#cb52-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rasterGrob</span>()</span>
<span id="cb52-174"><a href="#cb52-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-175"><a href="#cb52-175" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot three image images next to each other</span></span>
<span id="cb52-176"><a href="#cb52-176" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(lowres_image, detected_image, coordinate_plot, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">align =</span> <span class="st">'h'</span>)</span>
<span id="cb52-177"><a href="#cb52-177" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-178"><a href="#cb52-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-179"><a href="#cb52-179" aria-hidden="true" tabindex="-1"></a><span class="fu">## Array Coordinate Space</span></span>
<span id="cb52-180"><a href="#cb52-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-181"><a href="#cb52-181" aria-hidden="true" tabindex="-1"></a>Let's dig a bit deeper into how the spatial data is stored. The storage of the data comes from a <span class="in">`parquet`</span> file that is provided with the data. We can read this parquet file using <span class="in">`read_parquet`</span> from the <span class="in">`arrow`</span> package:</span>
<span id="cb52-182"><a href="#cb52-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-185"><a href="#cb52-185" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-186"><a href="#cb52-186" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: read_parquet_file</span></span>
<span id="cb52-187"><a href="#cb52-187" aria-hidden="true" tabindex="-1"></a><span class="co"># Read parquet file</span></span>
<span id="cb52-188"><a href="#cb52-188" aria-hidden="true" tabindex="-1"></a>parquet_df <span class="ot">&lt;-</span> <span class="fu">read_parquet</span>(<span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_positions.parquet"</span>)</span>
<span id="cb52-189"><a href="#cb52-189" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-190"><a href="#cb52-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-191"><a href="#cb52-191" aria-hidden="true" tabindex="-1"></a>We can inspect what this <span class="in">`parquet`</span> file looks like with:</span>
<span id="cb52-192"><a href="#cb52-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-195"><a href="#cb52-195" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-196"><a href="#cb52-196" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: view_parquet_file</span></span>
<span id="cb52-197"><a href="#cb52-197" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-198"><a href="#cb52-198" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect parquet file</span></span>
<span id="cb52-199"><a href="#cb52-199" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(parquet_df)</span>
<span id="cb52-200"><a href="#cb52-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-201"><a href="#cb52-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-202"><a href="#cb52-202" aria-hidden="true" tabindex="-1"></a>The first few dozen lines will look like:</span>
<span id="cb52-203"><a href="#cb52-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-206"><a href="#cb52-206" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-207"><a href="#cb52-207" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: table_parquet_file</span></span>
<span id="cb52-208"><a href="#cb52-208" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-209"><a href="#cb52-209" aria-hidden="true" tabindex="-1"></a><span class="co"># View the first 60 lines of the parquet file in a data table</span></span>
<span id="cb52-210"><a href="#cb52-210" aria-hidden="true" tabindex="-1"></a>parquet_df <span class="sc">%&gt;%</span> </span>
<span id="cb52-211"><a href="#cb52-211" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="at">n =</span> <span class="dv">60</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb52-212"><a href="#cb52-212" aria-hidden="true" tabindex="-1"></a>  <span class="fu">datatable</span>(<span class="at">options =</span> <span class="fu">list</span>(<span class="at">pageLength =</span> <span class="dv">5</span>))</span>
<span id="cb52-213"><a href="#cb52-213" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-214"><a href="#cb52-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-215"><a href="#cb52-215" aria-hidden="true" tabindex="-1"></a>In order to get a sense for this table, let's investigate each column:</span>
<span id="cb52-216"><a href="#cb52-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-217"><a href="#cb52-217" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`barcode`</span> - This is the barcode for the square on the slide. The format is:</span>
<span id="cb52-218"><a href="#cb52-218" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="co">[</span><span class="ot">ident</span><span class="co">]</span><span class="sc">\_</span><span class="co">[</span><span class="ot">bin_size</span><span class="co">]</span><span class="sc">\_</span><span class="co">[</span><span class="ot">array_row</span><span class="co">]</span><span class="sc">\_</span><span class="co">[</span><span class="ot">array_col</span><span class="co">]</span>-<span class="co">[</span><span class="ot">suffix</span><span class="co">]</span> where:</span>
<span id="cb52-219"><a href="#cb52-219" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span><span class="in">`ident`</span> - The ident</span>
<span id="cb52-220"><a href="#cb52-220" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span><span class="in">`bin_size`</span> - The size of the bin</span>
<span id="cb52-221"><a href="#cb52-221" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span><span class="in">`array_row`</span> - The row number in the array for a given bin size</span>
<span id="cb52-222"><a href="#cb52-222" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span><span class="in">`array_col`</span> - The column number in the array for a given bin size</span>
<span id="cb52-223"><a href="#cb52-223" aria-hidden="true" tabindex="-1"></a><span class="ss">    - </span><span class="in">`suffix`</span> - An optional suffix reserved for technical reasons (usually is <span class="in">`-1`</span>)</span>
<span id="cb52-224"><a href="#cb52-224" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`in_tissue`</span> - A binary descriptor for whether Space Ranger believes this is in the tissue or not</span>
<span id="cb52-225"><a href="#cb52-225" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="in">`0`</span> - Not in the tissue</span>
<span id="cb52-226"><a href="#cb52-226" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span><span class="in">`1`</span> - In the tissue</span>
<span id="cb52-227"><a href="#cb52-227" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`array_row`</span> - The row number in the array for a given bin size</span>
<span id="cb52-228"><a href="#cb52-228" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`array_col`</span> - The column number in the array for a given bin size</span>
<span id="cb52-229"><a href="#cb52-229" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`pxl_row_in_fullres`</span> - Location of row in pixel space</span>
<span id="cb52-230"><a href="#cb52-230" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`pxl_col_in_fullres`</span> - Location of column in pixel space</span>
<span id="cb52-231"><a href="#cb52-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-232"><a href="#cb52-232" aria-hidden="true" tabindex="-1"></a><span class="fu">## Drawing Connections Between the Two Coordinate Spaces</span></span>
<span id="cb52-233"><a href="#cb52-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-234"><a href="#cb52-234" aria-hidden="true" tabindex="-1"></a>The <span class="in">`array_row`</span> and <span class="in">`array_col`</span> are directly tied to <span class="in">`pxl_row_in_fullres`</span> and <span class="in">`pxl_col_in_fullres`</span>, respectively. However the data has undergone an affine transformation within Space Ranger that considers a scale factor (how many pixels per bin), an offset and a rotation. In short, the <span class="in">`array_row`</span> and <span class="in">`array_col`</span> are ralated to <span class="in">`pxl_row_in_fullres`</span> and <span class="in">`pxl_col_in_fullres`</span> and the <span class="in">`parquet`</span> file ties them together. Let's evaluate this by looking at the first line from the <span class="in">`parquet`</span> file:</span>
<span id="cb52-235"><a href="#cb52-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-238"><a href="#cb52-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-239"><a href="#cb52-239" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">pillar.sigfig =</span> <span class="dv">7</span>)</span>
<span id="cb52-240"><a href="#cb52-240" aria-hidden="true" tabindex="-1"></a>parquet_df[parquet_df<span class="sc">$</span>barcode <span class="sc">==</span> <span class="st">"s_016um_00107_00066-1"</span>,]</span>
<span id="cb52-241"><a href="#cb52-241" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-242"><a href="#cb52-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-243"><a href="#cb52-243" aria-hidden="true" tabindex="-1"></a>Now let's look at that same cell in our <span class="in">`coordinates`</span> object</span>
<span id="cb52-244"><a href="#cb52-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-247"><a href="#cb52-247" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-248"><a href="#cb52-248" aria-hidden="true" tabindex="-1"></a>coordinates[coordinates<span class="sc">$</span>cell <span class="sc">==</span> <span class="st">"s_016um_00107_00066-1"</span>,]</span>
<span id="cb52-249"><a href="#cb52-249" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-250"><a href="#cb52-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-251"><a href="#cb52-251" aria-hidden="true" tabindex="-1"></a>We can observe that <span class="in">`pxl_row_in_fullres`</span> in the <span class="in">`parquet`</span> file is the <span class="in">`x`</span> column from Seurat's <span class="in">`GetTissueCoordinates()`</span> function and <span class="in">`pxl_col_in_fullres`</span> in the <span class="in">`parquet`</span> file is the <span class="in">`y`</span> column from Seurat's <span class="in">`GetTissueCoordinates()`</span> function.</span>
<span id="cb52-252"><a href="#cb52-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-253"><a href="#cb52-253" aria-hidden="true" tabindex="-1"></a><span class="fu"># Scope of data</span></span>
<span id="cb52-254"><a href="#cb52-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-255"><a href="#cb52-255" aria-hidden="true" tabindex="-1"></a>We can inspect the number of bins that Space Ranger determine were in a tissue by inspecting the Seurat object:</span>
<span id="cb52-256"><a href="#cb52-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-259"><a href="#cb52-259" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-260"><a href="#cb52-260" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(seurat_obj)</span>
<span id="cb52-261"><a href="#cb52-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-262"><a href="#cb52-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-263"><a href="#cb52-263" aria-hidden="true" tabindex="-1"></a>This data is also encoded in the <span class="in">`parquet`</span> file within the <span class="in">`in_tissue`</span> column where <span class="in">`0`</span> represent not in the tissue and <span class="in">`1`</span> represents in the tissue. </span>
<span id="cb52-264"><a href="#cb52-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-267"><a href="#cb52-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-268"><a href="#cb52-268" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(parquet_df<span class="sc">$</span>in_tissue)</span>
<span id="cb52-269"><a href="#cb52-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-270"><a href="#cb52-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-271"><a href="#cb52-271" aria-hidden="true" tabindex="-1"></a>From this we can see that Seurat has subsampled our dataset to only include the bins that Space Ranger detected as being within the tissue.</span>
<span id="cb52-272"><a href="#cb52-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-273"><a href="#cb52-273" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb52-274"><a href="#cb52-274" aria-hidden="true" tabindex="-1"></a>If we wanted to include all pixels (with and without tissue), we could</span>
<span id="cb52-275"><a href="#cb52-275" aria-hidden="true" tabindex="-1"></a>have used the <span class="in">`filter.matrix = FALSE`</span> option in <span class="in">`Load10X_Spatial()`</span>)</span>
<span id="cb52-276"><a href="#cb52-276" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb52-277"><a href="#cb52-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-278"><a href="#cb52-278" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb52-279"><a href="#cb52-279" aria-hidden="true" tabindex="-1"></a><span class="fu"># Exercise </span></span>
<span id="cb52-280"><a href="#cb52-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-281"><a href="#cb52-281" aria-hidden="true" tabindex="-1"></a>Run the <span class="in">`Load10X_Spatial()`</span> function again and also add the <span class="in">`filter.matrix = FALSE`</span> option and assign it to <span class="in">`exercise_obj`</span>. Pull the coordinates with the <span class="in">`GetTissueCoordinates()`</span> function and assigning them to <span class="in">`exercise_coordinates`</span>.</span>
<span id="cb52-282"><a href="#cb52-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-283"><a href="#cb52-283" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Determine the <span class="in">`exercise_coordinates`</span> object to determine the number of rows of and the number of.</span>
<span id="cb52-284"><a href="#cb52-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-287"><a href="#cb52-287" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-288"><a href="#cb52-288" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_exercise_object</span></span>
<span id="cb52-289"><a href="#cb52-289" aria-hidden="true" tabindex="-1"></a><span class="co"># Load spatial data into Seurat object</span></span>
<span id="cb52-290"><a href="#cb52-290" aria-hidden="true" tabindex="-1"></a>exercise_obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-291"><a href="#cb52-291" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.dir =</span> <span class="st">"data/fresh_frozen"</span>,</span>
<span id="cb52-292"><a href="#cb52-292" aria-hidden="true" tabindex="-1"></a>  <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb52-293"><a href="#cb52-293" aria-hidden="true" tabindex="-1"></a>  <span class="at">slice =</span> <span class="st">"fresh_frozen"</span>,</span>
<span id="cb52-294"><a href="#cb52-294" aria-hidden="true" tabindex="-1"></a>  <span class="at">filter.matrix =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-295"><a href="#cb52-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-296"><a href="#cb52-296" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract tissue coordinates from the Seurat object</span></span>
<span id="cb52-297"><a href="#cb52-297" aria-hidden="true" tabindex="-1"></a>exercise_coordinates <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(exercise_obj)</span>
<span id="cb52-298"><a href="#cb52-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-299"><a href="#cb52-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-300"><a href="#cb52-300" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>In a single 6.5mm x 6.5mm capture area from Visium HD, we would</span>
<span id="cb52-301"><a href="#cb52-301" aria-hidden="true" tabindex="-1"></a>    expect that to correspond to about 20,000 pixels by 20,000 pixels.</span>
<span id="cb52-302"><a href="#cb52-302" aria-hidden="true" tabindex="-1"></a>    Do our expectations match our observed data?</span>
<span id="cb52-303"><a href="#cb52-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-304"><a href="#cb52-304" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>By comparing the dimensions of <span class="in">`coordinates`</span>, which only includes</span>
<span id="cb52-305"><a href="#cb52-305" aria-hidden="true" tabindex="-1"></a>    coordinates for locations that Space Ranger detected tissue, and</span>
<span id="cb52-306"><a href="#cb52-306" aria-hidden="true" tabindex="-1"></a>    <span class="in">`exercise_coordinates`</span>, which contains all coordinates with and</span>
<span id="cb52-307"><a href="#cb52-307" aria-hidden="true" tabindex="-1"></a>    without tissue detected by Space Ranger, how many pixels didn't have</span>
<span id="cb52-308"><a href="#cb52-308" aria-hidden="true" tabindex="-1"></a>    tissue detected in them?</span>
<span id="cb52-309"><a href="#cb52-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-310"><a href="#cb52-310" aria-hidden="true" tabindex="-1"></a>After completing this exercise, be sure to remove <span class="in">`exercise_obj`</span> and</span>
<span id="cb52-311"><a href="#cb52-311" aria-hidden="true" tabindex="-1"></a><span class="in">`exercise_coordinates`</span> from our R environment to save on memory by</span>
<span id="cb52-312"><a href="#cb52-312" aria-hidden="true" tabindex="-1"></a>using:</span>
<span id="cb52-313"><a href="#cb52-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-316"><a href="#cb52-316" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-317"><a href="#cb52-317" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: remove_exercise_objects</span></span>
<span id="cb52-318"><a href="#cb52-318" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove exercise Seurat object</span></span>
<span id="cb52-319"><a href="#cb52-319" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(exercise_obj)</span>
<span id="cb52-320"><a href="#cb52-320" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove exercise coordinate object</span></span>
<span id="cb52-321"><a href="#cb52-321" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(exercise_coordinates)</span>
<span id="cb52-322"><a href="#cb52-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-323"><a href="#cb52-323" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb52-324"><a href="#cb52-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-325"><a href="#cb52-325" aria-hidden="true" tabindex="-1"></a><span class="fu"># Subsetting the Seurat Object</span></span>
<span id="cb52-326"><a href="#cb52-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-327"><a href="#cb52-327" aria-hidden="true" tabindex="-1"></a>Now that we have an understanding of the coordinates, let's go ahead and subset our coordinates based upon the region we will like to focus on in this workshop:</span>
<span id="cb52-328"><a href="#cb52-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-331"><a href="#cb52-331" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-332"><a href="#cb52-332" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: subset_coords</span></span>
<span id="cb52-333"><a href="#cb52-333" aria-hidden="true" tabindex="-1"></a><span class="co"># Set x and y bounds for subsetting image</span></span>
<span id="cb52-334"><a href="#cb52-334" aria-hidden="true" tabindex="-1"></a>x_min <span class="ot">&lt;-</span> <span class="sc">-</span><span class="cn">Inf</span> </span>
<span id="cb52-335"><a href="#cb52-335" aria-hidden="true" tabindex="-1"></a>x_max <span class="ot">&lt;-</span> <span class="dv">15114</span></span>
<span id="cb52-336"><a href="#cb52-336" aria-hidden="true" tabindex="-1"></a>y_min <span class="ot">&lt;-</span> <span class="sc">-</span><span class="cn">Inf</span></span>
<span id="cb52-337"><a href="#cb52-337" aria-hidden="true" tabindex="-1"></a>y_max <span class="ot">&lt;-</span> <span class="dv">21517</span></span>
<span id="cb52-338"><a href="#cb52-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-339"><a href="#cb52-339" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract bin barcodes based on the defined x and y bounds</span></span>
<span id="cb52-340"><a href="#cb52-340" aria-hidden="true" tabindex="-1"></a>subsetted_bins <span class="ot">&lt;-</span> coordinates <span class="sc">%&gt;%</span> </span>
<span id="cb52-341"><a href="#cb52-341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">subset</span>((x <span class="sc">&gt;</span> x_min <span class="sc">&amp;</span> x <span class="sc">&lt;</span> x_max) <span class="sc">&amp;</span> </span>
<span id="cb52-342"><a href="#cb52-342" aria-hidden="true" tabindex="-1"></a>           (y <span class="sc">&gt;</span> y_min <span class="sc">&amp;</span> y <span class="sc">&lt;</span> y_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb52-343"><a href="#cb52-343" aria-hidden="true" tabindex="-1"></a>    <span class="fu">row.names</span>()</span>
<span id="cb52-344"><a href="#cb52-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-345"><a href="#cb52-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-346"><a href="#cb52-346" aria-hidden="true" tabindex="-1"></a>The output of this subsetting should be bin names that fit this critera for <span class="in">`x`</span> and <span class="in">`y`</span> bounds. We can inspect the object with:</span>
<span id="cb52-347"><a href="#cb52-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-350"><a href="#cb52-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-351"><a href="#cb52-351" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: inspect_subsetted_bins</span></span>
<span id="cb52-352"><a href="#cb52-352" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect subsetted bin names</span></span>
<span id="cb52-353"><a href="#cb52-353" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(subsetted_bins)</span>
<span id="cb52-354"><a href="#cb52-354" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-355"><a href="#cb52-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-356"><a href="#cb52-356" aria-hidden="true" tabindex="-1"></a>We can confirm that this subsetting was done appropriately, by subsetting the <span class="in">`coordinates`</span> object by our <span class="in">`subsetted_bins`</span> names. </span>
<span id="cb52-357"><a href="#cb52-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-360"><a href="#cb52-360" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-361"><a href="#cb52-361" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: subsetting_sanity_checks</span></span>
<span id="cb52-362"><a href="#cb52-362" aria-hidden="true" tabindex="-1"></a>coordinates[subsetted_bins,] <span class="sc">%&gt;%</span> </span>
<span id="cb52-363"><a href="#cb52-363" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span>
<span id="cb52-364"><a href="#cb52-364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-365"><a href="#cb52-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-366"><a href="#cb52-366" aria-hidden="true" tabindex="-1"></a>From here we can see our pixels are all within the pixel range that we subsetted by, so we feel confident that our subsetting was appropriate. </span>
<span id="cb52-367"><a href="#cb52-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-368"><a href="#cb52-368" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb52-369"><a href="#cb52-369" aria-hidden="true" tabindex="-1"></a>Sanity checks like the above check may feel unnecessary but they are really important for good data analysis practices. It oftentimes takes less than a few moments to inspect and can save you lots of time downstream when you are debugging code.</span>
<span id="cb52-370"><a href="#cb52-370" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb52-371"><a href="#cb52-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-372"><a href="#cb52-372" aria-hidden="true" tabindex="-1"></a>Let's go ahead and start subsetting our Seurat object. This will happen in three parts:</span>
<span id="cb52-373"><a href="#cb52-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-374"><a href="#cb52-374" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Creating a new column, <span class="in">`filter`</span>, in <span class="in">`meta.data`</span> of the Seurat object</span>
<span id="cb52-375"><a href="#cb52-375" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Assigning the values of <span class="in">`TRUE`</span> or <span class="in">`FALSE`</span> for bins to be removed or kept, respectively</span>
<span id="cb52-376"><a href="#cb52-376" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Subsetting the Seurat object based about the data in the <span class="in">`filter`</span> column of <span class="in">`meta.data`</span></span>
<span id="cb52-377"><a href="#cb52-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-380"><a href="#cb52-380" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-381"><a href="#cb52-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: subsetting_seurat_object</span></span>
<span id="cb52-382"><a href="#cb52-382" aria-hidden="true" tabindex="-1"></a><span class="co"># Label a column called filter for TRUE</span></span>
<span id="cb52-383"><a href="#cb52-383" aria-hidden="true" tabindex="-1"></a>seurat_obj<span class="sc">$</span>filter <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb52-384"><a href="#cb52-384" aria-hidden="true" tabindex="-1"></a><span class="co"># Re-assign the value of FALSE to the filter column if the bin is within subsetted_bins</span></span>
<span id="cb52-385"><a href="#cb52-385" aria-hidden="true" tabindex="-1"></a>seurat_obj<span class="sc">@</span>meta.data[subsetted_bins, <span class="st">"filter"</span>] <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb52-386"><a href="#cb52-386" aria-hidden="true" tabindex="-1"></a><span class="co"># Retain the bins which have FALSE for filter in the meta data</span></span>
<span id="cb52-387"><a href="#cb52-387" aria-hidden="true" tabindex="-1"></a>seurat_obj_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_obj, filter <span class="sc">==</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-388"><a href="#cb52-388" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-389"><a href="#cb52-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-390"><a href="#cb52-390" aria-hidden="true" tabindex="-1"></a>Just like we did previously when subsetting coordinates, let's confirm that we are subsetting the correct bins.</span>
<span id="cb52-391"><a href="#cb52-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-394"><a href="#cb52-394" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-395"><a href="#cb52-395" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: subsetted_seurat_sanity_check</span></span>
<span id="cb52-396"><a href="#cb52-396" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the first few lines of coordinates to ensure they match our expectation</span></span>
<span id="cb52-397"><a href="#cb52-397" aria-hidden="true" tabindex="-1"></a><span class="fu">GetTissueCoordinates</span>(seurat_obj_filt) <span class="sc">%&gt;%</span>  </span>
<span id="cb52-398"><a href="#cb52-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span>
<span id="cb52-399"><a href="#cb52-399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-400"><a href="#cb52-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-401"><a href="#cb52-401" aria-hidden="true" tabindex="-1"></a>Last, we can use the <span class="in">`SpatialFeaturePlot()`</span> function from Seurat to compare the before and after of subsetting.</span>
<span id="cb52-402"><a href="#cb52-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-405"><a href="#cb52-405" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-406"><a href="#cb52-406" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: visualize_subsetted_bins</span></span>
<span id="cb52-407"><a href="#cb52-407" aria-hidden="true" tabindex="-1"></a><span class="co"># Visual the original image and the subsetted image next to each other</span></span>
<span id="cb52-408"><a href="#cb52-408" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(seurat_obj, <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb52-409"><a href="#cb52-409" aria-hidden="true" tabindex="-1"></a>       <span class="fu">SpatialFeaturePlot</span>(seurat_obj_filt, <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-410"><a href="#cb52-410" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-411"><a href="#cb52-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-412"><a href="#cb52-412" aria-hidden="true" tabindex="-1"></a><span class="fu"># Writing Subsetted Data Directory</span></span>
<span id="cb52-413"><a href="#cb52-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-414"><a href="#cb52-414" aria-hidden="true" tabindex="-1"></a>It looks like we have correctly subsetted out data and now we are ready to create a directory to hold our data. This will make it easier in the future to work with our data, because we can use the <span class="in">`Load10X_Spatial()`</span> function from Seurat to read in our subsetted data rather than performing the subsetting again.</span>
<span id="cb52-415"><a href="#cb52-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-416"><a href="#cb52-416" aria-hidden="true" tabindex="-1"></a>We are going to start by creating a directory structure to hold our subsetted data:</span>
<span id="cb52-417"><a href="#cb52-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-420"><a href="#cb52-420" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-421"><a href="#cb52-421" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: establish_directory_structure</span></span>
<span id="cb52-422"><a href="#cb52-422" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>))</span>
<span id="cb52-423"><a href="#cb52-423" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/"</span>))</span>
<span id="cb52-424"><a href="#cb52-424" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/"</span>))</span>
<span id="cb52-425"><a href="#cb52-425" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/"</span>))</span>
<span id="cb52-426"><a href="#cb52-426" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>))</span>
<span id="cb52-427"><a href="#cb52-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-428"><a href="#cb52-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-429"><a href="#cb52-429" aria-hidden="true" tabindex="-1"></a>The directory structure should look like:</span>
<span id="cb52-430"><a href="#cb52-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-433"><a href="#cb52-433" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-434"><a href="#cb52-434" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: directory_structure_before</span></span>
<span id="cb52-435"><a href="#cb52-435" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-436"><a href="#cb52-436" aria-hidden="true" tabindex="-1"></a><span class="fu">dir_tree</span>(<span class="st">"data_filtered/"</span>)</span>
<span id="cb52-437"><a href="#cb52-437" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-438"><a href="#cb52-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-439"><a href="#cb52-439" aria-hidden="true" tabindex="-1"></a>Now we will populate this directory with the files needed to be able to run the <span class="in">`Load10X_Spatial()`</span> function in the future.</span>
<span id="cb52-440"><a href="#cb52-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-443"><a href="#cb52-443" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-444"><a href="#cb52-444" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: write_10X_data</span></span>
<span id="cb52-445"><a href="#cb52-445" aria-hidden="true" tabindex="-1"></a><span class="co">#| output: false</span></span>
<span id="cb52-446"><a href="#cb52-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-447"><a href="#cb52-447" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy scale factors JSON to new directory</span></span>
<span id="cb52-448"><a href="#cb52-448" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb52-449"><a href="#cb52-449" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/scalefactors_json.json"</span>,</span>
<span id="cb52-450"><a href="#cb52-450" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb52-451"><a href="#cb52-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-452"><a href="#cb52-452" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy tissue positions parquet file to new directory</span></span>
<span id="cb52-453"><a href="#cb52-453" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb52-454"><a href="#cb52-454" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_positions.parquet"</span>,</span>
<span id="cb52-455"><a href="#cb52-455" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb52-456"><a href="#cb52-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-457"><a href="#cb52-457" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy low resolution PNG to new directory</span></span>
<span id="cb52-458"><a href="#cb52-458" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(</span>
<span id="cb52-459"><a href="#cb52-459" aria-hidden="true" tabindex="-1"></a>  <span class="at">from =</span> <span class="st">"data/fresh_frozen/binned_outputs/square_016um/spatial/tissue_lowres_image.png"</span>,</span>
<span id="cb52-460"><a href="#cb52-460" aria-hidden="true" tabindex="-1"></a>  <span class="at">to =</span> <span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/spatial/"</span>)</span>
<span id="cb52-461"><a href="#cb52-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-462"><a href="#cb52-462" aria-hidden="true" tabindex="-1"></a><span class="co"># Write counts to hd5 file</span></span>
<span id="cb52-463"><a href="#cb52-463" aria-hidden="true" tabindex="-1"></a><span class="fu">write10xCounts</span>(<span class="st">"data_filtered/fresh_frozen/binned_outputs/square_016um/filtered_feature_bc_matrix.h5"</span>,</span>
<span id="cb52-464"><a href="#cb52-464" aria-hidden="true" tabindex="-1"></a>  <span class="at">x=</span><span class="fu">LayerData</span>(seurat_obj_filt),</span>
<span id="cb52-465"><a href="#cb52-465" aria-hidden="true" tabindex="-1"></a>  <span class="at">barcodes=</span><span class="fu">colnames</span>(seurat_obj_filt),</span>
<span id="cb52-466"><a href="#cb52-466" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene.id=</span><span class="fu">rownames</span>(seurat_obj_filt),</span>
<span id="cb52-467"><a href="#cb52-467" aria-hidden="true" tabindex="-1"></a>  <span class="at">version=</span><span class="st">"3"</span>,</span>
<span id="cb52-468"><a href="#cb52-468" aria-hidden="true" tabindex="-1"></a>  <span class="at">type=</span><span class="st">"HDF5"</span>,</span>
<span id="cb52-469"><a href="#cb52-469" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-470"><a href="#cb52-470" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-471"><a href="#cb52-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-472"><a href="#cb52-472" aria-hidden="true" tabindex="-1"></a>After writing and copying the files the file directory should look like this:</span>
<span id="cb52-473"><a href="#cb52-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-476"><a href="#cb52-476" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-477"><a href="#cb52-477" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: directory_structure_after</span></span>
<span id="cb52-478"><a href="#cb52-478" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-479"><a href="#cb52-479" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the directory structure after copying and writing files</span></span>
<span id="cb52-480"><a href="#cb52-480" aria-hidden="true" tabindex="-1"></a><span class="fu">dir_tree</span>(<span class="st">"data_filtered/"</span>)</span>
<span id="cb52-481"><a href="#cb52-481" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-482"><a href="#cb52-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-483"><a href="#cb52-483" aria-hidden="true" tabindex="-1"></a><span class="fu"># Automate subsetting by looping through samples and bins</span></span>
<span id="cb52-484"><a href="#cb52-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-485"><a href="#cb52-485" aria-hidden="true" tabindex="-1"></a>Now we have processed a single bin (16μm) for a single sample (fresh_frozen). However, we would like to process our data for 8μm and 16μm for both "fresh_frozen" and "fixed_frozen" This process can be automated by using the same process as above, but looping through both bin sizes and both samples. **Due to memory constraints on some laptops, you many not be able to run the code for the remaining parts of the lesson**. It requires ~32Gb of RAM to execute.</span>
<span id="cb52-486"><a href="#cb52-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-487"><a href="#cb52-487" aria-hidden="true" tabindex="-1"></a><span class="fu">## Setting variables</span></span>
<span id="cb52-488"><a href="#cb52-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-489"><a href="#cb52-489" aria-hidden="true" tabindex="-1"></a>We are going to set-up the the variables that we would like to loop through (bin size and sample) and also set our boundaries for where to subset our sample for. The index of the <span class="in">`folders`</span> vector is the sample index for the boundary vectors (<span class="in">`x_min`</span>, <span class="in">`x_max`</span>, <span class="in">`y_min`</span>, <span class="in">`y_max`</span>).For example, the <span class="in">`xmin`</span> of <span class="in">`-Inf`</span> corresponds to the fresh_frozen sample while the <span class="in">`xmin`</span> of <span class="in">`7205.1`</span> corresponds to the fixed_frozen sample.</span>
<span id="cb52-490"><a href="#cb52-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-491"><a href="#cb52-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-494"><a href="#cb52-494" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-495"><a href="#cb52-495" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: set_bins_sizes</span></span>
<span id="cb52-496"><a href="#cb52-496" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-497"><a href="#cb52-497" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-498"><a href="#cb52-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-499"><a href="#cb52-499" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"008"</span>, <span class="st">"016"</span>)</span>
<span id="cb52-500"><a href="#cb52-500" aria-hidden="true" tabindex="-1"></a>folders <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"fresh_frozen"</span>, <span class="st">"fixed_frozen"</span>)</span>
<span id="cb52-501"><a href="#cb52-501" aria-hidden="true" tabindex="-1"></a>x_min <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">7205.1</span>) </span>
<span id="cb52-502"><a href="#cb52-502" aria-hidden="true" tabindex="-1"></a>x_max <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15114</span>, <span class="cn">Inf</span>)</span>
<span id="cb52-503"><a href="#cb52-503" aria-hidden="true" tabindex="-1"></a>y_min <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="sc">-</span><span class="dv">1592</span>)</span>
<span id="cb52-504"><a href="#cb52-504" aria-hidden="true" tabindex="-1"></a>y_max <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">21517</span>, <span class="dv">17735</span>)</span>
<span id="cb52-505"><a href="#cb52-505" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-506"><a href="#cb52-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-507"><a href="#cb52-507" aria-hidden="true" tabindex="-1"></a>The code below will allow us to loop through the different samples and bin sizes, adding them to a <span class="in">`data_filtered/`</span>. </span>
<span id="cb52-508"><a href="#cb52-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-511"><a href="#cb52-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-512"><a href="#cb52-512" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: reading_in_samples</span></span>
<span id="cb52-513"><a href="#cb52-513" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-514"><a href="#cb52-514" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-515"><a href="#cb52-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-516"><a href="#cb52-516" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (folder <span class="cf">in</span> folders) {</span>
<span id="cb52-517"><a href="#cb52-517" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dir.create</span>(<span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>, folder, <span class="st">"/binned_outputs/"</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-518"><a href="#cb52-518" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (bin <span class="cf">in</span> bins) {</span>
<span id="cb52-519"><a href="#cb52-519" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-520"><a href="#cb52-520" aria-hidden="true" tabindex="-1"></a>    path_bin <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data_filtered/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, <span class="st">"um"</span>)</span>
<span id="cb52-521"><a href="#cb52-521" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(path_bin, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-522"><a href="#cb52-522" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-523"><a href="#cb52-523" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Copy spatial images from original folder</span></span>
<span id="cb52-524"><a href="#cb52-524" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(<span class="fu">paste0</span>(path_bin, <span class="st">"/spatial"</span>), <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-525"><a href="#cb52-525" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-526"><a href="#cb52-526" aria-hidden="true" tabindex="-1"></a>    path_json <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb52-527"><a href="#cb52-527" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/scalefactors_json.json"</span>)</span>
<span id="cb52-528"><a href="#cb52-528" aria-hidden="true" tabindex="-1"></a>    path_parq <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb52-529"><a href="#cb52-529" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/tissue_positions.parquet"</span>)</span>
<span id="cb52-530"><a href="#cb52-530" aria-hidden="true" tabindex="-1"></a>    path_png  <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder, <span class="st">"/binned_outputs/square_"</span>, bin, </span>
<span id="cb52-531"><a href="#cb52-531" aria-hidden="true" tabindex="-1"></a>                      <span class="st">"um/spatial/tissue_lowres_image.png"</span>)</span>
<span id="cb52-532"><a href="#cb52-532" aria-hidden="true" tabindex="-1"></a>    files_spatial <span class="ot">&lt;-</span> <span class="fu">c</span>(path_json, path_parq, path_png)</span>
<span id="cb52-533"><a href="#cb52-533" aria-hidden="true" tabindex="-1"></a>    <span class="fu">file.copy</span>(files_spatial,<span class="fu">paste0</span>(path_bin, <span class="st">"/spatial"</span>))</span>
<span id="cb52-534"><a href="#cb52-534" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-535"><a href="#cb52-535" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Load object</span></span>
<span id="cb52-536"><a href="#cb52-536" aria-hidden="true" tabindex="-1"></a>    obj <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-537"><a href="#cb52-537" aria-hidden="true" tabindex="-1"></a>      <span class="at">data.dir =</span> <span class="fu">paste0</span>(<span class="st">"data/"</span>, folder),</span>
<span id="cb52-538"><a href="#cb52-538" aria-hidden="true" tabindex="-1"></a>      <span class="at">bin.size =</span> <span class="fu">c</span>(<span class="fu">as.integer</span>(bin)),</span>
<span id="cb52-539"><a href="#cb52-539" aria-hidden="true" tabindex="-1"></a>      <span class="at">slice =</span> folder)</span>
<span id="cb52-540"><a href="#cb52-540" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-541"><a href="#cb52-541" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Cells to keep based on coordinates</span></span>
<span id="cb52-542"><a href="#cb52-542" aria-hidden="true" tabindex="-1"></a>    coords <span class="ot">&lt;-</span> <span class="fu">GetTissueCoordinates</span>(obj)</span>
<span id="cb52-543"><a href="#cb52-543" aria-hidden="true" tabindex="-1"></a>    cells <span class="ot">&lt;-</span> coords <span class="sc">%&gt;%</span> </span>
<span id="cb52-544"><a href="#cb52-544" aria-hidden="true" tabindex="-1"></a>      <span class="fu">subset</span>((x <span class="sc">&gt;</span> x_min[<span class="fu">which</span>(folders <span class="sc">==</span> folder)] <span class="sc">&amp;</span> x <span class="sc">&lt;</span> x_max[<span class="fu">which</span>(folders <span class="sc">==</span> folder)]) <span class="sc">&amp;</span> </span>
<span id="cb52-545"><a href="#cb52-545" aria-hidden="true" tabindex="-1"></a>             (y <span class="sc">&gt;</span> y_min[<span class="fu">which</span>(folders <span class="sc">==</span> folder)] <span class="sc">&amp;</span> y <span class="sc">&lt;</span> y_max[<span class="fu">which</span>(folders <span class="sc">==</span> folder)])) <span class="sc">%&gt;%</span> </span>
<span id="cb52-546"><a href="#cb52-546" aria-hidden="true" tabindex="-1"></a>      <span class="fu">row.names</span>()</span>
<span id="cb52-547"><a href="#cb52-547" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-548"><a href="#cb52-548" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Subset object</span></span>
<span id="cb52-549"><a href="#cb52-549" aria-hidden="true" tabindex="-1"></a>    obj<span class="sc">$</span>filter <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb52-550"><a href="#cb52-550" aria-hidden="true" tabindex="-1"></a>    obj<span class="sc">@</span>meta.data[cells, <span class="st">"filter"</span>] <span class="ot">&lt;-</span> <span class="cn">FALSE</span></span>
<span id="cb52-551"><a href="#cb52-551" aria-hidden="true" tabindex="-1"></a>    obj_filt <span class="ot">&lt;-</span> <span class="fu">subset</span>(obj, filter <span class="sc">==</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-552"><a href="#cb52-552" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-553"><a href="#cb52-553" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">SpatialFeaturePlot</span>(obj, <span class="fu">paste0</span>(<span class="st">"nCount_Spatial."</span>, bin, <span class="st">"um"</span>), <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb52-554"><a href="#cb52-554" aria-hidden="true" tabindex="-1"></a>         <span class="fu">SpatialFeaturePlot</span>(obj_filt, <span class="fu">paste0</span>(<span class="st">"nCount_Spatial."</span>, bin, <span class="st">"um"</span>), <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-555"><a href="#cb52-555" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(p)</span>
<span id="cb52-556"><a href="#cb52-556" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-557"><a href="#cb52-557" aria-hidden="true" tabindex="-1"></a>    path_h5 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(path_bin, <span class="st">"/filtered_feature_bc_matrix.h5"</span>)</span>
<span id="cb52-558"><a href="#cb52-558" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write10xCounts</span>(path_h5,</span>
<span id="cb52-559"><a href="#cb52-559" aria-hidden="true" tabindex="-1"></a>                   <span class="at">x=</span><span class="fu">LayerData</span>(obj_filt),</span>
<span id="cb52-560"><a href="#cb52-560" aria-hidden="true" tabindex="-1"></a>                   <span class="at">barcodes=</span><span class="fu">colnames</span>(obj_filt),</span>
<span id="cb52-561"><a href="#cb52-561" aria-hidden="true" tabindex="-1"></a>                   <span class="at">gene.id=</span><span class="fu">rownames</span>(obj_filt),</span>
<span id="cb52-562"><a href="#cb52-562" aria-hidden="true" tabindex="-1"></a>                   <span class="at">version=</span><span class="st">"3"</span>,</span>
<span id="cb52-563"><a href="#cb52-563" aria-hidden="true" tabindex="-1"></a>                   <span class="at">type=</span><span class="st">"HDF5"</span>,</span>
<span id="cb52-564"><a href="#cb52-564" aria-hidden="true" tabindex="-1"></a>                   <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span>
<span id="cb52-565"><a href="#cb52-565" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-566"><a href="#cb52-566" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(obj)</span>
<span id="cb52-567"><a href="#cb52-567" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(obj_filt)</span>
<span id="cb52-568"><a href="#cb52-568" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb52-569"><a href="#cb52-569" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-570"><a href="#cb52-570" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-571"><a href="#cb52-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-572"><a href="#cb52-572" aria-hidden="true" tabindex="-1"></a><span class="fu">## Test Reading in `fresh_frozen` Bin Sizes</span></span>
<span id="cb52-573"><a href="#cb52-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-574"><a href="#cb52-574" aria-hidden="true" tabindex="-1"></a>If we had run been able to run the above for loops, our directory structure would now look like:</span>
<span id="cb52-575"><a href="#cb52-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-578"><a href="#cb52-578" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-579"><a href="#cb52-579" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: read_in_tree</span></span>
<span id="cb52-580"><a href="#cb52-580" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-581"><a href="#cb52-581" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-582"><a href="#cb52-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-583"><a href="#cb52-583" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-584"><a href="#cb52-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-587"><a href="#cb52-587" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-588"><a href="#cb52-588" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: print_tree</span></span>
<span id="cb52-589"><a href="#cb52-589" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-590"><a href="#cb52-590" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb52-591"><a href="#cb52-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-592"><a href="#cb52-592" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-593"><a href="#cb52-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-594"><a href="#cb52-594" aria-hidden="true" tabindex="-1"></a>With this directory structure, let's go ahead and load in the "fresh_frozen" sample for the 16μm bin size:</span>
<span id="cb52-595"><a href="#cb52-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-598"><a href="#cb52-598" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-599"><a href="#cb52-599" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_fresh_16μm</span></span>
<span id="cb52-600"><a href="#cb52-600" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-601"><a href="#cb52-601" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-602"><a href="#cb52-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-603"><a href="#cb52-603" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fresh_frozen 16μm</span></span>
<span id="cb52-604"><a href="#cb52-604" aria-hidden="true" tabindex="-1"></a>fresh_16 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-605"><a href="#cb52-605" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fresh_frozen"</span>,</span>
<span id="cb52-606"><a href="#cb52-606" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb52-607"><a href="#cb52-607" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fresh"</span>)</span>
<span id="cb52-608"><a href="#cb52-608" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-609"><a href="#cb52-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-610"><a href="#cb52-610" aria-hidden="true" tabindex="-1"></a>Next we will load in the "fresh_frozen" sample for the 8μm bin size:</span>
<span id="cb52-611"><a href="#cb52-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-614"><a href="#cb52-614" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-615"><a href="#cb52-615" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_fresh_8μm</span></span>
<span id="cb52-616"><a href="#cb52-616" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-617"><a href="#cb52-617" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-618"><a href="#cb52-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-619"><a href="#cb52-619" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fresh_frozen 8μm</span></span>
<span id="cb52-620"><a href="#cb52-620" aria-hidden="true" tabindex="-1"></a>fresh_8 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-621"><a href="#cb52-621" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fresh_frozen"</span>,</span>
<span id="cb52-622"><a href="#cb52-622" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">8</span>,</span>
<span id="cb52-623"><a href="#cb52-623" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fresh"</span>)</span>
<span id="cb52-624"><a href="#cb52-624" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-625"><a href="#cb52-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-626"><a href="#cb52-626" aria-hidden="true" tabindex="-1"></a>We can confirm that the number of bins in the <span class="in">`fresh_8`</span> is greater than the number of bins in the <span class="in">`fresh_16`</span> as we would expect:</span>
<span id="cb52-627"><a href="#cb52-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-630"><a href="#cb52-630" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-631"><a href="#cb52-631" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fresh_compare_cells</span></span>
<span id="cb52-632"><a href="#cb52-632" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-633"><a href="#cb52-633" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-634"><a href="#cb52-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-635"><a href="#cb52-635" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 8μm fresh_frozen sample</span></span>
<span id="cb52-636"><a href="#cb52-636" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fresh_8)</span>
<span id="cb52-637"><a href="#cb52-637" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 16μm fresh_frozen sample</span></span>
<span id="cb52-638"><a href="#cb52-638" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fresh_16)</span>
<span id="cb52-639"><a href="#cb52-639" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-640"><a href="#cb52-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-641"><a href="#cb52-641" aria-hidden="true" tabindex="-1"></a>Lastly, we can visualize both bin sizes for our "fresh_frozen" samples side-by-side using <span class="in">`SpatialFeaturePlot()`</span>:</span>
<span id="cb52-642"><a href="#cb52-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-645"><a href="#cb52-645" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-646"><a href="#cb52-646" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fresh_compare_feature_plot</span></span>
<span id="cb52-647"><a href="#cb52-647" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-648"><a href="#cb52-648" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-649"><a href="#cb52-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-650"><a href="#cb52-650" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare SpatialFeaturePlot of fresh_8 and fresh_16</span></span>
<span id="cb52-651"><a href="#cb52-651" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(fresh_8,  <span class="st">"nCount_Spatial.008um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">|</span></span>
<span id="cb52-652"><a href="#cb52-652" aria-hidden="true" tabindex="-1"></a>  <span class="fu">SpatialFeaturePlot</span>(fresh_16,  <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-653"><a href="#cb52-653" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-654"><a href="#cb52-654" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-655"><a href="#cb52-655" aria-hidden="true" tabindex="-1"></a><span class="fu">## Test Reading in `fixed_frozen` Bin Sizes</span></span>
<span id="cb52-656"><a href="#cb52-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-657"><a href="#cb52-657" aria-hidden="true" tabindex="-1"></a>We will iterate those steps for the "fixed_frozen" sample:. First, we will read in the 16μm bin size:</span>
<span id="cb52-658"><a href="#cb52-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-661"><a href="#cb52-661" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-662"><a href="#cb52-662" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_fized_16μm</span></span>
<span id="cb52-663"><a href="#cb52-663" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-664"><a href="#cb52-664" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-665"><a href="#cb52-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-666"><a href="#cb52-666" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fixed_frozen 16μm</span></span>
<span id="cb52-667"><a href="#cb52-667" aria-hidden="true" tabindex="-1"></a>fixed_16 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-668"><a href="#cb52-668" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fixed_frozen"</span>,</span>
<span id="cb52-669"><a href="#cb52-669" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">16</span>,</span>
<span id="cb52-670"><a href="#cb52-670" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fixed"</span>)</span>
<span id="cb52-671"><a href="#cb52-671" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-672"><a href="#cb52-672" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-673"><a href="#cb52-673" aria-hidden="true" tabindex="-1"></a>Next, we will read in the 8μm bin size for the "fixed_frozen" sample:</span>
<span id="cb52-674"><a href="#cb52-674" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-677"><a href="#cb52-677" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-678"><a href="#cb52-678" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load_fixed_8μm</span></span>
<span id="cb52-679"><a href="#cb52-679" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-680"><a href="#cb52-680" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-681"><a href="#cb52-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-682"><a href="#cb52-682" aria-hidden="true" tabindex="-1"></a><span class="co"># Load fixed_frozen 8μm</span></span>
<span id="cb52-683"><a href="#cb52-683" aria-hidden="true" tabindex="-1"></a>fixed_8 <span class="ot">&lt;-</span> <span class="fu">Load10X_Spatial</span>(</span>
<span id="cb52-684"><a href="#cb52-684" aria-hidden="true" tabindex="-1"></a>    <span class="at">data.dir =</span> <span class="st">"data_filtered/fixed_frozen"</span>,</span>
<span id="cb52-685"><a href="#cb52-685" aria-hidden="true" tabindex="-1"></a>    <span class="at">bin.size =</span> <span class="dv">8</span>,</span>
<span id="cb52-686"><a href="#cb52-686" aria-hidden="true" tabindex="-1"></a>    <span class="at">slice =</span> <span class="st">"fixed"</span>)</span>
<span id="cb52-687"><a href="#cb52-687" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-688"><a href="#cb52-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-689"><a href="#cb52-689" aria-hidden="true" tabindex="-1"></a>Once again, we will do a check to ensure that the number of bins in the <span class="in">`fixed_8`</span> is greater than the number of bins in the <span class="in">`fixed_16`</span> as we would expect:</span>
<span id="cb52-690"><a href="#cb52-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-693"><a href="#cb52-693" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-694"><a href="#cb52-694" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fixed_compare_cells</span></span>
<span id="cb52-695"><a href="#cb52-695" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-696"><a href="#cb52-696" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-697"><a href="#cb52-697" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-698"><a href="#cb52-698" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 8μm fixed_frozen sample</span></span>
<span id="cb52-699"><a href="#cb52-699" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fixed_8)</span>
<span id="cb52-700"><a href="#cb52-700" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of bins in the 16μm fixed_frozen sample</span></span>
<span id="cb52-701"><a href="#cb52-701" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(fixed_16)</span>
<span id="cb52-702"><a href="#cb52-702" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-703"><a href="#cb52-703" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-704"><a href="#cb52-704" aria-hidden="true" tabindex="-1"></a>Lastly, we can check that both bin sizes have been subsetted correctly and view them with <span class="in">`SpatialFeaturePlot()`</span>: </span>
<span id="cb52-705"><a href="#cb52-705" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-708"><a href="#cb52-708" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-709"><a href="#cb52-709" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fixed_compare_feature_plot</span></span>
<span id="cb52-710"><a href="#cb52-710" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-711"><a href="#cb52-711" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-712"><a href="#cb52-712" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-713"><a href="#cb52-713" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare SpatialFeaturePlot of fixed_8 and fixed_16</span></span>
<span id="cb52-714"><a href="#cb52-714" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(fixed_8,  <span class="st">"nCount_Spatial.008um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>) <span class="sc">|</span></span>
<span id="cb52-715"><a href="#cb52-715" aria-hidden="true" tabindex="-1"></a>  <span class="fu">SpatialFeaturePlot</span>(fixed_16,  <span class="st">"nCount_Spatial.016um"</span>, <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-716"><a href="#cb52-716" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-717"><a href="#cb52-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-718"><a href="#cb52-718" aria-hidden="true" tabindex="-1"></a><span class="fu"># Merging a Bin Size</span></span>
<span id="cb52-719"><a href="#cb52-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-720"><a href="#cb52-720" aria-hidden="true" tabindex="-1"></a>With our sample looking appropriate, we can now merge the samples with the same bin size into a single Seurat object. This can be done using Seurat's <span class="in">`merge()`</span> function.</span>
<span id="cb52-721"><a href="#cb52-721" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-722"><a href="#cb52-722" aria-hidden="true" tabindex="-1"></a><span class="fu">## Merging the 8μm Bin Size</span></span>
<span id="cb52-723"><a href="#cb52-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-724"><a href="#cb52-724" aria-hidden="true" tabindex="-1"></a>First, we will merge the 8μm bin size:</span>
<span id="cb52-725"><a href="#cb52-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-728"><a href="#cb52-728" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-729"><a href="#cb52-729" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: merge_8um</span></span>
<span id="cb52-730"><a href="#cb52-730" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-731"><a href="#cb52-731" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-732"><a href="#cb52-732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-733"><a href="#cb52-733" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge two sample together</span></span>
<span id="cb52-734"><a href="#cb52-734" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">merge</span>(fixed_8, fresh_8)</span>
<span id="cb52-735"><a href="#cb52-735" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-736"><a href="#cb52-736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-737"><a href="#cb52-737" aria-hidden="true" tabindex="-1"></a>However, when we use the <span class="in">`merge()`</span> function in Seurat, it keeps the counts matrices of previously <span class="in">`fixed_8`</span> and <span class="in">`fresh_8`</span> in separate layers. We would like them to be in the same layer for our analysis, so we will use the <span class="in">`JoinLayers()`</span> function from Seurat to do this:</span>
<span id="cb52-738"><a href="#cb52-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-741"><a href="#cb52-741" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-742"><a href="#cb52-742" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: joinlayers_8um</span></span>
<span id="cb52-743"><a href="#cb52-743" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-744"><a href="#cb52-744" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-745"><a href="#cb52-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-746"><a href="#cb52-746" aria-hidden="true" tabindex="-1"></a><span class="co"># Combines combines the count matrices from different layers into a single layer</span></span>
<span id="cb52-747"><a href="#cb52-747" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_8)</span>
<span id="cb52-748"><a href="#cb52-748" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-749"><a href="#cb52-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-750"><a href="#cb52-750" aria-hidden="true" tabindex="-1"></a>We would like to visualize our data to ensure that everything looks like it was done correctly, but in order to do this appropriately, we will need to normalize the data using Seuart's <span class="in">`NormalizeData()`</span> function:</span>
<span id="cb52-751"><a href="#cb52-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-754"><a href="#cb52-754" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-755"><a href="#cb52-755" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: normalizedata_8um</span></span>
<span id="cb52-756"><a href="#cb52-756" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-757"><a href="#cb52-757" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-758"><a href="#cb52-758" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-759"><a href="#cb52-759" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data for the merged 8μm bin size</span></span>
<span id="cb52-760"><a href="#cb52-760" aria-hidden="true" tabindex="-1"></a>merged_8 <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(merged_8)</span>
<span id="cb52-761"><a href="#cb52-761" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-762"><a href="#cb52-762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-763"><a href="#cb52-763" aria-hidden="true" tabindex="-1"></a>We can visualize a gene that we are potentially interested in "Cck" using <span class="in">`SpatialFeaturePlot()`</span>:</span>
<span id="cb52-764"><a href="#cb52-764" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-767"><a href="#cb52-767" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-768"><a href="#cb52-768" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: visualize_Cck_8um</span></span>
<span id="cb52-769"><a href="#cb52-769" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-770"><a href="#cb52-770" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-771"><a href="#cb52-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-772"><a href="#cb52-772" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Cck for our merged 8μm bin size</span></span>
<span id="cb52-773"><a href="#cb52-773" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(merged_8, </span>
<span id="cb52-774"><a href="#cb52-774" aria-hidden="true" tabindex="-1"></a>                   <span class="at">features =</span> <span class="st">"Cck"</span>, </span>
<span id="cb52-775"><a href="#cb52-775" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-776"><a href="#cb52-776" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-777"><a href="#cb52-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-778"><a href="#cb52-778" aria-hidden="true" tabindex="-1"></a><span class="fu">## Merging the 16μm Bin Size</span></span>
<span id="cb52-779"><a href="#cb52-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-780"><a href="#cb52-780" aria-hidden="true" tabindex="-1"></a>Now, we will merge the 16μm bin size following the same steps:</span>
<span id="cb52-781"><a href="#cb52-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-784"><a href="#cb52-784" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-785"><a href="#cb52-785" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: merge_16um</span></span>
<span id="cb52-786"><a href="#cb52-786" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-787"><a href="#cb52-787" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-788"><a href="#cb52-788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-789"><a href="#cb52-789" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge two sample together</span></span>
<span id="cb52-790"><a href="#cb52-790" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">merge</span>(fixed_16, fresh_16)</span>
<span id="cb52-791"><a href="#cb52-791" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-792"><a href="#cb52-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-793"><a href="#cb52-793" aria-hidden="true" tabindex="-1"></a>Again, we will use the <span class="in">`JoinLayers()`</span> to combine the counts matrices of the previously  <span class="in">`fixed_16`</span> and <span class="in">`fresh_16`</span> :</span>
<span id="cb52-794"><a href="#cb52-794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-797"><a href="#cb52-797" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-798"><a href="#cb52-798" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: joinlayers_16um</span></span>
<span id="cb52-799"><a href="#cb52-799" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-800"><a href="#cb52-800" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-801"><a href="#cb52-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-802"><a href="#cb52-802" aria-hidden="true" tabindex="-1"></a><span class="co"># Combines combines the count matrices from different layers into a single layer</span></span>
<span id="cb52-803"><a href="#cb52-803" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(merged_16)</span>
<span id="cb52-804"><a href="#cb52-804" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-805"><a href="#cb52-805" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-806"><a href="#cb52-806" aria-hidden="true" tabindex="-1"></a>We will normalize our 16μm data for visualization with Seuart's <span class="in">`NormalizeData()`</span> function:</span>
<span id="cb52-807"><a href="#cb52-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-810"><a href="#cb52-810" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-811"><a href="#cb52-811" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: normalizedata_16um</span></span>
<span id="cb52-812"><a href="#cb52-812" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-813"><a href="#cb52-813" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-814"><a href="#cb52-814" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-815"><a href="#cb52-815" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data for the merged 16μm bin size</span></span>
<span id="cb52-816"><a href="#cb52-816" aria-hidden="true" tabindex="-1"></a>merged_16 <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(merged_16)</span>
<span id="cb52-817"><a href="#cb52-817" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-818"><a href="#cb52-818" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-819"><a href="#cb52-819" aria-hidden="true" tabindex="-1"></a>We can visualize a gene that we are potentially interested in "Cck" using <span class="in">`SpatialFeaturePlot()`</span> on the 16μm data:</span>
<span id="cb52-820"><a href="#cb52-820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-823"><a href="#cb52-823" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb52-824"><a href="#cb52-824" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: visualize_Cck_16um</span></span>
<span id="cb52-825"><a href="#cb52-825" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb52-826"><a href="#cb52-826" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb52-827"><a href="#cb52-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-828"><a href="#cb52-828" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Cck for our merged 16μm bin size</span></span>
<span id="cb52-829"><a href="#cb52-829" aria-hidden="true" tabindex="-1"></a><span class="fu">SpatialFeaturePlot</span>(merged_16, </span>
<span id="cb52-830"><a href="#cb52-830" aria-hidden="true" tabindex="-1"></a>                   <span class="at">features =</span> <span class="st">"Cck"</span>, </span>
<span id="cb52-831"><a href="#cb52-831" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pt.size.factor =</span> <span class="dv">3</span>)</span>
<span id="cb52-832"><a href="#cb52-832" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb52-833"><a href="#cb52-833" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-834"><a href="#cb52-834" aria-hidden="true" tabindex="-1"></a>At this point we have successfully subsetted out data for two different bin sizes and from two separate samples while performing appropriate checks along the way to ensure that our subsetting has been done correctly.</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br> A portion of these materials and hands-on activities were adapted from the <a href="https://satijalab.org/">Satija Lab’s</a> <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html">Seurat - Guided Clustering Tutorial</a></p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>