---
title: "scRNA-seq Workflow"
authors: "Will Gammerdinger, Noor Sohail, Zhu Zhuo, James Billingsley, Shannan Ho Sui"
date: "Tuesday, July 22, 2025"
editor_options: 
  markdown: 
    wrap: 72
---

Approximate time: XY minutes

# Learning Objectives

-   LO 1
-   LO 2
-   LO 3

# Single-Cell RNAseq (scRNA-seq) Workflow

We will be peforming a standard scRNA-seq workflow on our sketch object. For more explicit details on each one of these steps, we have materials on how to conduct a [scRNA-seq analysis](https://hbctraining.github.io/Intro-to-scRNAseq-Quarto/).

Here we show the overarching

<img src="../img/sc_workflow_2022.jpg" width="630">


1. `FindVariableFeatures`: As before, this generates a list of highly variable genes, which may be slightly different for the sketched dataset than for the full dataset
2. `ScaleData`: Highly variable genes will be confounded with the most highly expressed genes, so we need to adjust for this
3. `RunPCA`: Perform a principal component analysis using our scaled data and variable genes. This will emphasize variation in gene expression as well as similarity across bins
4. `FindNeighbors`: Determine the Euclidean distance between bins in PCA space
5. `FindClusters`: Iteratively group bins together based on neighborhood distances. Higher resolution will yield more groups. 


## Sub-topic 1

## Sub-topic 2

## Sub-topic 3

::: callout-tip
# [**Exercise 1**](06_clustering_Answer-key.qmd#exercise-1)

Exercise 1
:::

# Topic 2

## Sub-topic 1

## Sub-topic 2

## Sub-topic 3

::: callout-tip
# [**Exercise 2**](06_clustering_Answer-key.qmd#exercise-2)

Exercise 1
:::

# Topic 3

## Sub-topic 1

## Sub-topic 2

## Sub-topic 3

::: callout-tip
# [**Exercise 3**](06_clustering_Answer-key.qmd#exercise-3)

Exercise 3
:::