---
title: "Integration"
authors: "Will Gammerdinger, Noor Sohail, Zhu Zhuo, James Billingsley, Shannan Ho Sui"
date: "Tuesday, July 22, 2025"
editor_options: 
  markdown: 
    wrap: 72
---

Approximate time: XY minutes

# Learning Objectives

-   When is it appropriate to integrate
-   LO 2
-   LO 3

# To Integrate or Not to Integrate

Generally, we always look at our clustering **without integration** before deciding whether we need to perform any alignment. It can be helpful to first run through clustering with samples from different sample classes together to see whether there are condition-specific clusters for cell types present in both conditions. Oftentimes, when clustering cells from multiple conditions there are condition-specific clusters and integration can help ensure the same cell types cluster together.

In this lesson, we will cover the integration of our samples across conditions, which is adapted from the [Seurat Guided Integration Tutorial](https://satijalab.org/seurat/articles/integration_introduction.html).

 **Do not just always perform integration because you think there might be differences - explore the data.** If we had performed the normalization on both conditions together in a Seurat object and visualized the similarity between cells, we would have seen in our dataset there is condition-specific clustering:

Condition-specific clustering of the cells indicates that we need to integrate the cells across conditions to ensure that cells of the same cell type cluster together. 
_**If cells cluster by sample, condition, batch, dataset, modality, performing integration can help align cells across the groups to greatly improve the clustering and the downstream analyses**._

```{r}
library(Seurat)
library(tidyverse)

seurat_processed <- readRDS("data/seurat_processed.RDS")
```

In our case, we have only one batch condition, which is our samples `P5CRC` and `P5NAT`. Here we assess both the PCA and the UMAP to see if there is a clear split based upon our batch variable.

```{r}
#| fig-width: 10
p1 <- DimPlot(seurat_processed,
              group.by="orig.ident",
              reduction = "full.pca.sketch") +
  NoLegend() + ggtitle("PCA (Projected)")

p2 <- DimPlot(seurat_processed,
              group.by="orig.ident",
              reduction = "full.umap.sketch") +
  ggtitle("UMAP (Projected)")

p1 | p2
```


**Why is it important that cells of the same cell type cluster together?** 

We want to identify  _**cell types which are present in all samples/conditions/modalities**_ within our dataset, and therefore would like to observe a representation of cells from both samples/conditions/modalities in every cluster. This will enable more interpretable results downstream (i.e. DE analysis, ligand-receptor analysis, differential abundance analysis, ...).

::: {.callout-note collapse=true}
# Alternative UMAP Visualization

Sometimes it can be helpful to look at the UMAP/PCA plots using the `split.by` parameter:

```{r}
DimPlot(seurat_processed, group.by = "orig.ident", split.by = "orig.ident" )
```
However it can be difficult to see where the cells are in one condition do not lay in the other condition. So what we can do is make our own custom ggplot where we color all the cells in the background as gray and overlay the condition cells on top. This makes it more clear where the differneces between conditions are.

```{r}
df <- FetchData(seurat_processed, c("fullumapsketch_1", 
                                    "fullumapsketch_2", 
                                    "orig.ident"))

# gray isn't working 
# TODO
p <- ggplot(df) +
  geom_point(aes(x=fullumapsketch_1, 
                 y=fullumapsketch_2),
             color = "lightgray", alpha=0.5, size = 0.1) +
  geom_point(aes(x=fullumapsketch_1, 
                 y=fullumapsketch_2, 
                 color=orig.ident),
             size = 0.1) +
  theme_void() +
  facet_wrap(~orig.ident)
p
```
:::


::: callout-tip
# [**Exercise 1**](07_clustering-quality-control_Answer-key.qmd#exercise-1)

1. Do you see a clear split quadrants of the UMAP and PCA? What is an alternative visualization that might make it easier to quantify the difference.
:::


# Integration Considerations

Now that we have thought some more about what could be causing our differences, we can start by assessing where the split in dataset originates. One of the clearest way is to take a look at the clusters that were calculated during the `FindClusters()` step of our workflow.

## Batch Distribution in Clusters

We do see a clear split in our dataset between our `NAT` and `CRC` samples. This is emphasized even more when we look at the distribution of cells in each cluster.

```{r}
#| label: cell_proportion_barplot
# Barplot of proportion of cells in each cluster by sample
ggplot(seurat_processed@meta.data) +
    geom_bar(aes(x=seurat_cluster.projected, fill=orig.ident), 
             position=position_fill())  +
    theme_classic()
```

## Possible Explanations

Before running any single-cell analysis, it is important to come in with some expectations of your cell population. Even better would be to make use of known marker genes curated from previous publications. Here, we have provided a few examples of genes that we would expect to represent some of the cell populations 

| Cell type     | Genes                                                                 |
|---------------|------------------------------------------------------------------------|
| Goblet cells  | CLCA1, FCGBP, MUC2                                                     |
| Goblet/Entero | PIGR                                                                   |
| Tumor         | CEACAM6, CEACAM5                                                       |
| Fibroblasts   | COL1A1                                                                 |
| Macrophages   | C1QC, SPP1, SELENOP                                                    |
| CAF           | COL1A1                                                                 |
| T cells       | TRAC, CD3E                                                             |
| Endothelial   | PECAM1                                                                 |
| Plasma cells  | IGKC                                                                   |


::: callout-tip
# [**Exercise 2**](07_clustering-quality-control_Answer-key.qmd#exercise-2)

Is there a population here that could potentially explain the differences between our samples?
:::

## Assessing Tumor Cells

We can use alternative visualizations to more clearly see

```{r}
#| fig-width: 15
celltype_markers <- list(
  goblet = c("CLCA1", "FCGBP", "MUC2"),
  goblet_enterocyte = c("PIGR"),
  tumor = c("CEACAM6", "CEACAM5"),
  fibroblast = c("COL1A1"),
  macrophage = c("C1QC", "SPP1", "SELENOP"),
  tcell = c("TRAC", "CD3E"),
  endothelial = c("PECAM1"),
  plasma = c("IGKC"))

p <- DotPlot(seurat_processed, celltype_markers)
p
```

```{r}
#| fig-width: 10
#| fig-height: 10
genes <- c("CLCA1", "FCGBP", "MUC2",
  "PIGR",
  "CEACAM6", "CEACAM5",
  "COL1A1",
  "C1QC", "SPP1", "SELENOP",
  "TRAC", "CD3E",
  "PECAM1",
  "IGKC")

p <- FeaturePlot(seurat_processed, genes, reduction="full.umap.sketch")
p
```

| Tumor Type | Genes              |
|------------|--------------------|
| Tumor I    | JCHAIN, RGMB       |
| Tumor II   | MUC17, ANPEP       |
| Tumor III  | DAB2, MME          |
| Tumor IV   | LYZ, IGHG1         |
| Tumor V    | MGST2, LRBA        |

```{r}
gene_sets <- list(
  Tumor_I   = c("JCHAIN", "RGMB"),
  Tumor_II  = c("MUC17", "ANPEP"),
  Tumor_III = c("DAB2", "MME"),
  Tumor_IV  = c("LYZ", "IGHG1"),
  Tumor_V   = c("MGST2", "LRBA")
)

DotPlot(seurat_processed, gene_sets)
```

```{r}
#| fig-width: 10
#| fig-height: 10
genes <- c("JCHAIN", "RGMB",
           "MUC17", "ANPEP",
           "DAB2", "MME",
           "LYZ", "IGHG1",
           "MGST2", "LRBA")

p <- DotPlot(seurat_processed, genes)
p
```



::: callout-tip
# [**Exercise 2**](07_clustering-quality-control_Answer-key.qmd#exercise-2)

Exercise 1
:::

# Integration Algorithms

While we may not be running integration on our own dataset, it is beneficial to understand some of the basic principles of the tools. Some of the most commonly used algorithms are described here:

::: {.panel-tabset}

## CCA

::: {.callout-note collapse=true}
# Code to run CCA
:::

## Harmony

::: {.callout-note collapse=true}
# Code to run Harmony
:::

## Mutual Nearest Neighbors

::: 

::: callout-tip
# [**Exercise 3**](07_clustering-quality-control_Answer-key.qmd#exercise-3)

Exercise 3
:::

