# Introduction to single-cell RNA-seq

## Learning Objectives

- LO 1
- LO 2
- LO 3
- LO 4

## Installations

1. [Follow the instructions linked here](../index.qmd#installation-requirements) to download R and RStudio + Install Packages from CRAN and Bioconductor

2. [Download this project]()**CREATE DROPBOX LINK STILL**

## Lessons

1. [Introduction to Spatial Single Cell RNA-sequencing](../lessons/01_introduction-to-spatial.qmd)
2. [Pre-processing of Spatial Data](../lessons/02_preprocessing-of-data.qmd)
3. [Quality Control](../lessons/03_quality-control.qmd)
4. [Theory of PCA](../lessons/04_theory_of_PCA.qmd)
5. [Normalization and regressing out unwanted variation](../lessons/05_normalization.qmd)
6. [Clustering](../lessons/06_clustering.qmd)
7. [Clustering quality control](../lessons/07_clustering-quality-control.qmd)
8. [Marker Identification](../lessons/08_marker-identification.qmd)
9. [Integration](../lessons/09_integration.qmd)
10. [Spatially Derived Value](../lessons/10_spatially-derived-values.qmd)
11. [Deconvolution](../lessons/11_deconvolution.qmd)
12. [Differential expression](../lessons/12_dge.qmd)
13. [Cell-Cell Comunication](../lessons/13_cell-cell-communication.qmd)

[Seurat Cheatsheet](../lessons/Cheatsheet_seurat.qmd)

# Additional Resources

## Building on this workshop

* Other online scRNA-seq courses:
  - [http://bioconductor.org/books/release/OSCA/](http://bioconductor.org/books/release/OSCA/)
  - [https://liulab-dfci.github.io/bioinfo-combio/](https://liulab-dfci.github.io/bioinfo-combio/)
  - [https://hemberg-lab.github.io/scRNA.seq.course/](https://hemberg-lab.github.io/scRNA.seq.course/)
  - [https://github.com/SingleCellTranscriptomics](https://github.com/SingleCellTranscriptomics)
  - [https://broadinstitute.github.io/2020_scWorkshop/](https://broadinstitute.github.io/2020_scWorkshop/)

* Resources for scRNA-seq Sample Prep:
  - [https://www.protocols.io/](https://www.protocols.io/)
  - [https://support.10xgenomics.com/single-cell-gene-expression/sample-prep](https://support.10xgenomics.com/single-cell-gene-expression/sample-prep)
  - [https://community.10xgenomics.com/](https://community.10xgenomics.com/)

***

## Resources
We have covered the analysis steps in quite a bit of detail for scRNA-seq exploration of cellular heterogeneity using the Seurat package. For more information on topics covered, we encourage you to take a look at the following resources:

* [Seurat vignettes](https://satijalab.org/seurat/vignettes.html)
* [Seurat cheatsheet](https://satijalab.org/seurat/essential_commands.html)
* [Satija Lab: Single Cell Genomics Day](https://satijalab.org/scgd21/)
* ["Principal Component Analysis (PCA) clearly explained"](https://www.youtube.com/watch?v=_UVHneBUBW0), a video from [Josh Starmer](https://twitter.com/joshuastarmer)
* [Additional information about cell cycle scoring](../lessons/Aside_cell_cycle_scoring.qmd)
* [Using R on the O2 cluster](https://hbctraining.github.io/Intro-to-Unix-QMB/lessons/R_on_o2.html)
* Highlighted papers for sample processing steps (pre-sequencing):
  - ["Sampling time-dependent artifacts in single-cell genomics studies."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02032-0) *Massoni-Badosa et al.* 2019
  - ["Dissociation of solid tumor tissues with cold active protease for single-cell RNA-seq minimizes conserved collagenase-associated stress responses."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1830-0) *O'Flanagan et al.* 2020
  - ["Systematic assessment of tissue dissociation and storage biases in single-cell and single-nucleus RNA-seq workflows."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02048-6) *Denisenko et al.* 2020
* [Best practices for single-cell analysis across modalities](https://www.nature.com/articles/s41576-023-00586-w)
