---
title: Introduction to Spatial Single-cell RNA-seq Schedule
sidebar: false
toc: false
execute: 
  eval: false
---

## Pre-reading

- [Pre-reading 1](../lessons/Pre-reading_01.qmd)
- [Pre-reading 2](../lessons/Pre-reading_02.qmd)
- [Pre-reading 3](../lessons/Pre-reading_03.qmd)

## Day 1

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 09:30 - 09:45 | [Workshop introduction](../slides/workshop_intro_slides.pdf) | Will |
| 09:45 - 10:45| [Introduction to Spatial Single Cell RNA-sequencing](../lessons/01_introduction-to-spatial.qmd) | [Dr. Mandovi Chatterjee](https://singlecellcore.hms.harvard.edu/people/mandovi-chatterjee) |
| 10:45 - 10:50 | Break |
| 10:50 - 11:15 | scRNA-seq pre-reading discussion | All |
| 11:00 - 11:45 | [Pre-processing of Spatial Data](../lessons/02_preprocessing-of-data.qmd) | Noor |
| 11:45 - 12:00 | Overview of self-learning materials and homework submission | Will |

### Before the next class:

I. Please **study the contents** and **work through all the code** within the following lessons:
 
   1. [Quality Control](../lessons/03_quality-control.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Brief description of the lesson<br><br>In this lesson you will:<br>
             - LO 1<br>
             - LO 2<br>
             - LO 3<br><br>
        </details>

  2. [Theory of PCA](../lessons/04_theory_of_PCA.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>Before we can begin the next steps of the workflow, we need to make sure you have a good understanding of Principal Components Analysis (PCA). This method will be utilized in the scRNA-seq analysis workflow, and this foundation will help you better navigate those steps and interpretation of results.<br><br>
        </details>
     

II. **Submit your work**:
   * Each lesson above contains exercises; please go through each of them.
   * **Submit your answers** to the exercises using [this Google form]()(**NEED LINK STILL**) on **the day *before* the next class**.


### Questions?
* ***If you get stuck due to an error*** while running code in the lesson, [email us](mailto:hbctraining@hsph.harvard.edu) 

***

## Day 2

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 09:30 - 10:00 | Self-learning lessons discussion | All |
| 10:00 - 11:00| [Normalization and Regressing Out Unwanted Variation](../lessons/05_normalization.qmd) | Noor |
| 11:00 - 11:05 | Break |
| 11:05 - 12:00| [Clustering](../lessons/06_clustering.qmd) | Will |

### Before the next class:
I. Please **study the contents** and **work through all the code** within the following lessons:

1. [Clustering Quality Control](../lessons/07_clustering-quality-control.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>After separating cells into clusters, it is crtical to evaluate whether they are biologically meaningful or not. At this point we can also decide if we need to re-cluster and/or potentialy go back to a previous QC step.
         <br><br>In this lesson you will:<br>
           - Check to see that clusters are not influenced by uninteresting sources of variation<br>
           - Check to see whether the major principal components are driving the different clusters<br>
           - Explore the cell type identities by looking at the expression for known markers across the clusters.<br>
        </details>

2. [Marker Identification](../lessons/08_marker-identification.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Brief description of the lesson<br><br>In this lesson you will:<br>
             - LO 1<br>
             - LO 2<br>
             - LO 3<br><br>
        </details>
        
3. [Integration](../lessons/09_integration.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Brief description of the lesson<br><br>In this lesson you will:<br>
             - LO 1<br>
             - LO 2<br>
             - LO 3<br><br>
        </details>


II. **Submit your work**:
   * Each lesson above contains exercises; please go through each of them.
   * **Submit your answers** to the exercises using [this Google form]()(**NEED LINK STILL**) on **the day *before* the next class**.

### Questions?
* ***If you get stuck due to an error*** while running code in the lesson, [email us](mailto:hbctraining@hsph.harvard.edu) 


***


## Day 3

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 9:30 - 10:15 | Self-learning lessons discussion | All |
| 10:15 - 11:00 | [Spatially Derived Values](../lessons/10_spatially-derived-values.qmd) | Noor |
| 11:00 - 11:05 | Break |
| 11:05 - 12:00 | [Deconvolution](../lessons/11_deconvolution.qmd) | Will |

### Before the next class:
I. Please **study the contents** and **work through all the code** within the following lessons:

1. [Differential Gene Expression](../lessons/12_dge.qmd)
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
       <br>Brief description of the lesson<br><br>In this lesson you will:<br>
             - LO 1<br>
             - LO 2<br>
             - LO 3<br><br>
        </details>

2. [Seurat Cheatsheet](../lessons/Cheatsheet_seurat.qmd) 
      <details>
       <summary><i>Click here for a preview of this lesson</i></summary>
         <br>At this point, we have populated our seurat object with many different pieces of information. Knowing how to access different values will allow you to interact more efficiently with your dataset.
         <br><br>In this lesson you will:<br>
           - Explore the different parts of a seurat object.<br>
           - Use the built-in functions from the Seurat package for visualizations and grabbing data.<br>
        </details>

***

## Day 4

| Time |  Topic  | Instructor |
|:-----------:|:----------:|:--------:|
| 9:30 - 10:15 | Self-learning lessons discussion | All |
| 10:15 - 11:15 | [Cell-Cell Comunication](../lessons/13_cell-cell-communication.qmd) | Noor |
| 11:15 - 11:20 | Break |
| 11:20 - 11:45 | Overview and Final Q & A | All |
| 11:45- 12:00 | [Wrap up](../slides/Workshop_wrapup.pdf) | Will |


***

## Resources
We have covered the analysis steps in quite a bit of detail for scRNA-seq exploration of cellular heterogeneity using the Seurat package. For more information on topics covered, we encourage you to take a look at the following resources:

### Seurat-focused
* [Seurat vignettes](https://satijalab.org/seurat/vignettes.html)
* [Seurat cheatsheet](https://satijalab.org/seurat/essential_commands.html)
* [Satija Lab: Single Cell Genomics Day](https://satijalab.org/scgd21/)
* [Additional information about cell cycle scoring](../lessons/Aside_cell_cycle_scoring.qmd)
* [A nice explanation on CCA and cell label transfer in Seurat](https://divingintogeneticsandgenomics.com/post/cca-alignment/)

### Scaling up: scRNA-seq analysis on HPC  
* Using RStudio on O2
    * [HMSRC wiki page](https://harvardmed.atlassian.net/wiki/spaces/O2/pages/1623425967/RStudio+on+O2)
    * [HBC RStudio on O2 tutorial](https://hbctraining.github.io/Intro-to-Unix-QMB/lessons/R_studio_on_02.html)

### Cell type annotation
- Databases with markers for manual annotation
  - [CellMarker 2.0](http://bio-bigdata.hrbmu.edu.cn/CellMarker/)
  - Cell type signature gene sets from [MSigDb](https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=C8)
  - [CELL x GENE from CZI](https://cellxgene.cziscience.com/gene-expression)
- Reference-based automated celltype annotation
  - [Azimuth](https://azimuth.hubmapconsortium.org/)
  - [Celltypist](https://www.celltypist.org/)
 
   
### Highlighted papers

- ["Sampling time-dependent artifacts in single-cell genomics studies."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02032-0) *Massoni-Badosa et al.* 2019
- ["Dissociation of solid tumor tissues with cold active protease for single-cell RNA-seq minimizes conserved collagenase-associated stress responses."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1830-0) *O'Flanagan et al.* 2020
- ["Systematic assessment of tissue dissociation and storage biases in single-cell and single-nucleus RNA-seq workflows."](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02048-6) *Denisenko et al.* 2020
- ["Confronting false discoveries in single-cell differential expression", _Nature Communications_ 2021](https://www.nature.com/articles/s41467-021-25960-2)
- [Single-nucleus and single-cell transcriptomes compared in matched cortical cell types](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6306246/)
- [A single-cell and single-nucleus RNA-Seq toolbox for fresh and frozen human tumors](https://www.nature.com/articles/s41591-020-0844-1)
- [Ligand-receptor analysis with CellphoneDB](https://www.nature.com/articles/s41576-020-00292-x)
- [Best practices for single-cell analysis across modalities](https://www.nature.com/articles/s41576-023-00586-w)



### Other online scRNA-seq courses:
  - [OSCA with Bioconductor](http://bioconductor.org/books/release/OSCA/)
  - [DFCI/Shirley Liu](https://liulab-dfci.github.io/bioinfo-combio/)
  - [Wellcome Sanger Institute/Hemmberg Lab](https://www.singlecellcourse.org/)
  - [ISCB Workshop](https://github.com/SingleCellTranscriptomics)
  - [Broad workshop](https://broadinstitute.github.io/2020_scWorkshop/)
  - [SciLifeLab workshop](https://nbisweden.github.io/workshop-scRNAseq/)
